---
title: "Quantitative genetic parameters, univariate models"
output: 
  github_document:
    toc: true
    df_print: kable
---

I experimentally selected tapeworms (*S. solidus*) for faster and slower development in their intermediate host. Throughout the multigeneration experiment, I tracked the parasite pedigree. Here, I estimate quantitative genetic parameters like heritability and maternal variance one trait at a time.

```{r setup, include=FALSE}
library(tidyverse)
library(MCMCglmm)
knitr::opts_chunk$set(warning = FALSE, message = FALSE, echo = FALSE)
options(stringsAsFactors = FALSE)
theme_set(theme_bw())
theme_update(panel.grid.minor = element_blank())
```

```{r}
load("02testing_selection_response.RData")
```
```{r}
rm(list = ls()[grep("m1|dat", ls(), invert = T)]) # remove everything that does not start with m1 or dat
```

```{r}
dat <- read.csv(file="../data/sel_exp_data.csv", header = TRUE)
```
```{r}
#Make some centered variables...
dat <- dat%>%
  mutate(cop_length_cen = scale(cop_length_mm, scale = FALSE),
         fish_initial_tl_cen = scale(fish_initial_tl, scale = FALSE),
         fish_initial_bm_cen = scale(fish_initial_bm_g, scale = FALSE),
         age_diss_cen = scale(age_dissection, scale = FALSE),
         fish_sex_cen = fish_sex - 0.5,
         proc_size2 = procercoid_size/1000, # avoid numerical problems
         proc_size_std = scale(procercoid_size, scale = TRUE),
         worm_bm_mg_std = scale(worm_bm_mg, scale = TRUE),
         tank_id = if_else(tank_id %in% c("_", "0_","1_","2_", "3_", "4_"), NA_character_, tank_id),
         gen_line = if_else(line !="",paste(gen, line, sep = "_"), NA_character_))%>%
  mutate(gen_line = factor(gen_line, levels = c("0_base", "1_control", "1_fast", "1_slow",
                                                "2_control", "2_fast", "2_slow", 
                                                "3_control", "3_fast", "3_slow",
                                                "4_control", "4_fast", "4_slow")))
```

For each parasite trait, we already fit a basic model that tried to account for nuisance variables, like inbreeding, copepod stage, or fish size (see [here](02testing_selection_response.md)). To this model, we added the selection line to look for correlated responses to selection. Another way to asses genetic variance and covariance among traits is to add genetic information to the model, such as full-sib family or even the full pedigree.

For each trait, we consider three additional models beyond the 'base' model accounting for things like starting condition (e.g. copepod stage, fish size) and shared environments (e.g. block, tank id). 

First, we add full-sib family to the model. Significant among-family variance suggests genetic variance in the trait. However, full-sib heritabilities can be biased upwards by dominance effects and shared parental effects (maternal but also paternal in this simultaneous hermaphrodite). Therefore, as a second model, we replace the random effect of 'family' with the full pedigree. This is the so-called animal model, and it uses all information on relatedness in the experiment to estimate the additive genetic variance. Finally, within full-sib families, parasites differ in which worm was the sire and dam. If some worms differentially invest in their offspring, then some of the variance within families could be accounted for by such 'parental' effects.

```{r}
# full pedigree and relatedness matrix
ped <- select(dat, id, dam, sire)%>%
  mutate_all(as.factor)

Ainv <- inverseA(ped)$Ainv
ib <- inverseA(ped)$inbreeding
```

# Traits in copepods

## Copepod infection rate

Parent-offspring regression does not work for survival (all parents survived to reproduce, so there is no variation in x-axis variable.)

```{r}
parents <- dat_cinf%>%
  group_by(eggs_clutch_id)%>%
  summarise(n_parent = n(), cop_inf_p = mean(cop_inf))%>%
  na.omit()

offspring <- dat_cinf%>%
  group_by(worm_full_sib_fam_id)%>%
  summarise(n_offspring = n(), cop_inf_o = mean(cop_inf, na.rm = T))%>%
  na.omit()

parent_offspring <- left_join(parents, offspring, by = c("eggs_clutch_id" = "worm_full_sib_fam_id"))
```

Here are the number of full-sib families and the number of offspring phenotyped per family.

```{r}
dat_cinf%>%
  group_by(worm_full_sib_fam_id)%>%
  summarize(n_pheno = sum(!is.na(cop_inf)))%>%
  ungroup()%>%
  filter(!is.na(worm_full_sib_fam_id))%>%
  summarize(fs_fams = n(),
            min_phenotyped = min(n_pheno),
            lwrIQR_phenotyped = quantile(n_pheno, prob = 0.25),
            avg_phenotyped = mean(n_pheno),
            uprIQR_phenotyped = quantile(n_pheno, prob = 0.75),
            max_phenotyped = max(n_pheno)
            )
```

Here are the number of offspring phenotyped per dam/sire.

```{r}
dat_cinf%>%
  group_by(maternal_match, worm_maternal_id)%>%
  summarize(n_pheno = sum(!is.na(cop_inf)))%>%
  group_by(maternal_match)%>%
  filter(!is.na(worm_maternal_id))%>%
  summarize(dams = n(),
            min_phenotyped = min(n_pheno),
            lwrIQR_phenotyped = quantile(n_pheno, prob = 0.25),
            avg_phenotyped = mean(n_pheno),
            uprIQR_phenotyped = quantile(n_pheno, prob = 0.75),
            max_phenotyped = max(n_pheno)
            )
```

```{r}
target_samples <- 2500
```

```{r}
p0_bi <- list(R = list(V = 1, fix = 1))

p1_bi <- list(R = list(V = 1, fix = 1),
              G = list(G1 = list(V = 1, nu = 1000, alpha.mu = 0, alpha.V = 1)))

p2_bi <- list(R = list(V = 1, fix = 1),
              G = list(G1 = list(V = 1, nu = 1000, alpha.mu = 0, alpha.V = 1),
                       G2 = list(V = 1, nu = 1000, alpha.mu = 0, alpha.V = 1)))


thin_cinf <- 100
burnin_cinf <- 1000
nit_cinf <- thin_cinf * target_samples + burnin_cinf


# fixed effects
m1_cinf_noib <- MCMCglmm(cop_inf ~ cop_stage_checking, 
                 random = ~cop_block,
                 family = "threshold",
                 data = dat_cinf,
                 prior = p1_bi,
                 verbose=FALSE,
                 nitt = nit_cinf, 
                 thin = thin_cinf,
                 burnin = burnin_cinf
                )
# fixed effects
m1_cinf <- MCMCglmm(cop_inf ~ ib + cop_stage_checking, 
                 random = ~cop_block,
                 family = "threshold",
                 data = dat_cinf,
                 prior = p1_bi,
                 verbose=FALSE,
                 nitt = nit_cinf, 
                 thin = thin_cinf,
                 burnin = burnin_cinf
                )
# full-sibs
m2_cinf_fam <- MCMCglmm(cop_inf ~ ib + cop_stage_checking, 
                 random = ~ cop_block + worm_full_sib_fam_id,
                 family = "threshold",
                 data = dat_cinf,
                 prior = p2_bi,
                 verbose=FALSE,
                 pr = T,
                 nitt = nit_cinf, 
                 thin = thin_cinf,
                 burnin = burnin_cinf
                )

# # full-sibs
# m2_cinf_fam_logistic <- MCMCglmm(cop_inf ~ ib + cop_stage_checking, 
#                  random = ~ cop_block + worm_full_sib_fam_id,
#                  family = "categorical",
#                  data = dat_cinf,
#                  prior = p2_bi,
#                  verbose=F,
#                  nitt = nit_cinf, 
#                  thin = thin_cinf,
#                  burnin = burnin_cinf
#                 )
# full sib x line
m2_cinf_famXline <- MCMCglmm(
  cop_inf ~ ib + cop_stage_checking, 
  random = ~ cop_block + idh(line):worm_full_sib_fam_id,
  family = "threshold",
  data = dat_cinf,
  prior = 
    list(R = list(V = 1, fix = 1),
         G = list(G1 = list(V = 1, nu = 1000, alpha.mu = 0, alpha.V = 1),
                  G2 = list(V = diag(4), nu = 0.002))),
  verbose=FALSE,
  pr = F,
  nitt = nit_cinf, 
  thin = thin_cinf,
  burnin = burnin_cinf
  )

# anim model
m2_cinf_anim <- MCMCglmm(cop_inf ~ ib + cop_stage_checking, 
                 random = ~cop_block + id,
                 ginverse = list(id = Ainv),
                 family = "threshold",
                 data = dat_cinf,
                 prior = p2_bi,
                 verbose=F,
                 nitt = nit_cinf, 
                 thin = thin_cinf,
                 burnin = burnin_cinf
                )
```

Adding full-sibs (red) is an improvement, and the animal model (green) looks like a further improvement.

```{r}
plot(mcmc.list(
  m1_cinf$Deviance, m2_cinf_fam$Deviance, m2_cinf_anim$Deviance
  ), density = F)
```

Here is the full sib model. Full sib explains more variation than experimental block.

```{r}
summary(m2_cinf_fam)
```

Here is the animal model.

```{r}
summary(m2_cinf_anim)
```

For, these models, we break down total variance into several components: fixed effects (starting conditions), shared environment, and genetic effects.

```{r}
# FIXED EFFECT VARIANCE
# manual prediction, correcting for set variance comp
c2 <- ((16 * sqrt(3))/(15 * pi))^2

# first model
pdx <- m1_cinf$X # model matrix, fixed effx
n <- dim(pdx)[1] # number of data points in model, all traits
nt <- n/1 # number of data points per trait
p_i <- which(is.na(dat_cinf$pred)) # points where we can predicted vals and cred int
pdx <- pdx[c(p_i),] # restrict to only points where we want preds
p_n <- dim(pdx)[1]/1 # number of points for each trait we want to predict
num_fe <- m1_cinf$Fixed$nfl # number of fixed effx, rand effx marginalized
p_all <- as.matrix(pdx) %*% t(m1_cinf$Sol[,1:num_fe]) # predicteds via matrix mult 
m1_Vf <- apply(p_all, MARGIN = 2, var)

# first model no ib
pdx <- m1_cinf_noib$X # model matrix, fixed effx
n <- dim(pdx)[1] # number of data points in model, all traits
nt <- n/1 # number of data points per trait
p_i <- which(is.na(dat_cinf$pred)) # points where we can predicted vals and cred int
pdx <- pdx[c(p_i),] # restrict to only points where we want preds
p_n <- dim(pdx)[1]/1 # number of points for each trait we want to predict
num_fe <- m1_cinf_noib$Fixed$nfl # number of fixed effx, rand effx marginalized
p_all <- as.matrix(pdx) %*% t(m1_cinf_noib$Sol[,1:num_fe]) # predicteds via matrix mult 
m1_Vf_noib <- apply(p_all, MARGIN = 2, var)

# post_sol <- m2_cinf_fam_logistic$Sol/sqrt(1 + c2 * m2_cinf_fam_logistic$VCV[, 'units']) # correction suggested in course notes
# pdx <- m2_cinf_fam_logistic$X # model matrix, fixed effx
# n <- dim(pdx)[1] # number of data points in model, all traits
# nt <- n/1 # number of data points per trait
# p_i <- which(is.na(dat_cinf$pred)) # points where we can predicted vals and cred int
# pdx <- pdx[c(p_i),] # restrict to only points where we want preds
# p_n <- dim(pdx)[1]/1 # number of points for each trait we want to predict
# num_fe <- m2_cinf_fam_logistic$Fixed$nfl # number of fixed effx, rand effx marginalized
# p_all <- as.matrix(pdx) %*% t(post_sol[,1:num_fe]) # predicteds via matrix mult 
# m2_Vf_fam_logistic <- apply(p_all, MARGIN = 2, var)

pdx <- m2_cinf_fam$X # model matrix, fixed effx
n <- dim(pdx)[1] # number of data points in model, all traits
nt <- n/1 # number of data points per trait
p_i <- which(is.na(dat_cinf$pred)) # points where we can predicted vals and cred int
pdx <- pdx[c(p_i),] # restrict to only points where we want preds
p_n <- dim(pdx)[1]/1 # number of points for each trait we want to predict
num_fe <- m2_cinf_fam$Fixed$nfl # number of fixed effx, rand effx marginalized
p_all <- as.matrix(pdx) %*% t(m2_cinf_fam$Sol[,1:num_fe]) # predicteds via matrix mult 
m2_Vf_fam <- apply(p_all, MARGIN = 2, var)

pdx <- m2_cinf_anim$X # model matrix, fixed effx
n <- dim(pdx)[1] # number of data points in model, all traits
nt <- n/1 # number of data points per trait
p_i <- which(is.na(dat_cinf$pred)) # points where we can predicted vals and cred int
pdx <- pdx[c(p_i),] # restrict to only points where we want preds
p_n <- dim(pdx)[1]/1 # number of points for each trait we want to predict
num_fe <- m2_cinf_anim$Fixed$nfl # number of fixed effx, rand effx marginalized
p_all <- as.matrix(pdx) %*% t(m2_cinf_anim$Sol[,1:num_fe]) # predicteds via matrix mult 
m2_Vf_anim <- apply(p_all, MARGIN = 2, var)
```
```{r}
# TOTAL PHENOTYPIC VARIANCE
m1_Vp <- m1_Vf + rowSums(m1_cinf$VCV)
m1_Vp_noib <- m1_Vf_noib + rowSums(m1_cinf_noib$VCV)
m2_Vp_fam <- m2_Vf_fam + rowSums(m2_cinf_fam$VCV)
# m2_Vp_fam_logistic <- m2_Vf_fam_logistic + rowSums(m2_cinf_fam_logistic$VCV) + (pi^2)/3
m2_Vp_anim <- m2_Vf_anim + rowSums(m2_cinf_anim$VCV)
```
```{r}
# vc for model 1, no ib
m1_vc_noib <- data.frame(mod = "1a_base_noIB", 
                    v_est = c(
                      m1_Vf_noib/m1_Vp_noib, 
                      m1_cinf_noib$VCV[,'cop_block']/m1_Vp_noib,
                      m1_cinf_noib$VCV[,'cop_block']/(m1_Vp_noib - m1_Vf_noib),
                      1-(m1_Vf_noib+m1_cinf_noib$VCV[,'cop_block'])/m1_Vp_noib,
                      1-(m1_cinf_noib$VCV[,'cop_block'])/(m1_Vp_noib - m1_Vf_noib)
                      ),
                    v_name = rep(c("Vf", "Ve", "Ve2", "Vr", "Vr2"), 
                                 times = rep(length(m1_Vf), 5))
                    )
# vc for model 1
m1_vc <- data.frame(mod = "1b_base", 
                    v_est = c(
                      m1_Vf/m1_Vp, 
                      m1_cinf$VCV[,'cop_block']/m1_Vp,
                      m1_cinf$VCV[,'cop_block']/(m1_Vp - m1_Vf),
                      1-(m1_Vf+m1_cinf$VCV[,'cop_block'])/m1_Vp,
                      1-(m1_cinf$VCV[,'cop_block'])/(m1_Vp - m1_Vf)
                      ),
                    v_name = rep(c("Vf", "Ve", "Ve2", "Vr", "Vr2"), 
                                 times = rep(length(m1_Vf), 5))
                    )

# vc for model 2 full sib
m2_vc <- data.frame(mod = "2a_full_sibs", 
                    v_est = c(
                      m2_Vf_fam/m2_Vp_fam, 
                      m2_cinf_fam$VCV[,'cop_block']/m2_Vp_fam, 
                      m2_cinf_fam$VCV[,'cop_block']/(m2_Vp_fam - m2_Vf_fam), 
                      2*m2_cinf_fam$VCV[,'worm_full_sib_fam_id']/m2_Vp_fam, 
                      2*m2_cinf_fam$VCV[,'worm_full_sib_fam_id']/(m2_Vp_fam - m2_Vf_fam),
                      1-(m2_Vf_fam+m2_cinf_fam$VCV[,'cop_block'] +
                           2*m2_cinf_fam$VCV[,'worm_full_sib_fam_id'])/m2_Vp_fam,
                      1-(m2_cinf_fam$VCV[,'cop_block'] +
                           2*m2_cinf_fam$VCV[,'worm_full_sib_fam_id'])/(m2_Vp_fam - m2_Vf_fam)
                      ),
                    v_name = rep(c("Vf", "Ve", "Ve2", "Vg", "Vg2", "Vr", "Vr2"), 
                            times = rep(length(m2_Vf_fam), 7))
                    )

# vc for model 2 anim
m2_vc_anim <- data.frame(mod = "2b_animal", 
                    v_est = c(
                      m2_Vf_anim/m2_Vp_anim, 
                      m2_cinf_anim$VCV[,'cop_block']/m2_Vp_anim, 
                      m2_cinf_anim$VCV[,'cop_block']/(m2_Vp_anim - m2_Vf_anim), 
                      m2_cinf_anim$VCV[,'id']/m2_Vp_anim, 
                      m2_cinf_anim$VCV[,'id']/(m2_Vp_anim - m2_Vf_anim),
                      1-(m2_Vf_anim+m2_cinf_anim$VCV[,'cop_block'] +
                           m2_cinf_anim$VCV[,'id'])/m2_Vp_anim,
                      1-(m2_cinf_anim$VCV[,'cop_block'] +
                           m2_cinf_anim$VCV[,'id'])/(m2_Vp_anim - m2_Vf_anim)
                      ),
                    v_name = rep(c("Vf", "Ve", "Ve2", "Vg", "Vg2", "Vr", "Vr2"), 
                            times = rep(length(m2_Vf_anim), 7))
                    )


m_vc <- bind_rows(
  m1_vc_noib, m1_vc,
  m2_vc, m2_vc_anim
)


m_vc_cinf <- m_vc%>%
  group_by(mod, v_name)%>%
  summarise(v_fit = median(v_est),
            v_lwr = quantile(v_est, probs = 0.025),
            v_upr = quantile(v_est, probs = 0.975),
            )%>%
  ungroup()%>%
  mutate(trait = "cinf")

m_vc_cinf2 <- m_vc_cinf%>%filter(v_name %in% c("Ve2", "Vg2", "Vm2", "Vr2")) # var with fixed effects
m_vc_cinf <- m_vc_cinf%>%filter(!v_name %in% c("Ve2", "Vg2", "Vm2", "Vr2")) # var excluding fixed effects
```

Here is the percent variation in copepod infection accounted for by copepod stage in the three models. It is consistently low.

```{r}
filter(m_vc_cinf, v_name == "Vf")%>%
  mutate(across(v_fit:v_upr, function(x){round(x,3)}))
```

Here is the percent variation in copepod infection accounted for by shared environments (copepod block). Some of the environmental variation in the base model is attributed to genes after adding full-sibs or the pedigree.

```{r}
filter(m_vc_cinf, v_name == "Ve")%>%
  mutate(across(v_fit:v_upr, function(x){round(x,3)}))
```

Here is the percent variation in copepod infection accounted for by genetic variance (i.e. the heritability). Note that the variance component for full-sibs was multipled by two since full-sibs share just half their alleles on average. The estimate from the animal model was quite a bit higher.

```{r}
filter(m_vc_cinf, v_name == "Vg")%>%
  mutate(across(v_fit:v_upr, function(x){round(x,3)}))
```

Here is the distribution of family-level random effects. It is a little left-skewed. This suggests the genetic variance represents primarily represents some families having lower than expected infection rates. 

```{r}
mean_cinf <- sum(dat_cinf$cop_inf, na.rm = T)/sum(!is.na(dat_cinf$cop_inf), na.rm = T)
re_fams_cinf <- dat_cinf%>%
  filter(is.na(pred))%>%
  group_by(worm_full_sib_fam_id)%>%
  summarize(fam_avg = qnorm(sum(cop_inf, na.rm = T)/sum(!is.na(cop_inf), na.rm = T)) - qnorm(mean_cinf))%>%
  ungroup()%>%
  mutate(approach="unadjusted")

re <- colMeans(m2_cinf_fam$Sol[,which(grepl("worm_full_sib", colnames(m2_cinf_fam$Sol)))])
re <- data.frame(worm_full_sib_fam_id = gsub("worm_full_sib_fam_id.", "", x = names(re)),
                 fam_avg = re,
                 approach="model-estimated")

re_fams_cinf <- bind_rows(
  re_fams_cinf,
  re
)%>%
  mutate(approach = fct_relevel(factor(approach), c("unadjusted", "model-estimated")))

ggplot(re_fams_cinf, aes(x = fam_avg)) +
  geom_histogram(aes(y=..density..),
                 colour = "black",
                 fill = "white") +
  geom_function(data = filter(re_fams_cinf, approach == "unadjusted"),
                fun = dnorm, 
                args = list(mean = 0, 
                            sd = re_fams_cinf%>%
                              filter(approach == "unadjusted", !is.infinite(fam_avg))%>%.$fam_avg%>%sd(., na.rm=T)
                            ),
                color = 'red',
                linetype = "dashed") +
  geom_function(data = filter(re_fams_cinf, approach == "model-estimated"),
                fun = dnorm, 
                args = list(mean = 0, 
                            sd = re_fams_cinf%>%filter(approach == "model-estimated")%>%.$fam_avg%>%sd(., na.rm=T)
                            ),
                color = 'red',
                linetype = "dashed") +
  geom_density() +
  geom_label(data = re_fams_cinf%>%
               filter(!is.infinite(fam_avg))%>%
               group_by(approach)%>%
               summarise(min_x = min(fam_avg, na.rm = T),
                         skew = e1071::skewness(fam_avg, na.rm = T))%>%
               mutate(skew_c = paste0("Skewness: ", round(skew, 2))),
             aes(label = skew_c,
                 x = min_x,
                 y = 0.5),
             hjust = -0.05, vjust = 0) +
  facet_wrap(~approach, scales = "free_x") +
  labs(y = "Density", x = "Family averages (probit transformed)")
```

Here is the total variance explained by each model.

```{r}
filter(m_vc_cinf, v_name == "Vr")%>%
  mutate(across(v_fit:v_upr, function(x){round(1-x,3)}))
```

The animal model could be biased upwards by maternal effects. Let's now account for parental influences. We again fit the full-sib and animal models, but we add a random effect of dam/sire.

```{r}
p3_bi <- list(R = list(V = 1, fix = 1),
              G = list(G1 = list(V = 1, nu = 1000, alpha.mu = 0, alpha.V = 1),
                       G2 = list(V = 1, nu = 1000, alpha.mu = 0, alpha.V = 1),
                       G3 = list(V = 1, nu = 1000, alpha.mu = 0, alpha.V = 1)))


# full-sibs + mat effx
m3_cinf_mat <- MCMCglmm(cop_inf ~ ib + cop_stage_checking, 
                 random = ~cop_block + worm_full_sib_fam_id + dam,
                 family = "threshold",
                 data = dat_cinf%>%mutate(dam = factor(dam)),
                 prior = p3_bi,
                 verbose=F,
                 pr=T,
                 nitt = nit_cinf, 
                 thin = thin_cinf,
                 burnin = burnin_cinf
                )
# anim model
m3_cinf_anim <- MCMCglmm(cop_inf ~ ib + cop_stage_checking, 
                 random = ~cop_block + id  + dam,
                 ginverse = list(id = Ainv),
                 family = "threshold",
                 data = dat_cinf%>%mutate(dam = factor(dam)),
                 prior = p3_bi,
                 verbose=F,
                 nitt = nit_cinf, 
                 thin = thin_cinf,
                 burnin = burnin_cinf
                )
```

The full-sib model suggests that some variance within clutches is explained by parental effects.

```{r}
plot(mcmc.list(
  m2_cinf_fam$Deviance, m3_cinf_mat$Deviance
  ), density = F)
```

By contrast, the addition of maternal effects seems to worsen the animal model.

```{r}
plot(mcmc.list(
  m2_cinf_anim$Deviance, m3_cinf_anim$Deviance
  ), density = F)
```

The additive genetic variance decreased substantially in the animal model with maternal effects.

```{r}
plot(mcmc.list(
  m2_cinf_anim$VCV[,'id'], m3_cinf_anim$VCV[,'id']
  ), density = F)
```

Let's see how much variation is explained by these models.

```{r}
# FIXED AND TOTAL VARIANCES
pdx <- m3_cinf_mat$X # model matrix, fixed effx
n <- dim(pdx)[1] # number of data points in model, all traits
nt <- n/1 # number of data points per trait
p_i <- which(is.na(dat_cinf$pred)) # points where we can predicted vals and cred int
pdx <- pdx[c(p_i),] # restrict to only points where we want preds
p_n <- dim(pdx)[1]/1 # number of points for each trait we want to predict
num_fe <- m3_cinf_mat$Fixed$nfl # number of fixed effx, rand effx marginalized
p_all <- as.matrix(pdx) %*% t(m3_cinf_mat$Sol[,1:num_fe]) # predicteds via matrix mult 
m3_Vf_mat <- apply(p_all, MARGIN = 2, var)


m3_Vp_mat <- m3_Vf_mat + rowSums(m3_cinf_mat$VCV)


pdx <- m3_cinf_anim$X # model matrix, fixed effx
n <- dim(pdx)[1] # number of data points in model, all traits
nt <- n/1 # number of data points per trait
p_i <- which(is.na(dat_cinf$pred)) # points where we can predicted vals and cred int
pdx <- pdx[c(p_i),] # restrict to only points where we want preds
p_n <- dim(pdx)[1]/1 # number of points for each trait we want to predict
num_fe <- m3_cinf_anim$Fixed$nfl # number of fixed effx, rand effx marginalized
p_all <- as.matrix(pdx) %*% t(m3_cinf_anim$Sol[,1:num_fe]) # predicteds via matrix mult 
m3_Vf_anim <- apply(p_all, MARGIN = 2, var)


m3_Vp_anim <- m3_Vf_anim + rowSums(m3_cinf_anim$VCV)
```
```{r}
# vc for model 3 full sib
m3_vc <- data.frame(mod = "3a_full_sibs", 
                    v_est = c(
                      m3_Vf_mat/m3_Vp_mat, 
                      m3_cinf_mat$VCV[,'cop_block']/m3_Vp_mat, 
                      m3_cinf_mat$VCV[,'cop_block']/(m3_Vp_mat - m3_Vf_mat), 
                      2*m3_cinf_mat$VCV[,'worm_full_sib_fam_id']/m3_Vp_mat, 
                      2*m3_cinf_mat$VCV[,'worm_full_sib_fam_id']/(m3_Vp_mat - m3_Vf_mat), 
                      m3_cinf_mat$VCV[,'dam']/m3_Vp_mat, 
                      m3_cinf_mat$VCV[,'dam']/(m3_Vp_mat - m3_Vf_mat),
                      1-(m3_Vf_mat+m3_cinf_mat$VCV[,'cop_block'] + 2*m3_cinf_mat$VCV[,'worm_full_sib_fam_id'] +
                           m3_cinf_mat$VCV[,'dam'])/m3_Vp_mat,
                      1-(m3_cinf_mat$VCV[,'cop_block'] + 2*m3_cinf_mat$VCV[,'worm_full_sib_fam_id'] +
                           m3_cinf_mat$VCV[,'dam'])/(m3_Vp_mat - m3_Vf_mat)
                      ),
                    v_name = rep(c("Vf", "Ve", "Ve2", "Vg", "Vg2", "Vm", "Vm2", "Vr", "Vr2"), 
                                 times = rep(length(m2_Vf_anim), 9))
                    )


# vc for model 3 anim
m3_vc_anim <- data.frame(mod = "3b_animal", 
                    v_est = c(
                      m3_Vf_anim/m3_Vp_anim, 
                      m3_cinf_anim$VCV[,'cop_block']/m3_Vp_anim, 
                      m3_cinf_anim$VCV[,'cop_block']/(m3_Vp_anim - m3_Vf_anim), 
                      m3_cinf_anim$VCV[,'id']/m3_Vp_anim, 
                      m3_cinf_anim$VCV[,'id']/(m3_Vp_anim - m3_Vf_anim), 
                      m3_cinf_anim$VCV[,'dam']/m3_Vp_anim, 
                      m3_cinf_anim$VCV[,'dam']/(m3_Vp_anim - m3_Vf_anim),
                      1-(m3_Vf_anim+m3_cinf_anim$VCV[,'cop_block'] + m3_cinf_anim$VCV[,'id'] +
                           m3_cinf_anim$VCV[,'dam'])/m3_Vp_anim,
                      1-(m3_cinf_anim$VCV[,'cop_block'] + m3_cinf_anim$VCV[,'id'] +
                           m3_cinf_anim$VCV[,'dam'])/(m3_Vp_anim - m3_Vf_anim)
                      ),
                    v_name = rep(c("Vf", "Ve", "Ve2", "Vg", "Vg2", "Vm", "Vm2", "Vr", "Vr2"), 
                                 times = rep(length(m2_Vf_anim), 9))
                    )



m_vc <- bind_rows(
  m3_vc, m3_vc_anim
)


m_vc <- m_vc%>%
  group_by(mod, v_name)%>%
  summarise(v_fit = median(v_est),
            v_lwr = quantile(v_est, probs = 0.025),
            v_upr = quantile(v_est, probs = 0.975),
            )%>%
  ungroup()%>%
  mutate(trait = "cinf")

m_vc_cinf2 <- bind_rows(
  m_vc_cinf2,
  m_vc%>%filter(v_name %in% c("Ve2", "Vg2", "Vm2", "Vr2"))
  ) # var with fixed effects

m_vc_cinf <- bind_rows(
  m_vc_cinf,
  m_vc%>%filter(!v_name %in% c("Ve2", "Vg2", "Vm2", "Vr2"))
  ) # var excluding fixed effects
```

Here is the percent variation in copepod infection accounted for by copepod stage in the five models:

```{r}
filter(m_vc_cinf, v_name == "Vf")%>%
  mutate(across(v_fit:v_upr, function(x){round(x,3)}))
```

Here is the percent variation in copepod infection accounted for by copepod block:

```{r}
filter(m_vc_cinf, v_name == "Ve")%>%
  mutate(across(v_fit:v_upr, function(x){round(x,3)}))
```

Here are the heritability estimates. The animal model estimate comes down quite a bit after adding maternal effects.

```{r}
filter(m_vc_cinf, v_name == "Vg")%>%
  mutate(across(v_fit:v_upr, function(x){round(x,3)}))
```

Here is the percent variation in copepod infection accounted for by parental effects in the last two models

```{r}
filter(m_vc_cinf, v_name == "Vm")%>%
  mutate(across(v_fit:v_upr, function(x){round(x,3)}))
```

Here is the total variance explained.

```{r}
filter(m_vc_cinf, v_name == "Vr")%>%
  mutate(across(v_fit:v_upr, function(x){round(1-x,3)}))
```

The fraction of phenotypic variance explained by each term in the different models is plotted here

```{r}
ggplot(m_vc_cinf%>%
         mutate(v_name = fct_relevel(factor(v_name), c("Vf", "Ve", "Vg", "Vm", "Vr"))),
       aes(v_name, v_fit, color = mod)) +
  geom_pointrange(aes(ymin = v_lwr, ymax = v_upr),
                  position = position_dodge(width = 0.6)) +
  labs(y = "Fraction variance explained", title = "Copepod infection rate") +
  scale_x_discrete(labels = c("Starting\nconditions", "Shared\nenvironment", "Additive\ngenetic", "Parental", "Residual")) +
  theme(panel.grid.major.x = element_blank(),
        axis.title.x = element_blank(),
        legend.title = element_blank())
```

We can make the same plot, but only consider the fraction of phenotypic variation explained after accounting for starting conditions (i.e. we leave fixed effect variance out of the denominator).

```{r}
ggplot(m_vc_cinf2%>%
         mutate(v_name = fct_relevel(factor(v_name), c("Ve2", "Vg2", "Vm2", "Vr2"))),
       aes(v_name, v_fit, color = mod)) +
  geom_pointrange(aes(ymin = v_lwr, ymax = v_upr),
                  position = position_dodge(width = 0.6)) +
  labs(y = "Fraction variance explained", title = "Copepod infection rate") +
  scale_x_discrete(labels = c("Shared\nenvironment", "Additive\ngenetic",
                              "Parental", "Residual")) +
  theme(panel.grid.major.x = element_blank(),
        axis.title.x = element_blank(),
        legend.title = element_blank())
```

## Cercomere presence

```{r}
parents <- dat_cerc%>%
  group_by(eggs_clutch_id)%>%
  summarise(n_parent = sum(!is.na(cerc_bi)), cerc_p = mean(cerc_bi))%>%
  na.omit()

offspring <- dat_cerc%>%
  group_by(worm_full_sib_fam_id)%>%
  summarise(n_offspring = sum(!is.na(cerc_bi)), cerc_o = mean(cerc_bi, na.rm = T))%>%
  na.omit()

parent_offspring <- left_join(parents, offspring, by = c("eggs_clutch_id" = "worm_full_sib_fam_id"))
```

The parent offspring regression for cercomere presence was positive, which is expected since we observed a significant response to selection.

```{r}
ggplot(parent_offspring, aes(cerc_p, cerc_o)) +
  geom_point(alpha = .2) + geom_smooth()
```

Here is the slope of the line. It is quite similar to the realized heritability.

```{r}
p_o_cerc <- lm(cerc_o ~ cerc_p, data = parent_offspring%>%filter(!is.na(cerc_o)))
summary(p_o_cerc)
```

This heritability is corrected based on the proportion of parasites, overall, that had a cercomere. See de Villemereuil et a. 2013.

```{r}
p_cerc <- sum(dat_cerc$cerc_bi, na.rm = T)/sum(!is.na(dat_cerc$cerc_bi))
(p_cerc*(1-p_cerc))/(dnorm(p_cerc))^2 * coef(p_o_cerc)[2]
```

Here are the number of full-sib families and the number of offspring phenotyped per family.

```{r}
dat_cerc%>%
  group_by(worm_full_sib_fam_id)%>%
  summarize(n_pheno = sum(!is.na(cerc_bi)))%>%
  ungroup()%>%
  filter(!is.na(worm_full_sib_fam_id))%>%
  summarize(fs_fams = n(),
            min_phenotyped = min(n_pheno),
            lwrIQR_phenotyped = quantile(n_pheno, prob = 0.25),
            avg_phenotyped = mean(n_pheno),
            uprIQR_phenotyped = quantile(n_pheno, prob = 0.75),
            max_phenotyped = max(n_pheno)
            )
```

Here are the number of offspring phenotyped per dam/sire.

```{r}
dat_cerc%>%
  group_by(maternal_match, worm_maternal_id)%>%
  summarize(n_pheno = sum(!is.na(cerc_bi)))%>%
  group_by(maternal_match)%>%
  filter(!is.na(worm_maternal_id))%>%
  summarize(dams = n(),
            min_phenotyped = min(n_pheno),
            lwrIQR_phenotyped = quantile(n_pheno, prob = 0.25),
            avg_phenotyped = mean(n_pheno),
            uprIQR_phenotyped = quantile(n_pheno, prob = 0.75),
            max_phenotyped = max(n_pheno)
            )
```

Again, we will fit 5 models.

```{r}
p0_bi <- list(R = list(V = 1, fix = 1))

p1_bi <- list(R = list(V = 1, fix = 1),
              G = list(G1 = list(V = 1, nu = 1000, alpha.mu = 0, alpha.V = 1)))

p2_bi <- list(R = list(V = 1, fix = 1),
              G = list(G1 = list(V = 1, nu = 1000, alpha.mu = 0, alpha.V = 1),
                       G2 = list(V = 1, nu = 1000, alpha.mu = 0, alpha.V = 1)))
p3_bi <- list(R = list(V = 1, fix = 1),
              G = list(G1 = list(V = 1, nu = 1000, alpha.mu = 0, alpha.V = 1),
                       G2 = list(V = 1, nu = 1000, alpha.mu = 0, alpha.V = 1),
                       G3 = list(V = 1, nu = 1000, alpha.mu = 0, alpha.V = 1)))


thin_cerc <- 75
burnin_cerc <- 1000
nit_cerc <- thin_cerc * target_samples + burnin_cerc


# fixed effects (no ib)
m1_cerc_noib <- MCMCglmm(cerc_bi ~ cop_stage_checking, 
                 random = ~cop_block,
                 family = "threshold",
                 data = dat_cerc,
                 prior = p1_bi,
                 verbose=FALSE,
                 nitt = nit_cerc, 
                 thin = thin_cerc,
                 burnin = burnin_cerc
                )
# fixed effects
m1_cerc <- MCMCglmm(cerc_bi ~ ib + cop_stage_checking, 
                 random = ~cop_block,
                 family = "threshold",
                 data = dat_cerc,
                 prior = p1_bi,
                 verbose=FALSE,
                 nitt = nit_cerc, 
                 thin = thin_cerc,
                 burnin = burnin_cerc
                )
# full-sibs
m2_cerc_fam <- MCMCglmm(cerc_bi ~ ib + cop_stage_checking, 
                 random = ~ cop_block + worm_full_sib_fam_id,
                 family = "threshold",
                 data = dat_cerc,
                 prior = p2_bi,
                 verbose=F,
                 pr = T,
                 nitt = nit_cerc, 
                 thin = thin_cerc,
                 burnin = burnin_cerc
                )
# full sib x line
m2_cerc_famXline <- MCMCglmm(
  cerc_bi ~ ib + cop_stage_checking, 
  random = ~ cop_block + idh(line):worm_full_sib_fam_id,
  family = "threshold",
  data = dat_cerc,
  prior = 
    list(R = list(V = 1, fix = 1),
         G = list(G1 = list(V = 1, nu = 1000, alpha.mu = 0, alpha.V = 1),
                  G2 = list(V = diag(4), nu = 0.002))),
  verbose=FALSE,
  pr = F,
  nitt = nit_cerc, 
  thin = thin_cerc,
  burnin = burnin_cerc
  )

# anim model
m2_cerc_anim <- MCMCglmm(cerc_bi ~ ib + cop_stage_checking, 
                 random = ~cop_block + id,
                 ginverse = list(id = Ainv),
                 family = "threshold",
                 data = dat_cerc,
                 prior = p2_bi,
                 verbose=F,
                 nitt = nit_cerc, 
                 thin = thin_cerc,
                 burnin = burnin_cerc
                )

# full-sibs + mat effx
m3_cerc_mat <- MCMCglmm(cerc_bi ~ ib + cop_stage_checking, 
                 random = ~cop_block + worm_full_sib_fam_id + dam,
                 family = "threshold",
                 data = dat_cerc%>%mutate(dam = factor(dam)),
                 prior = p3_bi,
                 verbose=F,
                 pr=T,
                 nitt = nit_cerc, 
                 thin = thin_cerc,
                 burnin = burnin_cerc
                )
# anim model
m3_cerc_anim <- MCMCglmm(cerc_bi ~ ib + cop_stage_checking, 
                 random = ~cop_block + id  + dam,
                 ginverse = list(id = Ainv),
                 family = "threshold",
                 data = dat_cerc%>%mutate(dam = factor(dam)),
                 prior = p3_bi,
                 verbose=F,
                 nitt = nit_cerc, 
                 thin = thin_cerc,
                 burnin = burnin_cerc
                )

```

Adding full-sibs (red) is an improvement, and the animal model (green) looks like a further improvement, though there must be more variation in the parameters.

```{r}
plot(mcmc.list(
  m1_cerc$Deviance, m2_cerc_fam$Deviance, m2_cerc_anim$Deviance
  ), density = F)
```

The full-sib model suggests that some variance within clutches is explained by parental effects.

```{r}
plot(mcmc.list(
  m2_cerc_fam$Deviance, m3_cerc_mat$Deviance
  ), density = F)
```

By contrast, the addition of maternal effects does not improve the animal model.

```{r}
plot(mcmc.list(
  m2_cerc_anim$Deviance, m3_cerc_anim$Deviance
  ), density = F)
```

For, these models, we break down total variance into several components: fixed effects (starting conditions), shared environment, and genetic effects.

```{r}
get_var_comps <- function(trait, daty, m1a, m1b, m2a, m2b, m3a, m3b){
  # FIXED EFFECT VARIANCE
  # first model no ib
  pdx <- m1a$X # model matrix, fixed effx
  n <- dim(pdx)[1] # number of data points in model, all traits
  nt <- n/1 # number of data points per trait
  p_i <- which(is.na(daty$pred)) # points where we can predicted vals and cred int
  pdx <- pdx[c(p_i),] # restrict to only points where we want preds
  p_n <- dim(pdx)[1]/1 # number of points for each trait we want to predict
  num_fe <- m1a$Fixed$nfl # number of fixed effx, rand effx marginalized
  p_all <- as.matrix(pdx) %*% t(m1a$Sol[,1:num_fe]) # predicteds via matrix mult 
  m1_Vf_noib <- apply(p_all, MARGIN = 2, var)

  # first model
  pdx <- m1b$X # model matrix, fixed effx
  n <- dim(pdx)[1] # number of data points in model, all traits
  nt <- n/1 # number of data points per trait
  p_i <- which(is.na(daty$pred)) # points where we can predicted vals and cred int
  pdx <- pdx[c(p_i),] # restrict to only points where we want preds
  p_n <- dim(pdx)[1]/1 # number of points for each trait we want to predict
  num_fe <- m1b$Fixed$nfl # number of fixed effx, rand effx marginalized
  p_all <- as.matrix(pdx) %*% t(m1b$Sol[,1:num_fe]) # predicteds via matrix mult 
  m1_Vf <- apply(p_all, MARGIN = 2, var)
  
  pdx <- m2a$X # model matrix, fixed effx
  n <- dim(pdx)[1] # number of data points in model, all traits
  nt <- n/1 # number of data points per trait
  p_i <- which(is.na(daty$pred)) # points where we can predicted vals and cred int
  pdx <- pdx[c(p_i),] # restrict to only points where we want preds
  p_n <- dim(pdx)[1]/1 # number of points for each trait we want to predict
  num_fe <- m2a$Fixed$nfl # number of fixed effx, rand effx marginalized
  p_all <- as.matrix(pdx) %*% t(m2a$Sol[,1:num_fe]) # predicteds via matrix mult 
  m2_Vf_fam <- apply(p_all, MARGIN = 2, var)
  
  pdx <- m2b$X # model matrix, fixed effx
  n <- dim(pdx)[1] # number of data points in model, all traits
  nt <- n/1 # number of data points per trait
  p_i <- which(is.na(daty$pred)) # points where we can predicted vals and cred int
  pdx <- pdx[c(p_i),] # restrict to only points where we want preds
  p_n <- dim(pdx)[1]/1 # number of points for each trait we want to predict
  num_fe <- m2b$Fixed$nfl # number of fixed effx, rand effx marginalized
  p_all <- as.matrix(pdx) %*% t(m2b$Sol[,1:num_fe]) # predicteds via matrix mult 
  m2_Vf_anim <- apply(p_all, MARGIN = 2, var)
  
  pdx <- m3a$X # model matrix, fixed effx
  n <- dim(pdx)[1] # number of data points in model, all traits
  nt <- n/1 # number of data points per trait
  p_i <- which(is.na(daty$pred)) # points where we can predicted vals and cred int
  pdx <- pdx[c(p_i),] # restrict to only points where we want preds
  p_n <- dim(pdx)[1]/1 # number of points for each trait we want to predict
  num_fe <- m3a$Fixed$nfl # number of fixed effx, rand effx marginalized
  p_all <- as.matrix(pdx) %*% t(m3a$Sol[,1:num_fe]) # predicteds via matrix mult 
  m3_Vf_mat <- apply(p_all, MARGIN = 2, var)
  
  pdx <- m3b$X # model matrix, fixed effx
  n <- dim(pdx)[1] # number of data points in model, all traits
  nt <- n/1 # number of data points per trait
  p_i <- which(is.na(daty$pred)) # points where we can predicted vals and cred int
  pdx <- pdx[c(p_i),] # restrict to only points where we want preds
  p_n <- dim(pdx)[1]/1 # number of points for each trait we want to predict
  num_fe <- m3b$Fixed$nfl # number of fixed effx, rand effx marginalized
  p_all <- as.matrix(pdx) %*% t(m3b$Sol[,1:num_fe]) # predicteds via matrix mult 
  m3_Vf_anim <- apply(p_all, MARGIN = 2, var)
  
  # TOTAL PHENOTYPIC VARIANCE
  m1_Vp <- m1_Vf + rowSums(m1b$VCV)
  m1_Vp_noib <- m1_Vf_noib + rowSums(m1a$VCV)
  m2_Vp_fam <- m2_Vf_fam + rowSums(m2a$VCV)
  m2_Vp_anim <- m2_Vf_anim + rowSums(m2b$VCV)
  m3_Vp_mat <- m3_Vf_mat + rowSums(m3a$VCV)
  m3_Vp_anim <- m3_Vf_anim + rowSums(m3b$VCV)
  
  # vc for model 1 without inbreeding
  m1_vc_noib <- data.frame(mod = "1a_base_noIB", 
                      v_est = c(
                        m1_Vf_noib/m1_Vp_noib, 
                        m1a$VCV[,'cop_block']/m1_Vp_noib,
                        m1a$VCV[,'cop_block']/(m1_Vp_noib - m1_Vf_noib),
                        1-(m1_Vf_noib+m1a$VCV[,'cop_block'])/m1_Vp_noib,
                        1-(m1a$VCV[,'cop_block'])/(m1_Vp_noib - m1_Vf_noib)
                        ),
                      v_name = rep(c("Vf", "Ve", "Ve2", "Vr", "Vr2"), 
                                   times = rep(length(m1_Vf_noib), 5))
                      )
  # vc for model 1
  m1_vc <- data.frame(mod = "1b_base", 
                      v_est = c(
                        m1_Vf/m1_Vp, 
                        m1b$VCV[,'cop_block']/m1_Vp,
                        m1b$VCV[,'cop_block']/(m1_Vp - m1_Vf),
                        1-(m1_Vf+m1b$VCV[,'cop_block'])/m1_Vp,
                        1-(m1b$VCV[,'cop_block'])/(m1_Vp - m1_Vf)
                        ),
                      v_name = rep(c("Vf", "Ve", "Ve2", "Vr", "Vr2"), 
                                   times = rep(length(m1_Vf), 5))
                      )
  
  # vc for model 2 full sib
  m2_vc <- data.frame(mod = "2a_full_sibs", 
                      v_est = c(
                        m2_Vf_fam/m2_Vp_fam, 
                        m2a$VCV[,'cop_block']/m2_Vp_fam, 
                        m2a$VCV[,'cop_block']/(m2_Vp_fam - m2_Vf_fam), 
                        2*m2a$VCV[,'worm_full_sib_fam_id']/m2_Vp_fam, 
                        2*m2a$VCV[,'worm_full_sib_fam_id']/(m2_Vp_fam - m2_Vf_fam),
                        1-(m2_Vf_fam+m2a$VCV[,'cop_block'] +
                             2*m2a$VCV[,'worm_full_sib_fam_id'])/m2_Vp_fam,
                        1-(m2a$VCV[,'cop_block'] +
                             2*m2a$VCV[,'worm_full_sib_fam_id'])/(m2_Vp_fam - m2_Vf_fam)
                        ),
                      v_name = rep(c("Vf", "Ve", "Ve2", "Vg", "Vg2", "Vr", "Vr2"), 
                              times = rep(length(m2_Vf_fam), 7))
                      )
  
  # vc for model 2 anim
  m2_vc_anim <- data.frame(mod = "2b_animal", 
                      v_est = c(
                        m2_Vf_anim/m2_Vp_anim, 
                        m2b$VCV[,'cop_block']/m2_Vp_anim, 
                        m2b$VCV[,'cop_block']/(m2_Vp_anim - m2_Vf_anim), 
                        m2b$VCV[,'id']/m2_Vp_anim, 
                        m2b$VCV[,'id']/(m2_Vp_anim - m2_Vf_anim),
                        1-(m2_Vf_anim+m2b$VCV[,'cop_block'] +
                             m2b$VCV[,'id'])/m2_Vp_anim,
                        1-(m2b$VCV[,'cop_block'] +
                             m2b$VCV[,'id'])/(m2_Vp_anim - m2_Vf_anim)
                        ),
                      v_name = rep(c("Vf", "Ve", "Ve2", "Vg", "Vg2", "Vr", "Vr2"), 
                              times = rep(length(m2_Vf_anim), 7))
                      )
  
  # vc for model 3 full sib
  m3_vc <- data.frame(mod = "3a_full_sibs", 
                      v_est = c(
                        m3_Vf_mat/m3_Vp_mat, 
                        m3a$VCV[,'cop_block']/m3_Vp_mat, 
                        m3a$VCV[,'cop_block']/(m3_Vp_mat - m3_Vf_mat), 
                        2*m3a$VCV[,'worm_full_sib_fam_id']/m3_Vp_mat, 
                        2*m3a$VCV[,'worm_full_sib_fam_id']/(m3_Vp_mat - m3_Vf_mat), 
                        m3a$VCV[,'dam']/m3_Vp_mat, 
                        m3a$VCV[,'dam']/(m3_Vp_mat - m3_Vf_mat),
                        1-(m3_Vf_mat+m3a$VCV[,'cop_block'] + 2*m3a$VCV[,'worm_full_sib_fam_id'] +
                             m3a$VCV[,'dam'])/m3_Vp_mat,
                        1-(m3a$VCV[,'cop_block'] + 2*m3a$VCV[,'worm_full_sib_fam_id'] +
                             m3a$VCV[,'dam'])/(m3_Vp_mat - m3_Vf_mat)
                        ),
                      v_name = rep(c("Vf", "Ve", "Ve2", "Vg", "Vg2", "Vm", "Vm2", "Vr", "Vr2"), 
                                   times = rep(length(m2_Vf_anim), 9))
                      )
  
  
  # vc for model 3 anim
  m3_vc_anim <- data.frame(mod = "3b_animal", 
                      v_est = c(
                        m3_Vf_anim/m3_Vp_anim, 
                        m3b$VCV[,'cop_block']/m3_Vp_anim, 
                        m3b$VCV[,'cop_block']/(m3_Vp_anim - m3_Vf_anim), 
                        m3b$VCV[,'id']/m3_Vp_anim, 
                        m3b$VCV[,'id']/(m3_Vp_anim - m3_Vf_anim), 
                        m3b$VCV[,'dam']/m3_Vp_anim, 
                        m3b$VCV[,'dam']/(m3_Vp_anim - m3_Vf_anim),
                        1-(m3_Vf_anim+m3b$VCV[,'cop_block'] + m3b$VCV[,'id'] +
                             m3b$VCV[,'dam'])/m3_Vp_anim,
                        1-(m3b$VCV[,'cop_block'] + m3b$VCV[,'id'] +
                             m3b$VCV[,'dam'])/(m3_Vp_anim - m3_Vf_anim)
                        ),
                      v_name = rep(c("Vf", "Ve", "Ve2", "Vg", "Vg2", "Vm", "Vm2", "Vr", "Vr2"), 
                                   times = rep(length(m2_Vf_anim), 9))
                      )
  
  
  m_vc <- bind_rows(
    m1_vc_noib, m1_vc,
    m2_vc, m2_vc_anim,
    m3_vc, m3_vc_anim
  )
  
  
  m_vc_out <- m_vc%>%
    group_by(mod, v_name)%>%
    summarise(v_fit = median(v_est),
              v_lwr = quantile(v_est, probs = 0.025),
              v_upr = quantile(v_est, probs = 0.975),
              )%>%
    ungroup()%>%
    mutate(trait = trait)

  return(m_vc_out)
    
}
```

```{r}
get_var_comps <- function(trait, daty, evar, m1a, m1b, m2a, m2b, m3a, m3b){
  # FIXED EFFECT VARIANCE
  # first model no ib
  pdx <- m1a$X # model matrix, fixed effx
  n <- dim(pdx)[1] # number of data points in model, all traits
  nt <- n/1 # number of data points per trait
  p_i <- which(is.na(daty$pred)) # points where we can predicted vals and cred int
  pdx <- pdx[c(p_i),] # restrict to only points where we want preds
  p_n <- dim(pdx)[1]/1 # number of points for each trait we want to predict
  num_fe <- m1a$Fixed$nfl # number of fixed effx, rand effx marginalized
  p_all <- as.matrix(pdx) %*% t(m1a$Sol[,1:num_fe]) # predicteds via matrix mult 
  m1_Vf_noib <- apply(p_all, MARGIN = 2, var)

  # first model
  pdx <- m1b$X # model matrix, fixed effx
  n <- dim(pdx)[1] # number of data points in model, all traits
  nt <- n/1 # number of data points per trait
  p_i <- which(is.na(daty$pred)) # points where we can predicted vals and cred int
  pdx <- pdx[c(p_i),] # restrict to only points where we want preds
  p_n <- dim(pdx)[1]/1 # number of points for each trait we want to predict
  num_fe <- m1b$Fixed$nfl # number of fixed effx, rand effx marginalized
  p_all <- as.matrix(pdx) %*% t(m1b$Sol[,1:num_fe]) # predicteds via matrix mult 
  m1_Vf <- apply(p_all, MARGIN = 2, var)
  
  pdx <- m2a$X # model matrix, fixed effx
  n <- dim(pdx)[1] # number of data points in model, all traits
  nt <- n/1 # number of data points per trait
  p_i <- which(is.na(daty$pred)) # points where we can predicted vals and cred int
  pdx <- pdx[c(p_i),] # restrict to only points where we want preds
  p_n <- dim(pdx)[1]/1 # number of points for each trait we want to predict
  num_fe <- m2a$Fixed$nfl # number of fixed effx, rand effx marginalized
  p_all <- as.matrix(pdx) %*% t(m2a$Sol[,1:num_fe]) # predicteds via matrix mult 
  m2_Vf_fam <- apply(p_all, MARGIN = 2, var)
  
  pdx <- m2b$X # model matrix, fixed effx
  n <- dim(pdx)[1] # number of data points in model, all traits
  nt <- n/1 # number of data points per trait
  p_i <- which(is.na(daty$pred)) # points where we can predicted vals and cred int
  pdx <- pdx[c(p_i),] # restrict to only points where we want preds
  p_n <- dim(pdx)[1]/1 # number of points for each trait we want to predict
  num_fe <- m2b$Fixed$nfl # number of fixed effx, rand effx marginalized
  p_all <- as.matrix(pdx) %*% t(m2b$Sol[,1:num_fe]) # predicteds via matrix mult 
  m2_Vf_anim <- apply(p_all, MARGIN = 2, var)
  
  pdx <- m3a$X # model matrix, fixed effx
  n <- dim(pdx)[1] # number of data points in model, all traits
  nt <- n/1 # number of data points per trait
  p_i <- which(is.na(daty$pred)) # points where we can predicted vals and cred int
  pdx <- pdx[c(p_i),] # restrict to only points where we want preds
  p_n <- dim(pdx)[1]/1 # number of points for each trait we want to predict
  num_fe <- m3a$Fixed$nfl # number of fixed effx, rand effx marginalized
  p_all <- as.matrix(pdx) %*% t(m3a$Sol[,1:num_fe]) # predicteds via matrix mult 
  m3_Vf_mat <- apply(p_all, MARGIN = 2, var)
  
  pdx <- m3b$X # model matrix, fixed effx
  n <- dim(pdx)[1] # number of data points in model, all traits
  nt <- n/1 # number of data points per trait
  p_i <- which(is.na(daty$pred)) # points where we can predicted vals and cred int
  pdx <- pdx[c(p_i),] # restrict to only points where we want preds
  p_n <- dim(pdx)[1]/1 # number of points for each trait we want to predict
  num_fe <- m3b$Fixed$nfl # number of fixed effx, rand effx marginalized
  p_all <- as.matrix(pdx) %*% t(m3b$Sol[,1:num_fe]) # predicteds via matrix mult 
  m3_Vf_anim <- apply(p_all, MARGIN = 2, var)
  
  # TOTAL PHENOTYPIC VARIANCE
  m1_Vp <- m1_Vf + rowSums(m1b$VCV)
  m1_Vp_noib <- m1_Vf_noib + rowSums(m1a$VCV)
  m2_Vp_fam <- m2_Vf_fam + rowSums(m2a$VCV)
  m2_Vp_anim <- m2_Vf_anim + rowSums(m2b$VCV)
  m3_Vp_mat <- m3_Vf_mat + rowSums(m3a$VCV)
  m3_Vp_anim <- m3_Vf_anim + rowSums(m3b$VCV)
  
  # vc for model 1 without inbreeding
  m1_vc_noib <- data.frame(mod = "1a_base_noIB", 
                      v_est = c(
                        m1_Vf_noib/m1_Vp_noib, 
                        m1a$VCV[,evar]/m1_Vp_noib,
                        m1a$VCV[,evar]/(m1_Vp_noib - m1_Vf_noib),
                        1-(m1_Vf_noib+m1a$VCV[,evar])/m1_Vp_noib,
                        1-(m1a$VCV[,evar])/(m1_Vp_noib - m1_Vf_noib)
                        ),
                      v_name = rep(c("Vf", "Ve", "Ve2", "Vr", "Vr2"), 
                                   times = rep(length(m1_Vf_noib), 5))
                      )
  # vc for model 1
  m1_vc <- data.frame(mod = "1b_base", 
                      v_est = c(
                        m1_Vf/m1_Vp, 
                        m1b$VCV[,evar]/m1_Vp,
                        m1b$VCV[,evar]/(m1_Vp - m1_Vf),
                        1-(m1_Vf+m1b$VCV[,evar])/m1_Vp,
                        1-(m1b$VCV[,evar])/(m1_Vp - m1_Vf)
                        ),
                      v_name = rep(c("Vf", "Ve", "Ve2", "Vr", "Vr2"), 
                                   times = rep(length(m1_Vf), 5))
                      )
  
  # vc for model 2 full sib
  m2_vc <- data.frame(mod = "2a_full_sibs", 
                      v_est = c(
                        m2_Vf_fam/m2_Vp_fam, 
                        m2a$VCV[,evar]/m2_Vp_fam, 
                        m2a$VCV[,evar]/(m2_Vp_fam - m2_Vf_fam), 
                        2*m2a$VCV[,'worm_full_sib_fam_id']/m2_Vp_fam, 
                        2*m2a$VCV[,'worm_full_sib_fam_id']/(m2_Vp_fam - m2_Vf_fam),
                        1-(m2_Vf_fam+m2a$VCV[,evar] +
                             2*m2a$VCV[,'worm_full_sib_fam_id'])/m2_Vp_fam,
                        1-(m2a$VCV[,evar] +
                             2*m2a$VCV[,'worm_full_sib_fam_id'])/(m2_Vp_fam - m2_Vf_fam)
                        ),
                      v_name = rep(c("Vf", "Ve", "Ve2", "Vg", "Vg2", "Vr", "Vr2"), 
                              times = rep(length(m2_Vf_fam), 7))
                      )
  
  # vc for model 2 anim
  m2_vc_anim <- data.frame(mod = "2b_animal", 
                      v_est = c(
                        m2_Vf_anim/m2_Vp_anim, 
                        m2b$VCV[,evar]/m2_Vp_anim, 
                        m2b$VCV[,evar]/(m2_Vp_anim - m2_Vf_anim), 
                        m2b$VCV[,'id']/m2_Vp_anim, 
                        m2b$VCV[,'id']/(m2_Vp_anim - m2_Vf_anim),
                        1-(m2_Vf_anim+m2b$VCV[,evar] +
                             m2b$VCV[,'id'])/m2_Vp_anim,
                        1-(m2b$VCV[,evar] +
                             m2b$VCV[,'id'])/(m2_Vp_anim - m2_Vf_anim)
                        ),
                      v_name = rep(c("Vf", "Ve", "Ve2", "Vg", "Vg2", "Vr", "Vr2"), 
                              times = rep(length(m2_Vf_anim), 7))
                      )
  
  # vc for model 3 full sib
  m3_vc <- data.frame(mod = "3a_full_sibs", 
                      v_est = c(
                        m3_Vf_mat/m3_Vp_mat, 
                        m3a$VCV[,evar]/m3_Vp_mat, 
                        m3a$VCV[,evar]/(m3_Vp_mat - m3_Vf_mat), 
                        2*m3a$VCV[,'worm_full_sib_fam_id']/m3_Vp_mat, 
                        2*m3a$VCV[,'worm_full_sib_fam_id']/(m3_Vp_mat - m3_Vf_mat), 
                        m3a$VCV[,'dam']/m3_Vp_mat, 
                        m3a$VCV[,'dam']/(m3_Vp_mat - m3_Vf_mat),
                        1-(m3_Vf_mat+m3a$VCV[,evar] + 2*m3a$VCV[,'worm_full_sib_fam_id'] +
                             m3a$VCV[,'dam'])/m3_Vp_mat,
                        1-(m3a$VCV[,evar] + 2*m3a$VCV[,'worm_full_sib_fam_id'] +
                             m3a$VCV[,'dam'])/(m3_Vp_mat - m3_Vf_mat)
                        ),
                      v_name = rep(c("Vf", "Ve", "Ve2", "Vg", "Vg2", "Vm", "Vm2", "Vr", "Vr2"), 
                                   times = rep(length(m2_Vf_anim), 9))
                      )
  
  
  # vc for model 3 anim
  m3_vc_anim <- data.frame(mod = "3b_animal", 
                      v_est = c(
                        m3_Vf_anim/m3_Vp_anim, 
                        m3b$VCV[,evar]/m3_Vp_anim, 
                        m3b$VCV[,evar]/(m3_Vp_anim - m3_Vf_anim), 
                        m3b$VCV[,'id']/m3_Vp_anim, 
                        m3b$VCV[,'id']/(m3_Vp_anim - m3_Vf_anim), 
                        m3b$VCV[,'dam']/m3_Vp_anim, 
                        m3b$VCV[,'dam']/(m3_Vp_anim - m3_Vf_anim),
                        1-(m3_Vf_anim+m3b$VCV[,evar] + m3b$VCV[,'id'] +
                             m3b$VCV[,'dam'])/m3_Vp_anim,
                        1-(m3b$VCV[,evar] + m3b$VCV[,'id'] +
                             m3b$VCV[,'dam'])/(m3_Vp_anim - m3_Vf_anim)
                        ),
                      v_name = rep(c("Vf", "Ve", "Ve2", "Vg", "Vg2", "Vm", "Vm2", "Vr", "Vr2"), 
                                   times = rep(length(m2_Vf_anim), 9))
                      )
  
  
  m_vc <- bind_rows(
    m1_vc_noib, m1_vc,
    m2_vc, m2_vc_anim,
    m3_vc, m3_vc_anim
  )
  
  
  m_vc_out <- m_vc%>%
    group_by(mod, v_name)%>%
    summarise(v_fit = median(v_est),
              v_lwr = quantile(v_est, probs = 0.025),
              v_upr = quantile(v_est, probs = 0.975),
              )%>%
    ungroup()%>%
    mutate(trait = trait)

  return(m_vc_out)
    
}
```

```{r}
m_vc_cerc <- get_var_comps("cerc", dat_cerc, 'cop_block',
                            m1_cerc_noib, m1_cerc,
                            m2_cerc_fam, m2_cerc_anim,
                            m3_cerc_mat, m3_cerc_anim)

m_vc_cerc2 <- m_vc_cerc%>%filter(v_name %in% c("Ve2", "Vg2", "Vm2", "Vr2")) # var with fixed effects
m_vc_cerc <- m_vc_cerc%>%filter(!v_name %in% c("Ve2", "Vg2", "Vm2", "Vr2")) # var excluding fixed effects
```



```{r}
# # FIXED EFFECT VARIANCE
# # first model
# pdx <- m1_cerc$X # model matrix, fixed effx
# n <- dim(pdx)[1] # number of data points in model, all traits
# nt <- n/1 # number of data points per trait
# p_i <- which(is.na(dat_cerc$pred)) # points where we can predicted vals and cred int
# pdx <- pdx[c(p_i),] # restrict to only points where we want preds
# p_n <- dim(pdx)[1]/1 # number of points for each trait we want to predict
# num_fe <- m1_cerc$Fixed$nfl # number of fixed effx, rand effx marginalized
# p_all <- as.matrix(pdx) %*% t(m1_cerc$Sol[,1:num_fe]) # predicteds via matrix mult 
# m1_Vf <- apply(p_all, MARGIN = 2, var)
# 
# pdx <- m2_cerc_fam$X # model matrix, fixed effx
# n <- dim(pdx)[1] # number of data points in model, all traits
# nt <- n/1 # number of data points per trait
# p_i <- which(is.na(dat_cerc$pred)) # points where we can predicted vals and cred int
# pdx <- pdx[c(p_i),] # restrict to only points where we want preds
# p_n <- dim(pdx)[1]/1 # number of points for each trait we want to predict
# num_fe <- m2_cerc_fam$Fixed$nfl # number of fixed effx, rand effx marginalized
# p_all <- as.matrix(pdx) %*% t(m2_cerc_fam$Sol[,1:num_fe]) # predicteds via matrix mult 
# m2_Vf_fam <- apply(p_all, MARGIN = 2, var)
# 
# pdx <- m2_cerc_anim$X # model matrix, fixed effx
# n <- dim(pdx)[1] # number of data points in model, all traits
# nt <- n/1 # number of data points per trait
# p_i <- which(is.na(dat_cerc$pred)) # points where we can predicted vals and cred int
# pdx <- pdx[c(p_i),] # restrict to only points where we want preds
# p_n <- dim(pdx)[1]/1 # number of points for each trait we want to predict
# num_fe <- m2_cerc_anim$Fixed$nfl # number of fixed effx, rand effx marginalized
# p_all <- as.matrix(pdx) %*% t(m2_cerc_anim$Sol[,1:num_fe]) # predicteds via matrix mult 
# m2_Vf_anim <- apply(p_all, MARGIN = 2, var)
# 
# pdx <- m3_cerc_mat$X # model matrix, fixed effx
# n <- dim(pdx)[1] # number of data points in model, all traits
# nt <- n/1 # number of data points per trait
# p_i <- which(is.na(dat_cerc$pred)) # points where we can predicted vals and cred int
# pdx <- pdx[c(p_i),] # restrict to only points where we want preds
# p_n <- dim(pdx)[1]/1 # number of points for each trait we want to predict
# num_fe <- m3_cerc_mat$Fixed$nfl # number of fixed effx, rand effx marginalized
# p_all <- as.matrix(pdx) %*% t(m3_cerc_mat$Sol[,1:num_fe]) # predicteds via matrix mult 
# m3_Vf_mat <- apply(p_all, MARGIN = 2, var)
# 
# pdx <- m3_cerc_anim$X # model matrix, fixed effx
# n <- dim(pdx)[1] # number of data points in model, all traits
# nt <- n/1 # number of data points per trait
# p_i <- which(is.na(dat_cerc$pred)) # points where we can predicted vals and cred int
# pdx <- pdx[c(p_i),] # restrict to only points where we want preds
# p_n <- dim(pdx)[1]/1 # number of points for each trait we want to predict
# num_fe <- m3_cerc_anim$Fixed$nfl # number of fixed effx, rand effx marginalized
# p_all <- as.matrix(pdx) %*% t(m3_cerc_anim$Sol[,1:num_fe]) # predicteds via matrix mult 
# m3_Vf_anim <- apply(p_all, MARGIN = 2, var)
# 
# # TOTAL PHENOTYPIC VARIANCE
# m1_Vp <- m1_Vf + rowSums(m1_cerc$VCV)
# m2_Vp_fam <- m2_Vf_fam + rowSums(m2_cerc_fam$VCV)
# m2_Vp_anim <- m2_Vf_anim + rowSums(m2_cerc_anim$VCV)
# m3_Vp_mat <- m3_Vf_mat + rowSums(m3_cerc_mat$VCV)
# m3_Vp_anim <- m3_Vf_anim + rowSums(m3_cerc_anim$VCV)
# 
# # vc for model 1
# m1_vc <- data.frame(mod = "1_base", 
#                     v_est = c(
#                       m1_Vf/m1_Vp, 
#                       m1_cerc$VCV[,'cop_block']/m1_Vp,
#                       m1_cerc$VCV[,'cop_block']/(m1_Vp - m1_Vf),
#                       1-(m1_Vf+m1_cerc$VCV[,'cop_block'])/m1_Vp,
#                       1-(m1_cerc$VCV[,'cop_block'])/(m1_Vp - m1_Vf)
#                       ),
#                     v_name = rep(c("Vf", "Ve", "Ve2", "Vr", "Vr2"), 
#                                  times = rep(length(m1_Vf), 5))
#                     )
# 
# # vc for model 2 full sib
# m2_vc <- data.frame(mod = "2a_full_sibs", 
#                     v_est = c(
#                       m2_Vf_fam/m2_Vp_fam, 
#                       m2_cerc_fam$VCV[,'cop_block']/m2_Vp_fam, 
#                       m2_cerc_fam$VCV[,'cop_block']/(m2_Vp_fam - m2_Vf_fam), 
#                       2*m2_cerc_fam$VCV[,'worm_full_sib_fam_id']/m2_Vp_fam, 
#                       2*m2_cerc_fam$VCV[,'worm_full_sib_fam_id']/(m2_Vp_fam - m2_Vf_fam),
#                       1-(m2_Vf_fam+m2_cerc_fam$VCV[,'cop_block'] +
#                            2*m2_cerc_fam$VCV[,'worm_full_sib_fam_id'])/m2_Vp_fam,
#                       1-(m2_cerc_fam$VCV[,'cop_block'] +
#                            2*m2_cerc_fam$VCV[,'worm_full_sib_fam_id'])/(m2_Vp_fam - m2_Vf_fam)
#                       ),
#                     v_name = rep(c("Vf", "Ve", "Ve2", "Vg", "Vg2", "Vr", "Vr2"), 
#                             times = rep(length(m2_Vf_fam), 7))
#                     )
# 
# # vc for model 2 anim
# m2_vc_anim <- data.frame(mod = "2b_animal", 
#                     v_est = c(
#                       m2_Vf_anim/m2_Vp_anim, 
#                       m2_cerc_anim$VCV[,'cop_block']/m2_Vp_anim, 
#                       m2_cerc_anim$VCV[,'cop_block']/(m2_Vp_anim - m2_Vf_anim), 
#                       m2_cerc_anim$VCV[,'id']/m2_Vp_anim, 
#                       m2_cerc_anim$VCV[,'id']/(m2_Vp_anim - m2_Vf_anim),
#                       1-(m2_Vf_anim+m2_cerc_anim$VCV[,'cop_block'] +
#                            m2_cerc_anim$VCV[,'id'])/m2_Vp_anim,
#                       1-(m2_cerc_anim$VCV[,'cop_block'] +
#                            m2_cerc_anim$VCV[,'id'])/(m2_Vp_anim - m2_Vf_anim)
#                       ),
#                     v_name = rep(c("Vf", "Ve", "Ve2", "Vg", "Vg2", "Vr", "Vr2"), 
#                             times = rep(length(m2_Vf_anim), 7))
#                     )
# 
# # vc for model 3 full sib
# m3_vc <- data.frame(mod = "3a_full_sibs", 
#                     v_est = c(
#                       m3_Vf_mat/m3_Vp_mat, 
#                       m3_cerc_mat$VCV[,'cop_block']/m3_Vp_mat, 
#                       m3_cerc_mat$VCV[,'cop_block']/(m3_Vp_mat - m3_Vf_mat), 
#                       2*m3_cerc_mat$VCV[,'worm_full_sib_fam_id']/m3_Vp_mat, 
#                       2*m3_cerc_mat$VCV[,'worm_full_sib_fam_id']/(m3_Vp_mat - m3_Vf_mat), 
#                       m3_cerc_mat$VCV[,'dam']/m3_Vp_mat, 
#                       m3_cerc_mat$VCV[,'dam']/(m3_Vp_mat - m3_Vf_mat),
#                       1-(m3_Vf_mat+m3_cerc_mat$VCV[,'cop_block'] + 2*m3_cerc_mat$VCV[,'worm_full_sib_fam_id'] +
#                            m3_cerc_mat$VCV[,'dam'])/m3_Vp_mat,
#                       1-(m3_cerc_mat$VCV[,'cop_block'] + 2*m3_cerc_mat$VCV[,'worm_full_sib_fam_id'] +
#                            m3_cerc_mat$VCV[,'dam'])/(m3_Vp_mat - m3_Vf_mat)
#                       ),
#                     v_name = rep(c("Vf", "Ve", "Ve2", "Vg", "Vg2", "Vm", "Vm2", "Vr", "Vr2"), 
#                                  times = rep(length(m2_Vf_anim), 9))
#                     )
# 
# 
# # vc for model 3 anim
# m3_vc_anim <- data.frame(mod = "3b_animal", 
#                     v_est = c(
#                       m3_Vf_anim/m3_Vp_anim, 
#                       m3_cerc_anim$VCV[,'cop_block']/m3_Vp_anim, 
#                       m3_cerc_anim$VCV[,'cop_block']/(m3_Vp_anim - m3_Vf_anim), 
#                       m3_cerc_anim$VCV[,'id']/m3_Vp_anim, 
#                       m3_cerc_anim$VCV[,'id']/(m3_Vp_anim - m3_Vf_anim), 
#                       m3_cerc_anim$VCV[,'dam']/m3_Vp_anim, 
#                       m3_cerc_anim$VCV[,'dam']/(m3_Vp_anim - m3_Vf_anim),
#                       1-(m3_Vf_anim+m3_cerc_anim$VCV[,'cop_block'] + m3_cerc_anim$VCV[,'id'] +
#                            m3_cerc_anim$VCV[,'dam'])/m3_Vp_anim,
#                       1-(m3_cerc_anim$VCV[,'cop_block'] + m3_cerc_anim$VCV[,'id'] +
#                            m3_cerc_anim$VCV[,'dam'])/(m3_Vp_anim - m3_Vf_anim)
#                       ),
#                     v_name = rep(c("Vf", "Ve", "Ve2", "Vg", "Vg2", "Vm", "Vm2", "Vr", "Vr2"), 
#                                  times = rep(length(m2_Vf_anim), 9))
#                     )
# 
# 
# m_vc <- bind_rows(
#   m1_vc,
#   m2_vc, m2_vc_anim,
#   m3_vc, m3_vc_anim
# )
# 
# 
# m_vc_cerc <- m_vc%>%
#   group_by(mod, v_name)%>%
#   summarise(v_fit = median(v_est),
#             v_lwr = quantile(v_est, probs = 0.025),
#             v_upr = quantile(v_est, probs = 0.975),
#             )%>%
#   ungroup()%>%
#   mutate(trait = "cerc")
# 
# m_vc_cerc2 <- m_vc_cerc%>%filter(v_name %in% c("Ve2", "Vg2", "Vm2", "Vr2")) # var with fixed effects
# m_vc_cerc <- m_vc_cerc%>%filter(!v_name %in% c("Ve2", "Vg2", "Vm2", "Vr2")) # var excluding fixed effects
```


Here is the percent variation in cercomere presence accounted for by copepod stage across models. It is very consistent.

```{r}
filter(m_vc_cerc, v_name == "Vf")%>%
  mutate(across(v_fit:v_upr, function(x){round(x,3)}))
```

Here is the percent variation in cercomere presence accounted for by shared environments (copepod block). Some of the environmental variation in the base model is attributed to genes after adding full-sibs or the pedigree.

```{r}
filter(m_vc_cerc, v_name == "Ve")%>%
  mutate(across(v_fit:v_upr, function(x){round(x,3)}))
```

Here is the percent variation in cercomere presence accounted for by genetic variance (i.e. the heritability). Note that the variance component for full-sibs was multipled by two since full-sibs share just half their alleles on average. The estimate from the animal model was lower than the full-sib models. The addition of maternal effects decreased heritability estimates by ~5 points.

```{r}
filter(m_vc_cerc, v_name == "Vg")%>%
  mutate(across(v_fit:v_upr, function(x){round(x,3)}))
```


```{r}
mean_cerc <- sum(dat_cerc$cerc_bi, na.rm = T)/sum(!is.na(dat_cerc$cerc_bi), na.rm = T)
re_fams_cerc <- dat_cerc%>%
  filter(is.na(pred))%>%
  group_by(worm_full_sib_fam_id, gen)%>%
  summarize(fam_avg = qnorm(sum(cerc_bi, na.rm = T)/sum(!is.na(cerc_bi), na.rm = T)) - qnorm(mean_cerc))%>%
  ungroup()%>%
  mutate(approach="unadjusted")

re <- colMeans(m2_cerc_fam$Sol[,which(grepl("worm_full_sib", colnames(m2_cerc_fam$Sol)))])
re <- data.frame(worm_full_sib_fam_id = gsub("worm_full_sib_fam_id.", "", x = names(re)),
                 fam_avg = re,
                 approach="model-estimated")
re$lwr <- apply(m2_cerc_fam$Sol[,which(grepl("worm_full_sib", colnames(m2_cerc_fam$Sol)))], MARGIN = 2, function(x){quantile(x, probs = 0.025 )})
re$upr <- apply(m2_cerc_fam$Sol[,which(grepl("worm_full_sib", colnames(m2_cerc_fam$Sol)))], MARGIN = 2, function(x){quantile(x, probs = 0.975 )})

re_fams_cerc <- bind_rows(
  re_fams_cerc,
  re
)%>%
  mutate(approach = fct_relevel(factor(approach), c("unadjusted", "model-estimated")))
re <- left_join(re, select(dat, worm_full_sib_fam_id, gen, line)%>%distinct())

ggplot(re_fams_cerc, aes(x = fam_avg)) +
  geom_histogram(aes(y=..density..),
                 colour = "black",
                 fill = "white") +
  geom_function(data = filter(re_fams_cerc, approach == "unadjusted"),
                fun = dnorm, 
                args = list(mean = 0, 
                            sd = re_fams_cerc%>%
                              filter(approach == "unadjusted", !is.infinite(fam_avg))%>%.$fam_avg%>%sd(., na.rm=T)
                            ),
                color = 'red',
                linetype = "dashed") +
  geom_function(data = filter(re_fams_cerc, approach == "model-estimated"),
                fun = dnorm, 
                args = list(mean = 0, 
                            sd = re_fams_cerc%>%filter(approach == "model-estimated")%>%.$fam_avg%>%sd(., na.rm=T)
                            ),
                color = 'red',
                linetype = "dashed") +
  geom_density() +
  geom_label(data = re_fams_cerc%>%
               filter(!is.infinite(fam_avg))%>%
               group_by(approach)%>%
               summarise(min_x = min(fam_avg, na.rm = T),
                         skew = e1071::skewness(fam_avg, na.rm = T))%>%
               mutate(skew_c = paste0("Skewness: ", round(skew, 2))),
             aes(label = skew_c,
                 x = min_x,
                 y = 0.5),
             hjust = -0.05, vjust = 0) +
  facet_wrap(~approach, scales = "free_x") +
  labs(y = "Density", x = "Family averages (probit transformed)")
```

Here is how families from the different line fall. By the final generation, the families from the fast line mostly have positive random effects whereas those from the slow line mostly have negative random effects.

```{r}
ggplot(re%>%mutate(worm_full_sib_fam_id = fct_reorder(worm_full_sib_fam_id, fam_avg)),
       aes(x = worm_full_sib_fam_id, y = fam_avg, color = line)) +
  geom_hline(yintercept = 0) +
  geom_pointrange(aes(ymin = lwr, ymax = upr)) +
  scale_color_manual(values = c("gray", "black", "red","blue")) +
  facet_wrap(~factor(gen), scales = "free_x") +
  theme(panel.grid = element_blank())
```

Here are the parental influences on cercomere presence (i.e. the effect of dam/sire).

```{r}
filter(m_vc_cerc, v_name == "Vm")%>%
  mutate(across(v_fit:v_upr, function(x){round(x,3)}))
```

Here is the total variance explained by each model.

```{r}
filter(m_vc_cerc, v_name == "Vr")%>%
  mutate(across(v_fit:v_upr, function(x){round(1-x,3)}))
```

The fraction of phenotypic variance explained by each term in the different models is plotted here

```{r}
ggplot(m_vc_cerc%>%
         mutate(v_name = fct_relevel(factor(v_name), c("Vf", "Ve", "Vg", "Vm", "Vr"))),
       aes(v_name, v_fit, color = mod)) +
  geom_pointrange(aes(ymin = v_lwr, ymax = v_upr),
                  position = position_dodge(width = 0.6)) +
  labs(y = "Fraction variance explained", title = "Cercomere presence") +
  scale_x_discrete(labels = c("Starting\nconditions", "Shared\nenvironment", "Additive\ngenetic",
                              "Parental", "Residual")) +
  theme(panel.grid.major.x = element_blank(),
        axis.title.x = element_blank(),
        legend.title = element_blank())
```

Here is the same plot, but just showing the fraction of variance explained after accounting for starting conditions (copepod stage).

```{r}
ggplot(m_vc_cerc2%>%
         mutate(v_name = fct_relevel(factor(v_name), c("Ve2", "Vg2", "Vm2", "Vr2"))),
       aes(v_name, v_fit, color = mod)) +
  geom_pointrange(aes(ymin = v_lwr, ymax = v_upr),
                  position = position_dodge(width = 0.6)) +
  labs(y = "Fraction variance explained", title = "Cercomere presence") +
  scale_x_discrete(labels = c("Shared\nenvironment", "Additive\ngenetic",
                              "Parental", "Residual")) +
  theme(panel.grid.major.x = element_blank(),
        axis.title.x = element_blank(),
        legend.title = element_blank())
```

## Procercoid Size

Here are the number of full-sib families and the number of offspring phenotyped per family.

```{r}
dat_proc%>%
  group_by(worm_full_sib_fam_id)%>%
  summarize(n_pheno = sum(!is.na(proc_size2)))%>%
  ungroup()%>%
  filter(!is.na(worm_full_sib_fam_id))%>%
  summarize(fs_fams = n(),
            min_phenotyped = min(n_pheno),
            lwrIQR_phenotyped = quantile(n_pheno, prob = 0.25),
            avg_phenotyped = mean(n_pheno),
            uprIQR_phenotyped = quantile(n_pheno, prob = 0.75),
            max_phenotyped = max(n_pheno)
            )
```

Here are the number of offspring phenotyped per dam/sire.

```{r}
dat_proc%>%
  group_by(maternal_match, worm_maternal_id)%>%
  summarize(n_pheno = sum(!is.na(proc_size2)))%>%
  group_by(maternal_match)%>%
  filter(!is.na(worm_maternal_id))%>%
  summarize(dams = n(),
            min_phenotyped = min(n_pheno),
            lwrIQR_phenotyped = quantile(n_pheno, prob = 0.25),
            avg_phenotyped = mean(n_pheno),
            uprIQR_phenotyped = quantile(n_pheno, prob = 0.75),
            max_phenotyped = max(n_pheno)
            )
```

```{r}
p_o_proc <- filter(dat_proc, !is.na(proc_size2))

parents <- dat_proc%>%
  group_by(eggs_clutch_id)%>%
  summarise(n_parent = n(), proc_p = mean(proc_size2))%>%
  na.omit()

offspring <- dat_proc%>%
  group_by(worm_full_sib_fam_id)%>%
  summarise(n_offspring = n(), proc_o = mean(proc_size2, na.rm = T))%>%
  na.omit()

parent_offspring <- left_join(parents, offspring, by = c("eggs_clutch_id" = "worm_full_sib_fam_id"))
```

For procercoid size, there is not much of a parenent-offspring regression.

```{r}
ggplot(parent_offspring, aes(proc_p, proc_o)) + geom_point() +geom_smooth(se=F)
```

The slope suggests a very low heritability. 

```{r}
summary(lm(proc_o ~ proc_p, data = parent_offspring%>%na.omit()))
```

But this does not account for environmental influences like copepod size or block. Thus, let's fit the same 5 models again.

```{r}
p1_g <- list(R = list(V = 1, nu = 0.002),
             G = list(G1 = list(V = 1, nu = 0.002)))
p2_g <- list(R = list(V = 1, nu = 0.002),
             G = list(G1 = list(V = 1, nu = 0.002),
                      G2 = list(V = 1, nu = 0.002)))
p3_g <- list(R = list(V = 1, nu = 0.002),
             G = list(G1 = list(V = 1, nu = 0.002),
                      G2 = list(V = 1, nu = 0.002),
                      G3 = list(V = 1, nu = 0.002)))

thin_proc <- 100
burnin_proc <- 1000
nit_proc <- thin_proc * target_samples + burnin_proc

# fixed effects, no ib
m1_proc_noib <- MCMCglmm(proc_size2 ~ cop_stage_checking, 
                 random = ~cop_block,
                 family = "gaussian",
                 data = dat_proc,
                 prior = p1_g,
                 verbose=FALSE,
                 nitt = nit_proc, 
                 thin = thin_proc,
                 burnin = burnin_proc
                )
# fixed effects
m1_proc <- MCMCglmm(proc_size2 ~ ib + cop_stage_checking, 
                 random = ~cop_block,
                 family = "gaussian",
                 data = dat_proc,
                 prior = p1_g,
                 verbose=FALSE,
                 nitt = nit_proc, 
                 thin = thin_proc,
                 burnin = burnin_proc
                )
# fam
m2_proc_fam <- MCMCglmm(proc_size2 ~ ib + cop_stage_checking, 
                 random = ~cop_block + worm_full_sib_fam_id,
                 family = "gaussian",
                 data = dat_proc,
                 prior = p2_g,
                 verbose=FALSE,
                 pr = T,
                 nitt = nit_proc, 
                 thin = thin_proc,
                 burnin = burnin_proc
                )
# full sib X line
m2_proc_famXline <- MCMCglmm(
  proc_size2 ~ ib + cop_stage_checking, 
  random = ~cop_block + idh(line):worm_full_sib_fam_id,
  family = "gaussian",
  data = dat_proc,
  prior = 
    list(R = list(V = 1, nu = 0.002),
             G = list(G1 = list(V = 1, nu = 0.002),
                      G2 = list(V = diag(4), nu = 0.002))),
  verbose=FALSE,
  pr = F,
  nitt = nit_proc, 
  thin = thin_proc,
  burnin = burnin_proc
  )
# anim model
m2_proc_anim <- MCMCglmm(proc_size2 ~ ib + cop_stage_checking, 
                 random = ~cop_block + id,
                 ginverse = list(id = Ainv),
                 family = "gaussian",
                 data = dat_proc,
                 prior = p2_g,
                 verbose=F,
                 nitt = nit_proc, 
                 thin = thin_proc,
                 burnin = burnin_proc
                )
# fam
m3_proc_mat <- MCMCglmm(proc_size2 ~ ib + cop_stage_checking, 
                 random = ~cop_block + worm_full_sib_fam_id + dam,
                 family = "gaussian",
                 data = dat_proc%>%mutate(dam = factor(dam)),
                 prior = p3_g,
                 verbose=FALSE,
                 pr=T,
                 nitt = nit_proc, 
                 thin = thin_proc,
                 burnin = burnin_proc
                )
# anim model
m3_proc_anim <- MCMCglmm(proc_size2 ~ ib + cop_stage_checking, 
                 random = ~cop_block + id + dam,
                 ginverse = list(id = Ainv),
                 family = "gaussian",
                 data = dat_proc%>%mutate(dam = factor(dam)),
                 prior = p3_g,
                 verbose=F,
                 nitt = nit_proc, 
                 thin = thin_proc,
                 burnin = burnin_proc
                )
```

As in the other models, adding full-sibs (red) and then the pedigree (green) appear to be improvements.

```{r}
plot(mcmc.list(
  m1_proc$Deviance, m2_proc_fam$Deviance, m2_proc_anim$Deviance
  ), density = F)
```

The full-sib model suggests that some variance within clutches is explained by parental effects.

```{r}
plot(mcmc.list(
  m2_proc_fam$Deviance, m3_proc_mat$Deviance
  ), density = F)
```

By contrast, the addition of maternal effects may even worsen the animal model.

```{r}
plot(mcmc.list(
  m2_proc_anim$Deviance, m3_proc_anim$Deviance
  ), density = F)
```

For, these models, we break down total variance into several components: fixed effects (starting conditions), shared environment, and genetic effects.

```{r}
m_vc_proc <- get_var_comps("proc", dat_proc, 'cop_block', 
                            m1_proc_noib, m1_proc,
                            m2_proc_fam, m2_proc_anim,
                            m3_proc_mat, m3_proc_anim)

m_vc_proc2 <- m_vc_proc%>%filter(v_name %in% c("Ve2", "Vg2", "Vm2", "Vr2")) # var with fixed effects
m_vc_proc <- m_vc_proc%>%filter(!v_name %in% c("Ve2", "Vg2", "Vm2", "Vr2")) # var excluding fixed effects
```

Here is the percent variation in procercoid size accounted for by copepod stage across models. It is very consistent.

```{r}
filter(m_vc_proc, v_name == "Vf")%>%
  mutate(across(v_fit:v_upr, function(x){round(x,3)}))
```

Here is the percent variation in procercoid size accounted for by shared environments (copepod block). Little 'block' variance is redistributed to genes after adding full-sibs or the pedigree.

```{r}
filter(m_vc_proc, v_name == "Ve")%>%
  mutate(across(v_fit:v_upr, function(x){round(x,3)}))
```

Here is the percent variation in procercoid size accounted for by genetic variance (i.e. the heritability). Note that the variance component for full-sibs was multipled by two since full-sibs share just half their alleles on average. The estimate from the animal model was higher than the full-sib models. The addition of maternal effects decreased heritability in the animal model more than the full-sib model.

```{r}
filter(m_vc_proc, v_name == "Vg")%>%
  mutate(across(v_fit:v_upr, function(x){round(x,3)}))
```

The distribution of family effects also looks good.

```{r}
mean_proc <- mean(dat_proc$proc_size2, na.rm = T)
re_fams_proc <- dat_proc%>%
  filter(is.na(pred))%>%
  group_by(worm_full_sib_fam_id, gen)%>%
  summarize(fam_avg = mean(proc_size2, na.rm = T) - mean_proc)%>%
  ungroup()%>%
  mutate(approach="unadjusted")

re <- colMeans(m2_proc_fam$Sol[,which(grepl("worm_full_sib", colnames(m2_proc_fam$Sol)))])
re <- data.frame(worm_full_sib_fam_id = gsub("worm_full_sib_fam_id.", "", x = names(re)),
                 fam_avg = re,
                 approach="model-estimated")

re_fams_proc <- bind_rows(
  re_fams_proc,
  re
)%>%
  mutate(approach = fct_relevel(factor(approach), c("unadjusted", "model-estimated")))

ggplot(re_fams_proc, aes(x = fam_avg)) +
  geom_histogram(aes(y=..density..),
                 colour = "black",
                 fill = "white") +
  geom_function(data = filter(re_fams_proc, approach == "unadjusted"),
                fun = dnorm, 
                args = list(mean = 0, 
                            sd = re_fams_proc%>%filter(approach == "unadjusted")%>%.$fam_avg%>%sd(., na.rm=T)
                            ),
                color = 'red',
                linetype = "dashed") +
  geom_function(data = filter(re_fams_proc, approach == "model-estimated"),
                fun = dnorm, 
                args = list(mean = 0, 
                            sd = re_fams_proc%>%filter(approach == "model-estimated")%>%.$fam_avg%>%sd(., na.rm=T)
                            ),
                color = 'red',
                linetype = "dashed") +
  geom_density() +
  geom_label(data = re_fams_proc%>%
               filter(!is.infinite(fam_avg))%>%
               group_by(approach)%>%
               summarise(min_x = min(fam_avg, na.rm = T),
                         skew = e1071::skewness(fam_avg, na.rm = T))%>%
               mutate(skew_c = paste0("Skewness: ", round(skew, 2))),
             aes(label = skew_c,
                 x = min_x,
                 y = 0.5),
             hjust = -0.05, vjust = 0) +
  facet_wrap(~approach, scales = "free_x") +
  labs(y = "Density", x = "Family averages")
```

The influence of 'parent' on procercoid size are small.

```{r}
filter(m_vc_proc, v_name == "Vm")%>%
  mutate(across(v_fit:v_upr, function(x){round(x,3)}))
```

Here is the total variance explained by each model.

```{r}
filter(m_vc_proc, v_name == "Vr")%>%
  mutate(across(v_fit:v_upr, function(x){round(1-x,3)}))
```

The fraction of phenotypic variance explained by each term in the different models is plotted here

```{r}
ggplot(m_vc_proc%>%
         mutate(v_name = fct_relevel(factor(v_name), c("Vf", "Ve", "Vg", "Vm", "Vr"))),
       aes(v_name, v_fit, color = mod)) +
  geom_pointrange(aes(ymin = v_lwr, ymax = v_upr),
                  position = position_dodge(width = 0.6)) +
  labs(y = "Fraction variance explained", title = "Procercoid size") +
  scale_x_discrete(labels = c("Starting\nconditions", "Shared\nenvironment", "Additive\ngenetic", "Parental", "Residual")) +
  theme(panel.grid.major.x = element_blank(),
        axis.title.x = element_blank(),
        legend.title = element_blank())
```

Procercoid size seems to be impacted more by starting conditions and shared environment effects than larval developmental rate. This may explain why the parent-offspring regression underestimates heritability. Here is a comparison of the variance explained in the two traits by copepod stage and block combined in the different models.

```{r}
bind_rows(
  m_vc_cerc, m_vc_proc
)%>%
  filter(v_name %in% c("Vf", "Ve"))%>%
  group_by(mod, trait)%>%
  summarise(v_fit = sum(v_fit))%>%
  ggplot(., aes(x=mod, y=v_fit, color=trait)) + 
  geom_point() +
  labs(x = "Model", y = "Fixed + Env var") +
  theme(panel.grid.major.x = element_blank())
```

Therefore, it is worth looking at variance components after excluding the effect of starting conditions.

```{r}
ggplot(m_vc_proc2%>%
         mutate(v_name = fct_relevel(factor(v_name), c("Ve2", "Vg2", "Vm2", "Vr2"))),
       aes(v_name, v_fit, color = mod)) +
  geom_pointrange(aes(ymin = v_lwr, ymax = v_upr),
                  position = position_dodge(width = 0.6)) +
  labs(y = "Fraction variance explained", title = "Procercoid size") +
  scale_x_discrete(labels = c("Shared\nenvironment", "Additive\ngenetic", "Parental", "Residual")) +
  theme(panel.grid.major.x = element_blank(),
        axis.title.x = element_blank(),
        legend.title = element_blank())
```

## Developmental defects

When measuring worms at 13 dpi, some worms had clearly not developed normally; they lacked a cercomere after 13 days or had not grown much. For such failed development in copepods, we cannot use parent-offspring regression to evaluate heritability because all parents successfully developed.

```{r}
parents <- dat_fail%>%
  group_by(eggs_clutch_id)%>%
  summarise(n_parent = sum(!is.na(failed_devo_bi)), fail_p = mean(failed_devo_bi))%>%
  na.omit()

offspring <- dat_fail%>%
  group_by(worm_full_sib_fam_id)%>%
  summarise(n_offspring = sum(!is.na(failed_devo_bi)), fail_o = mean(failed_devo_bi, na.rm = T))%>%
  na.omit()

parent_offspring <- left_join(parents, offspring, by = c("eggs_clutch_id" = "worm_full_sib_fam_id"))
```

Here are the number of full-sib families and the number of offspring phenotyped per family.

```{r}
dat_fail%>%
  group_by(worm_full_sib_fam_id)%>%
  summarize(n_pheno = sum(!is.na(failed_devo_bi)))%>%
  ungroup()%>%
  filter(!is.na(worm_full_sib_fam_id))%>%
  summarize(fs_fams = n(),
            min_phenotyped = min(n_pheno),
            lwrIQR_phenotyped = quantile(n_pheno, prob = 0.25),
            avg_phenotyped = mean(n_pheno),
            uprIQR_phenotyped = quantile(n_pheno, prob = 0.75),
            max_phenotyped = max(n_pheno)
            )
```

Here are the number of offspring phenotyped per dam/sire.

```{r}
dat_fail%>%
  group_by(maternal_match, worm_maternal_id)%>%
  summarize(n_pheno = sum(!is.na(failed_devo_bi)))%>%
  group_by(maternal_match)%>%
  filter(!is.na(worm_maternal_id))%>%
  summarize(dams = n(),
            min_phenotyped = min(n_pheno),
            lwrIQR_phenotyped = quantile(n_pheno, prob = 0.25),
            avg_phenotyped = mean(n_pheno),
            uprIQR_phenotyped = quantile(n_pheno, prob = 0.75),
            max_phenotyped = max(n_pheno)
            )
```

Again, we will fit 5 models.

```{r}
p0_bi <- list(R = list(V = 1, fix = 1))

p1_bi <- list(R = list(V = 1, fix = 1),
              G = list(G1 = list(V = 1, nu = 1000, alpha.mu = 0, alpha.V = 1)))

p2_bi <- list(R = list(V = 1, fix = 1),
              G = list(G1 = list(V = 1, nu = 1000, alpha.mu = 0, alpha.V = 1),
                       G2 = list(V = 1, nu = 1000, alpha.mu = 0, alpha.V = 1)))
p3_bi <- list(R = list(V = 1, fix = 1),
              G = list(G1 = list(V = 1, nu = 1000, alpha.mu = 0, alpha.V = 1),
                       G2 = list(V = 1, nu = 1000, alpha.mu = 0, alpha.V = 1),
                       G3 = list(V = 1, nu = 1000, alpha.mu = 0, alpha.V = 1)))


thin_fail <- 250
burnin_fail <- 1000
nit_fail <- thin_fail * target_samples + burnin_fail


# fixed effects
m1_fail_noib <- MCMCglmm(failed_devo_bi ~ cop_stage_checking, 
                 random = ~cop_block,
                 family = "threshold",
                 data = dat_fail,
                 prior = p1_bi,
                 verbose=FALSE,
                 nitt = nit_fail, 
                 thin = thin_fail,
                 burnin = burnin_fail
                )
# fixed effects
m1_fail <- MCMCglmm(failed_devo_bi ~ ib + cop_stage_checking, 
                 random = ~cop_block,
                 family = "threshold",
                 data = dat_fail,
                 prior = p1_bi,
                 verbose=FALSE,
                 nitt = nit_fail, 
                 thin = thin_fail,
                 burnin = burnin_fail
                )
# full-sibs
m2_fail_fam <- MCMCglmm(failed_devo_bi ~ ib + cop_stage_checking, 
                 random = ~ cop_block + worm_full_sib_fam_id,
                 family = "threshold",
                 data = dat_fail,
                 prior = p2_bi,
                 verbose=F,
                 pr = T,
                 nitt = nit_fail, 
                 thin = thin_fail,
                 burnin = burnin_fail
                )
# full-sibs x line
m2_fail_famXline <- MCMCglmm(
  failed_devo_bi ~ ib + cop_stage_checking, 
  random = ~ cop_block + idh(line):worm_full_sib_fam_id,
  family = "threshold",
  data = dat_fail,
  prior = 
    list(R = list(V = 1, fix = 1),
         G = list(G1 = list(V = 1, nu = 1000, alpha.mu = 0, alpha.V = 1),
                  G2 = list(V = diag(4), nu = 0.002))),
  verbose=FALSE,
  pr = F,
  nitt = nit_fail, 
  thin = thin_fail,
  burnin = burnin_fail
  )
# anim model
m2_fail_anim <- MCMCglmm(failed_devo_bi ~ ib + cop_stage_checking, 
                 random = ~cop_block + id,
                 ginverse = list(id = Ainv),
                 family = "threshold",
                 data = dat_fail,
                 prior = p2_bi,
                 verbose=F,
                 nitt = nit_fail, 
                 thin = thin_fail,
                 burnin = burnin_fail
                )

# full-sibs + mat effx
m3_fail_mat <- MCMCglmm(failed_devo_bi ~ ib + cop_stage_checking, 
                 random = ~cop_block + worm_full_sib_fam_id + dam,
                 family = "threshold",
                 data = dat_fail%>%mutate(dam = factor(dam)),
                 prior = p3_bi,
                 verbose=F,
                 pr=T,
                 nitt = nit_fail, 
                 thin = thin_fail,
                 burnin = burnin_fail
                )
# anim model
m3_fail_anim <- MCMCglmm(failed_devo_bi ~ ib + cop_stage_checking, 
                 random = ~cop_block + id  + dam,
                 ginverse = list(id = Ainv),
                 family = "threshold",
                 data = dat_fail%>%mutate(dam = factor(dam)),
                 prior = p3_bi,
                 verbose=F,
                 nitt = nit_fail, 
                 thin = thin_fail,
                 burnin = burnin_fail
                )
```

Adding full-sibs (red) is an improvement, and the animal model (green) looks like a further improvement, though there must be more variation in the parameters.

```{r}
plot(mcmc.list(
  m1_fail$Deviance, m2_fail_fam$Deviance, m2_fail_anim$Deviance
  ), density = F)
```

The full-sib model suggests that there is little variance within clutches.

```{r}
plot(mcmc.list(
  m2_fail_fam$Deviance, m3_fail_mat$Deviance
  ), density = F)
```

By contrast, the addition of maternal effects does not improve the animal model.

```{r}
plot(mcmc.list(
  m2_fail_anim$Deviance, m3_fail_anim$Deviance
  ), density = F)
```

For, these models, we break down total variance into several components: fixed effects (starting conditions), shared environment, genetic effects, and parental effects.

```{r}
m_vc_fail <- get_var_comps("fail", dat_fail, 'cop_block', 
                            m1_fail_noib, m1_fail,
                            m2_fail_fam, m2_fail_anim,
                            m3_fail_mat, m3_fail_anim)

m_vc_fail2 <- m_vc_fail%>%filter(v_name %in% c("Ve2", "Vg2", "Vm2", "Vr2")) # var with fixed effects
m_vc_fail <- m_vc_fail%>%filter(!v_name %in% c("Ve2", "Vg2", "Vm2", "Vr2")) # var excluding fixed effects
```

Here is the percent variation in failed development accounted for by copepod stage across models. It is consistently low.

```{r}
filter(m_vc_fail, v_name == "Vf")%>%
  mutate(across(v_fit:v_upr, function(x){round(x,3)}))
```

Here is the percent variation in failed development accounted for by shared environments (copepod block). It is also low.

```{r}
filter(m_vc_fail, v_name == "Ve")%>%
  mutate(across(v_fit:v_upr, function(x){round(x,3)}))
```

Here is the percent variation in failed development accounted for by genetic variance (i.e. the heritability). It is high, indicating that defects run in families. The addition of maternal effects did not decrease heritability in the full sib model, but it did in the animal model.

```{r}
filter(m_vc_fail, v_name == "Vg")%>%
  mutate(across(v_fit:v_upr, function(x){round(x,3)}))
```

Here is the distribution of family effects. It is strongly skewed - most families have a very low likelihood of defects, but it is higher in some families.

```{r}
mean_fail <- sum(dat_fail$failed_devo_bi, na.rm = T)/sum(!is.na(dat_fail$failed_devo_bi), na.rm = T)
re_fams_fail <- dat_fail%>%
  filter(is.na(pred))%>%
  group_by(worm_full_sib_fam_id, gen)%>%
  summarize(fam_avg = -1*(qnorm(sum(failed_devo_bi, na.rm = T)/sum(!is.na(failed_devo_bi), na.rm = T)) - qnorm(mean_fail)))%>%
  ungroup()%>%
  mutate(approach="unadjusted")

re <- colMeans(m2_fail_fam$Sol[,which(grepl("worm_full_sib", colnames(m2_fail_fam$Sol)))])
re <- data.frame(worm_full_sib_fam_id = gsub("worm_full_sib_fam_id.", "", x = names(re)),
                 fam_avg = -1*re,
                 approach="model-estimated")

re_fams_fail <- bind_rows(
  re_fams_fail,
  re
)%>%
  mutate(approach = fct_relevel(factor(approach), c("unadjusted", "model-estimated")))

ggplot(re_fams_fail, aes(x = fam_avg)) +
  geom_histogram(aes(y=..density..),
                 colour = "black",
                 fill = "white") +
  geom_function(data = filter(re_fams_fail, approach == "unadjusted"),
                fun = dnorm, 
                args = list(mean = 0, 
                            sd = re_fams_fail%>%
                              filter(approach == "unadjusted", !is.infinite(fam_avg))%>%.$fam_avg%>%sd(., na.rm=T)
                            ),
                color = 'red',
                linetype = "dashed") +
  geom_function(data = filter(re_fams_fail, approach == "model-estimated"),
                fun = dnorm, 
                args = list(mean = 0, 
                            sd = re_fams_fail%>%filter(approach == "model-estimated")%>%.$fam_avg%>%sd(., na.rm=T)
                            ),
                color = 'red',
                linetype = "dashed") +
  geom_density() +
  geom_label(data = re_fams_fail%>%
               filter(!is.infinite(fam_avg))%>%
               group_by(approach)%>%
               summarise(min_x = min(fam_avg, na.rm = T),
                         skew = e1071::skewness(fam_avg, na.rm = T))%>%
               mutate(skew_c = paste0("Skewness: ", round(skew, 2))),
             aes(label = skew_c,
                 x = min_x,
                 y = 0.5),
             hjust = -0.05, vjust = 0) +
  facet_wrap(~approach, scales = "free_x") +
  labs(y = "Density", x = "Family averages (probit transformed)")
```

Here are the parental influences on failed development (i.e. the effect of dam/sire).

```{r}
filter(m_vc_fail, v_name == "Vm")%>%
  mutate(across(v_fit:v_upr, function(x){round(x,3)}))
```

Here is the total variance explained by each model.

```{r}
filter(m_vc_fail, v_name == "Vr")%>%
  mutate(across(v_fit:v_upr, function(x){round(1-x,3)}))
```


The fraction of phenotypic variance explained by each term in the different models is plotted here. Failed development is mostly a genetic pattern.

```{r}
ggplot(m_vc_fail%>%
         mutate(v_name = fct_relevel(factor(v_name), c("Vf", "Ve", "Vg", "Vm", "Vr"))),
       aes(v_name, v_fit, color = mod)) +
  geom_pointrange(aes(ymin = v_lwr, ymax = v_upr),
                  position = position_dodge(width = 0.6)) +
  labs(y = "Fraction variance explained", title = "Failed development") +
  scale_x_discrete(labels = c("Starting\nconditions", 
                              "Shared\nenvironment",
                              "Additive\ngenetic", "Parental", "Residual")) +
  theme(panel.grid.major.x = element_blank(),
        axis.title.x = element_blank(),
        legend.title = element_blank())
```

Since starting conditions explained little variation in developmental defects, little changes when it is excluded from variance calculations.

```{r}
ggplot(m_vc_fail2%>%
         mutate(v_name = fct_relevel(factor(v_name), c("Ve2", "Vg2", "Vm2", "Vr2"))),
       aes(v_name, v_fit, color = mod)) +
  geom_pointrange(aes(ymin = v_lwr, ymax = v_upr),
                  position = position_dodge(width = 0.6)) +
  labs(y = "Fraction variance explained", title = "Failed development") +
  scale_x_discrete(labels = c("Shared\nenvironment", "Additive\ngenetic", "Parental", "Residual")) +
  theme(panel.grid.major.x = element_blank(),
        axis.title.x = element_blank(),
        legend.title = element_blank())
```

## Copepod survival to day 13

On day 13 post infection, copepods were checked for survival. Usually this was done as part of measuring worm size, before exposing fish on day 14. However, in the final generation, some infected copepods were taken to infect fish at 12 dpe (to test the evolution of faster development). These copepods were thus excluded from the survival measurement. Since only surviving worms could parent the next generation, we cannot use parent-offspring regression to estimate heritability of host survival. Also, we only want to focus on the copepods that were infected because parasite genes will only impact the survival of infected copepods.

Here are the number of full-sib families and the number of offspring phenotyped per family.

```{r}
dat_surv13_2%>%
  group_by(worm_full_sib_fam_id)%>%
  summarize(n_pheno = sum(!is.na(cop_dead_13dpe)))%>%
  ungroup()%>%
  filter(!is.na(worm_full_sib_fam_id))%>%
  summarize(fs_fams = n(),
            min_phenotyped = min(n_pheno),
            lwrIQR_phenotyped = quantile(n_pheno, prob = 0.25),
            avg_phenotyped = mean(n_pheno),
            uprIQR_phenotyped = quantile(n_pheno, prob = 0.75),
            max_phenotyped = max(n_pheno)
            )
```

Here are the number of offspring phenotyped per dam/sire.

```{r}
dat_surv13_2%>%
  group_by(maternal_match, worm_maternal_id)%>%
  summarize(n_pheno = sum(!is.na(cop_dead_13dpe)))%>%
  group_by(maternal_match)%>%
  filter(!is.na(worm_maternal_id))%>%
  summarize(dams = n(),
            min_phenotyped = min(n_pheno),
            lwrIQR_phenotyped = quantile(n_pheno, prob = 0.25),
            avg_phenotyped = mean(n_pheno),
            uprIQR_phenotyped = quantile(n_pheno, prob = 0.75),
            max_phenotyped = max(n_pheno)
            )
```

Again, we will fit 5 models.

```{r}
p0_bi <- list(R = list(V = 1, fix = 1))

p1_bi <- list(R = list(V = 1, fix = 1),
              G = list(G1 = list(V = 1, nu = 1000, alpha.mu = 0, alpha.V = 1)))

p2_bi <- list(R = list(V = 1, fix = 1),
              G = list(G1 = list(V = 1, nu = 1000, alpha.mu = 0, alpha.V = 1),
                       G2 = list(V = 1, nu = 1000, alpha.mu = 0, alpha.V = 1)))
p3_bi <- list(R = list(V = 1, fix = 1),
              G = list(G1 = list(V = 1, nu = 1000, alpha.mu = 0, alpha.V = 1),
                       G2 = list(V = 1, nu = 1000, alpha.mu = 0, alpha.V = 1),
                       G3 = list(V = 1, nu = 1000, alpha.mu = 0, alpha.V = 1)))


thin_surv <- 75
burnin_surv <- 1000
nit_surv <- thin_surv * target_samples + burnin_surv


# fixed effects, no ib
m1_surv_noib <- MCMCglmm(cop_dead_13dpe ~ cop_stage_checking, 
                 random = ~cop_block,
                 family = "threshold",
                 data = dat_surv13_2,
                 prior = p1_bi,
                 verbose=FALSE,
                 nitt = nit_surv, 
                 thin = thin_surv,
                 burnin = burnin_surv
                )
# fixed effects
m1_surv <- MCMCglmm(cop_dead_13dpe ~ ib + cop_stage_checking, 
                 random = ~cop_block,
                 family = "threshold",
                 data = dat_surv13_2,
                 prior = p1_bi,
                 verbose=FALSE,
                 nitt = nit_surv, 
                 thin = thin_surv,
                 burnin = burnin_surv
                )
# full-sibs
m2_surv_fam <- MCMCglmm(cop_dead_13dpe ~ ib + cop_stage_checking, 
                 random = ~ cop_block + worm_full_sib_fam_id,
                 family = "threshold",
                 data = dat_surv13_2,
                 prior = p2_bi,
                 verbose=F,
                 pr = T,
                 nitt = nit_surv, 
                 thin = thin_surv,
                 burnin = burnin_surv
                )
# full-sibs x line
m2_surv_famXline <- MCMCglmm(
  cop_dead_13dpe ~ ib + cop_stage_checking,
  random = ~ cop_block + idh(line):worm_full_sib_fam_id,
  family = "threshold",
  data = dat_surv13_2,
  prior = 
    list(R = list(V = 1, fix = 1),
         G = list(G1 = list(V = 1, nu = 1000, alpha.mu = 0, alpha.V = 1),
                  G2 = list(V = diag(4), nu = 0.002))),
  verbose=FALSE,
  pr = F,
  nitt = nit_surv, 
  thin = thin_surv,
  burnin = burnin_surv
  )
# anim model
m2_surv_anim <- MCMCglmm(cop_dead_13dpe ~ ib + cop_stage_checking, 
                 random = ~cop_block + id,
                 ginverse = list(id = Ainv),
                 family = "threshold",
                 data = dat_surv13_2,
                 prior = p2_bi,
                 verbose=F,
                 nitt = nit_surv, 
                 thin = thin_surv,
                 burnin = burnin_surv
                )

# full-sibs + mat effx
m3_surv_mat <- MCMCglmm(cop_dead_13dpe ~ ib + cop_stage_checking, 
                 random = ~cop_block + worm_full_sib_fam_id + dam,
                 family = "threshold",
                 data = dat_surv13_2%>%mutate(dam = factor(dam)),
                 prior = p3_bi,
                 verbose=F,
                 pr=T,
                 nitt = nit_surv, 
                 thin = thin_surv,
                 burnin = burnin_surv
                )
# anim model
m3_surv_anim <- MCMCglmm(cop_dead_13dpe ~ ib + cop_stage_checking, 
                 random = ~cop_block + id  + dam,
                 ginverse = list(id = Ainv),
                 family = "threshold",
                 data = dat_surv13_2%>%mutate(dam = factor(dam)),
                 prior = p3_bi,
                 verbose=F,
                 nitt = nit_surv, 
                 thin = thin_surv,
                 burnin = burnin_surv
                )
```

Adding full-sibs (red) and/or the pedigree (green) looks like a slight improvement.

```{r}
plot(mcmc.list(
  m1_surv$Deviance, m2_surv_fam$Deviance, m2_surv_anim$Deviance
  ), density = F)
```

There does not appear to be any parental effects on copepod survival in the full sib model...

```{r}
plot(mcmc.list(
  m2_surv_fam$Deviance, m3_surv_mat$Deviance
  ), density = F)
```

...or the animal model.

```{r}
plot(mcmc.list(
  m2_surv_anim$Deviance, m3_surv_anim$Deviance
  ), density = F)
```

For, these models, we break down total variance into several components: fixed effects (starting conditions), shared environment, genetic effects, and parental effects.

```{r}
m_vc_surv <- get_var_comps("surv", dat_surv13_2, 'cop_block', 
                            m1_surv_noib, m1_surv,
                            m2_surv_fam, m2_surv_anim,
                            m3_surv_mat, m3_surv_anim)

m_vc_surv2 <- m_vc_surv%>%filter(v_name %in% c("Ve2", "Vg2", "Vm2", "Vr2")) # var with fixed effects
m_vc_surv <- m_vc_surv%>%filter(!v_name %in% c("Ve2", "Vg2", "Vm2", "Vr2")) # var excluding fixed effects
```

Here is the percent variation in copepod survival accounted for by copepod stage across models. It is consistent.

```{r}
filter(m_vc_surv, v_name == "Vf")%>%
  mutate(across(v_fit:v_upr, function(x){round(x,3)}))
```

Here is the percent variation in copepod survival accounted for by shared environments (copepod block). It is also consistent.

```{r}
filter(m_vc_surv, v_name == "Ve")%>%
  mutate(across(v_fit:v_upr, function(x){round(x,3)}))
```

Here is the percent variation in copepod survival accounted for by genetic variance (i.e. the heritability). It is low.

```{r}
filter(m_vc_surv, v_name == "Vg")%>%
  mutate(across(v_fit:v_upr, function(x){round(x,3)}))
```

And the estimated family effects are normally distributed.

```{r}
mean_surv <- sum(dat_surv13_2$cop_dead_13dpe, na.rm = T)/sum(!is.na(dat_surv13_2$cop_dead_13dpe), na.rm = T)
re_fams_surv <- dat_surv13_2%>%
  filter(is.na(pred))%>%
  group_by(worm_full_sib_fam_id, gen)%>%
  summarize(fam_avg = -1*(qnorm(sum(cop_dead_13dpe, na.rm = T)/sum(!is.na(cop_dead_13dpe), na.rm = T)) - qnorm(mean_surv)))%>%
  ungroup()%>%
  mutate(approach="unadjusted")

re <- colMeans(m2_surv_fam$Sol[,which(grepl("worm_full_sib", colnames(m2_surv_fam$Sol)))])
re <- data.frame(worm_full_sib_fam_id = gsub("worm_full_sib_fam_id.", "", x = names(re)),
                 fam_avg = -1*re,
                 approach="model-estimated")

re_fams_surv <- bind_rows(
  re_fams_surv,
  re
)%>%
  mutate(approach = fct_relevel(factor(approach), c("unadjusted", "model-estimated")))

ggplot(re_fams_surv, aes(x = fam_avg)) +
  geom_histogram(aes(y=..density..),
                 colour = "black",
                 fill = "white") +
  geom_function(data = filter(re_fams_surv, approach == "unadjusted"),
                fun = dnorm, 
                args = list(mean = 0, 
                            sd = re_fams_surv%>%
                              filter(approach == "unadjusted", !is.infinite(fam_avg))%>%.$fam_avg%>%sd(., na.rm=T)
                            ),
                color = 'red',
                linetype = "dashed") +
  geom_function(data = filter(re_fams_surv, approach == "model-estimated"),
                fun = dnorm, 
                args = list(mean = 0, 
                            sd = re_fams_surv%>%filter(approach == "model-estimated")%>%.$fam_avg%>%sd(., na.rm=T)
                            ),
                color = 'red',
                linetype = "dashed") +
  geom_density() +
  geom_label(data = re_fams_surv%>%
               filter(!is.infinite(fam_avg))%>%
               group_by(approach)%>%
               summarise(min_x = min(fam_avg, na.rm = T),
                         skew = e1071::skewness(fam_avg, na.rm = T))%>%
               mutate(skew_c = paste0("Skewness: ", round(skew, 2))),
             aes(label = skew_c,
                 x = min_x,
                 y = 2),
             hjust = -0.05, vjust = 0) +
  facet_wrap(~approach, scales = "free_x") +
  labs(y = "Density", x = "Family averages (probit transformed)")
```

And there is essentially no influence of dam/sire.

```{r}
filter(m_vc_surv, v_name == "Vm")%>%
  mutate(across(v_fit:v_upr, function(x){round(x,3)}))
```

Here is the total variance explained by each model.

```{r}
filter(m_vc_surv, v_name == "Vr")%>%
  mutate(across(v_fit:v_upr, function(x){round(1-x,3)}))
```

The fraction of phenotypic variance explained by each term in the different models is plotted here. copepod survival is mostly a genetic pattern.

```{r}
ggplot(m_vc_surv%>%
         mutate(v_name = fct_relevel(factor(v_name), c("Vf", "Ve", "Vg", "Vm", "Vr"))),
       aes(v_name, v_fit, color = mod)) +
  geom_pointrange(aes(ymin = v_lwr, ymax = v_upr),
                  position = position_dodge(width = 0.6)) +
  labs(y = "Fraction variance explained", title = "Copepod survival") +
  scale_x_discrete(labels = c("Starting\nconditions", 
                              "Shared\nenvironment",
                              "Additive\ngenetic", "Parental", "Residual")) +
  theme(panel.grid.major.x = element_blank(),
        axis.title.x = element_blank(),
        legend.title = element_blank())
```

This values do not change much when we exclude the effect of the starting conditions.

```{r}
ggplot(m_vc_surv2%>%
         mutate(v_name = fct_relevel(factor(v_name), c("Ve2", "Vg2", "Vm2", "Vr2"))),
       aes(v_name, v_fit, color = mod)) +
  geom_pointrange(aes(ymin = v_lwr, ymax = v_upr),
                  position = position_dodge(width = 0.6)) +
  labs(y = "Fraction variance explained", title = "Copepod survival") +
  scale_x_discrete(labels = c("Shared\nenvironment", "Additive\ngenetic", "Parental", "Residual")) +
  theme(panel.grid.major.x = element_blank(),
        axis.title.x = element_blank(),
        legend.title = element_blank())
```

# Traits in fish

Let's turn our attention from the first host to the second host, the stickleback. 

## Fish infection rate

All parental worms successfully infected fish, so we cannot use parent-offspring regression to estimate heritability.

Here are the number of full-sib families and the number of offspring phenotyped per family.

```{r}
dat_finf%>%
  group_by(worm_full_sib_fam_id)%>%
  summarize(n_pheno = sum(!is.na(schisto_inf)))%>%
  ungroup()%>%
  filter(!is.na(worm_full_sib_fam_id))%>%
  summarize(fs_fams = n(),
            min_phenotyped = min(n_pheno),
            lwrIQR_phenotyped = quantile(n_pheno, prob = 0.25),
            avg_phenotyped = mean(n_pheno),
            uprIQR_phenotyped = quantile(n_pheno, prob = 0.75),
            max_phenotyped = max(n_pheno)
            )
```

Here are the number of offspring phenotyped per dam/sire. There are only about 6 per dam. Fewer than this, makes the variance component challenging to estimate.

```{r}
dat_finf%>%
  group_by(maternal_match, worm_maternal_id)%>%
  summarize(n_pheno = sum(!is.na(schisto_inf)))%>%
  group_by(maternal_match)%>%
  filter(!is.na(worm_maternal_id))%>%
  summarize(dams = n(),
            min_phenotyped = min(n_pheno),
            lwrIQR_phenotyped = quantile(n_pheno, prob = 0.25),
            avg_phenotyped = mean(n_pheno),
            uprIQR_phenotyped = quantile(n_pheno, prob = 0.75),
            max_phenotyped = max(n_pheno)
            )
```

Again, we will fit 5 models.

```{r}
p0_bi <- list(R = list(V = 1, fix = 1))

p1_bi <- list(R = list(V = 1, fix = 1),
              G = list(G1 = list(V = 1, nu = 1000, alpha.mu = 0, alpha.V = 1)))

p2_bi <- list(R = list(V = 1, fix = 1),
              G = list(G1 = list(V = 1, nu = 1000, alpha.mu = 0, alpha.V = 1),
                       G2 = list(V = 1, nu = 1000, alpha.mu = 0, alpha.V = 1)))
p3_bi <- list(R = list(V = 1, fix = 1),
              G = list(G1 = list(V = 1, nu = 1000, alpha.mu = 0, alpha.V = 1),
                       G2 = list(V = 1, nu = 1000, alpha.mu = 0, alpha.V = 1),
                       G3 = list(V = 1, nu = 1000, alpha.mu = 0, alpha.V = 1)))


thin_finf <- 75
burnin_finf <- 1000
nit_finf <- thin_finf * target_samples + burnin_finf


# fixed effects, no ib
m1_finf_noib <- MCMCglmm(schisto_inf ~ fish_initial_bm_cen + fish_sex_cen + fish_exp_on_cop_dpe_fct,  
                 random = ~cop_block,
                 family = "threshold",
                 data = dat_finf,
                 prior = p1_bi,
                 verbose=FALSE,
                 nitt = nit_finf, 
                 thin = thin_finf,
                 burnin = burnin_finf
                )
# fixed effects
m1_finf <- MCMCglmm(schisto_inf ~ ib + fish_initial_bm_cen + fish_sex_cen + fish_exp_on_cop_dpe_fct,  
                 random = ~cop_block,
                 family = "threshold",
                 data = dat_finf,
                 prior = p1_bi,
                 verbose=FALSE,
                 nitt = nit_finf, 
                 thin = thin_finf,
                 burnin = burnin_finf
                )
# full-sibs
m2_finf_fam <- MCMCglmm(schisto_inf ~ ib + fish_initial_bm_cen + fish_sex_cen + fish_exp_on_cop_dpe_fct,  
                 random = ~ cop_block + worm_full_sib_fam_id,
                 family = "threshold",
                 data = dat_finf,
                 prior = p2_bi,
                 verbose=F,
                 pr=T,
                 nitt = nit_finf, 
                 thin = thin_finf,
                 burnin = burnin_finf
                )
# full-sibs x line
m2_finf_famXline <- MCMCglmm(
  schisto_inf ~ ib + fish_initial_bm_cen + fish_sex_cen + fish_exp_on_cop_dpe_fct,
  random = ~ cop_block + idh(line):worm_full_sib_fam_id,
  family = "threshold",
  data = dat_finf,
  prior = 
    list(R = list(V = 1, fix = 1),
         G = list(G1 = list(V = 1, nu = 1000, alpha.mu = 0, alpha.V = 1),
                  G2 = list(V = diag(4), nu = 0.002))),
  verbose=FALSE,
  pr = F,
  nitt = nit_finf, 
  thin = thin_finf,
  burnin = burnin_finf
  )
# anim model
m2_finf_anim <- MCMCglmm(schisto_inf ~ ib + fish_initial_bm_cen + fish_sex_cen + fish_exp_on_cop_dpe_fct, 
                 random = ~cop_block + id,
                 ginverse = list(id = Ainv),
                 family = "threshold",
                 data = dat_finf,
                 prior = p2_bi,
                 verbose=F,
                 nitt = nit_finf, 
                 thin = thin_finf,
                 burnin = burnin_finf
                )

# full-sibs + mat effx
m3_finf_mat <- MCMCglmm(schisto_inf ~ ib + fish_initial_bm_cen + fish_sex_cen + fish_exp_on_cop_dpe_fct, 
                 random = ~cop_block + worm_full_sib_fam_id + dam,
                 family = "threshold",
                 data = dat_finf%>%mutate(dam = factor(dam)),
                 prior = p3_bi,
                 verbose=F,
                 nitt = nit_finf, 
                 thin = thin_finf,
                 burnin = burnin_finf
                )
# anim model
m3_finf_anim <- MCMCglmm(schisto_inf ~ ib + fish_initial_bm_cen + fish_sex_cen + fish_exp_on_cop_dpe_fct, 
                 random = ~cop_block + id  + dam,
                 ginverse = list(id = Ainv),
                 family = "threshold",
                 data = dat_finf%>%mutate(dam = factor(dam)),
                 prior = p3_bi,
                 verbose=F,
                 nitt = nit_finf, 
                 thin = thin_finf,
                 burnin = burnin_finf
                )
```

Adding full-sibs (red) is an improvement, and the animal model (green) looks like a further (noisy) improvement.

```{r}
plot(mcmc.list(
  m1_finf$Deviance, m2_finf_fam$Deviance, m2_finf_anim$Deviance
  ), density = F)
```

The full-sib model suggests that there is little variance within clutches.

```{r}
plot(mcmc.list(
  m2_finf_fam$Deviance, m3_finf_mat$Deviance
  ), density = F)
```

By contrast, the addition of maternal effects does not improve the animal model.

```{r}
plot(mcmc.list(
  m2_finf_anim$Deviance, m3_finf_anim$Deviance
  ), density = F)
```

For, these models, we break down total variance into several components: fixed effects (starting conditions), shared environment, genetic effects, and parental effects.

```{r}
m_vc_finf <- get_var_comps("finf", dat_finf, 'cop_block', 
                            m1_finf_noib, m1_finf,
                            m2_finf_fam, m2_finf_anim,
                            m3_finf_mat, m3_finf_anim)

m_vc_finf2 <- m_vc_finf%>%filter(v_name %in% c("Ve2", "Vg2", "Vm2", "Vr2")) # var with fixed effects
m_vc_finf <- m_vc_finf%>%filter(!v_name %in% c("Ve2", "Vg2", "Vm2", "Vr2")) # var excluding fixed effects
```

Here is the percent variation in fish infection accounted for by starting conditions (initial fish size, sex, and day of exposure). It is positive but low.

```{r}
filter(m_vc_finf, v_name == "Vf")%>%
  mutate(across(v_fit:v_upr, function(x){round(x,3)}))
```

Here is the percent variation in fish infection accounted for by shared environments (copepod block).

```{r}
filter(m_vc_finf, v_name == "Ve")%>%
  mutate(across(v_fit:v_upr, function(x){round(x,3)}))
```

Here is the percent variation in fish infection accounted for by genetic variance (i.e. the heritability). It is rather low, though it is a few points higher in the animal model. The addition of maternal effects did not decrease heritability much.

```{r}
filter(m_vc_finf, v_name == "Vg")%>%
  mutate(across(v_fit:v_upr, function(x){round(x,3)}))
```

The distribution of family effects is left-skewed, suggesting there is more genetic variance for low infectivity than high infectivity.

```{r}
mean_finf <- sum(dat_finf$schisto_inf, na.rm = T)/sum(!is.na(dat_finf$schisto_inf), na.rm = T)
re_fams_finf <- dat_finf%>%
  filter(is.na(pred))%>%
  group_by(worm_full_sib_fam_id, gen)%>%
  summarize(fam_avg = qnorm(sum(schisto_inf, na.rm = T)/sum(!is.na(schisto_inf), na.rm = T)) - qnorm(mean_finf))%>%
  ungroup()%>%
  mutate(approach="unadjusted")

re <- colMeans(m2_finf_fam$Sol[,which(grepl("worm_full_sib", colnames(m2_finf_fam$Sol)))])
re <- data.frame(worm_full_sib_fam_id = gsub("worm_full_sib_fam_id.", "", x = names(re)),
                 fam_avg = re,
                 approach="model-estimated")

re_fams_finf <- bind_rows(
  re_fams_finf,
  re
)%>%
  mutate(approach = fct_relevel(factor(approach), c("unadjusted", "model-estimated")))

ggplot(re_fams_finf, aes(x = fam_avg)) +
  geom_histogram(aes(y=..density..),
                 colour = "black",
                 fill = "white") +
  geom_function(data = filter(re_fams_finf, approach == "unadjusted"),
                fun = dnorm, 
                args = list(mean = 0, 
                            sd = re_fams_finf%>%
                              filter(approach == "unadjusted", !is.infinite(fam_avg))%>%.$fam_avg%>%sd(., na.rm=T)
                            ),
                color = 'red',
                linetype = "dashed") +
  geom_function(data = filter(re_fams_finf, approach == "model-estimated"),
                fun = dnorm, 
                args = list(mean = 0, 
                            sd = re_fams_finf%>%filter(approach == "model-estimated")%>%.$fam_avg%>%sd(., na.rm=T)
                            ),
                color = 'red',
                linetype = "dashed") +
  geom_density() +
  geom_label(data = re_fams_finf%>%
               filter(!is.infinite(fam_avg))%>%
               group_by(approach)%>%
               summarise(min_x = min(fam_avg, na.rm = T),
                         skew = e1071::skewness(fam_avg, na.rm = T))%>%
               mutate(skew_c = paste0("Skewness: ", round(skew, 2))),
             aes(label = skew_c,
                 x = min_x,
                 y = 2),
             hjust = -0.05, vjust = 0) +
  facet_wrap(~approach, scales = "free_x") +
  labs(y = "Density", x = "Family averages (probit transformed)")
```

Here are the parental influences on fish infection (i.e. the effect of dam/sire).

```{r}
filter(m_vc_finf, v_name == "Vm")%>%
  mutate(across(v_fit:v_upr, function(x){round(x,3)}))
```

Here is the total variance explained by each model.

```{r}
filter(m_vc_finf, v_name == "Vr")%>%
  mutate(across(v_fit:v_upr, function(x){round(1-x,3)}))
```

The fraction of phenotypic variance explained by each term in the different models is plotted here. 

```{r}
ggplot(m_vc_finf%>%
         mutate(v_name = fct_relevel(factor(v_name), c("Vf", "Ve", "Vg", "Vm", "Vr"))),
       aes(v_name, v_fit, color = mod)) +
  geom_pointrange(aes(ymin = v_lwr, ymax = v_upr),
                  position = position_dodge(width = 0.6)) +
  labs(y = "Fraction variance explained", title = "Fish infection rate") +
  scale_x_discrete(labels = c("Starting\nconditions", 
                              "Shared\nenvironment",
                              "Additive\ngenetic", "Parental", "Residual")) +
  theme(panel.grid.major.x = element_blank(),
        axis.title.x = element_blank(),
        legend.title = element_blank())
```

Since fish size and sex had little impact on infectivity, removing the fixed effects did not increase heritability much.

```{r}
ggplot(m_vc_finf2%>%
         mutate(v_name = fct_relevel(factor(v_name), c("Ve2", "Vg2", "Vm2", "Vr2"))),
       aes(v_name, v_fit, color = mod)) +
  geom_pointrange(aes(ymin = v_lwr, ymax = v_upr),
                  position = position_dodge(width = 0.6)) +
  labs(y = "Fraction variance explained", title = "Fish infection rate") +
  scale_x_discrete(labels = c("Shared\nenvironment", "Additive\ngenetic", "Parental", "Residual")) +
  theme(panel.grid.major.x = element_blank(),
        axis.title.x = element_blank(),
        legend.title = element_blank())
```

## Fish survival

All parental worms were from fish that survived until dissection, so we cannot use parent-offspring regression to estimate heritability.

Here are the number of full-sib families and the number of offspring phenotyped per family.

```{r}
dat_fsurv2_75dpi%>%
  group_by(worm_full_sib_fam_id)%>%
  summarize(n_pheno = sum(!is.na(fish_surv75)))%>%
  ungroup()%>%
  filter(!is.na(worm_full_sib_fam_id))%>%
  summarize(fs_fams = n(),
            min_phenotyped = min(n_pheno),
            lwrIQR_phenotyped = quantile(n_pheno, prob = 0.25),
            avg_phenotyped = mean(n_pheno),
            uprIQR_phenotyped = quantile(n_pheno, prob = 0.75),
            max_phenotyped = max(n_pheno)
            )
```

Here are the number of offspring phenotyped per dam/sire. There are only about 3 per dam, so it is challenging to estimate this variance component.

```{r}
dat_fsurv2_75dpi%>%
  group_by(maternal_match, worm_maternal_id)%>%
  summarize(n_pheno = sum(!is.na(fish_surv75)))%>%
  group_by(maternal_match)%>%
  filter(!is.na(worm_maternal_id))%>%
  summarize(dams = n(),
            min_phenotyped = min(n_pheno),
            lwrIQR_phenotyped = quantile(n_pheno, prob = 0.25),
            avg_phenotyped = mean(n_pheno),
            uprIQR_phenotyped = quantile(n_pheno, prob = 0.75),
            max_phenotyped = max(n_pheno)
            )
```

Again, we will fit 5 models.

```{r}
thin_fsurv <- 200
burnin_fsurv <- 1000
nit_fsurv <- thin_fsurv * target_samples + burnin_fsurv


# fixed effects, no ib
m1_fsurv_noib <- MCMCglmm(fish_surv75 ~ fish_initial_bm_cen + fish_sex_cen,  
                 random = ~tank_id,
                 family = "threshold",
                 data = dat_fsurv2_75dpi,
                 prior = p1_bi,
                 verbose=FALSE,
                 nitt = nit_fsurv, 
                 thin = thin_fsurv,
                 burnin = burnin_fsurv
                )
# fixed effects
m1_fsurv <- MCMCglmm(fish_surv75 ~ ib + fish_initial_bm_cen + fish_sex_cen,  
                 random = ~tank_id,
                 family = "threshold",
                 data = dat_fsurv2_75dpi,
                 prior = p1_bi,
                 verbose=FALSE,
                 nitt = nit_fsurv, 
                 thin = thin_fsurv,
                 burnin = burnin_fsurv
                )
# full-sibs
m2_fsurv_fam <- MCMCglmm(fish_surv75 ~ ib + fish_initial_bm_cen + fish_sex_cen,  
                 random = ~ tank_id + worm_full_sib_fam_id,
                 family = "threshold",
                 data = dat_fsurv2_75dpi,
                 prior = p2_bi,
                 verbose=F,
                 pr=T,
                 nitt = nit_fsurv, 
                 thin = thin_fsurv,
                 burnin = burnin_fsurv
                )
# full-sibs x line
m2_fsurv_famXline <- MCMCglmm(
  fish_surv75 ~ ib + fish_initial_bm_cen + fish_sex_cen,
  random = ~ tank_id + idh(line):worm_full_sib_fam_id,
  family = "threshold",
  data = dat_fsurv2_75dpi,
  prior = 
    list(R = list(V = 1, fix = 1),
         G = list(G1 = list(V = 1, nu = 1000, alpha.mu = 0, alpha.V = 1),
                  G2 = list(V = diag(4), nu = 0.002))),
  verbose=FALSE,
  pr = F,
  nitt = nit_fsurv, 
  thin = thin_fsurv,
  burnin = burnin_fsurv
  )
# anim model
m2_fsurv_anim <- MCMCglmm(fish_surv75 ~ ib + fish_initial_bm_cen + fish_sex_cen, 
                 random = ~tank_id + id,
                 ginverse = list(id = Ainv),
                 family = "threshold",
                 data = dat_fsurv2_75dpi,
                 prior = p2_bi,
                 verbose=F,
                 nitt = nit_fsurv, 
                 thin = thin_fsurv,
                 burnin = burnin_fsurv
                )

# full-sibs + mat effx
m3_fsurv_mat <- MCMCglmm(fish_surv75 ~ ib + fish_initial_bm_cen + fish_sex_cen, 
                 random = ~tank_id + worm_full_sib_fam_id + dam,
                 family = "threshold",
                 data = dat_fsurv2_75dpi%>%mutate(dam = factor(dam)),
                 prior = p3_bi,
                 verbose=F,
                 nitt = nit_fsurv, 
                 thin = thin_fsurv,
                 burnin = burnin_fsurv
                )
# anim model
m3_fsurv_anim <- MCMCglmm(fish_surv75 ~ ib + fish_initial_bm_cen + fish_sex_cen,
                 random = ~tank_id + id  + dam,
                 ginverse = list(id = Ainv),
                 family = "threshold",
                 data = dat_fsurv2_75dpi%>%mutate(dam = factor(dam)),
                 prior = p3_bi,
                 verbose=F,
                 nitt = nit_fsurv,
                 thin = thin_fsurv,
                 burnin = burnin_fsurv
                )
```

Adding full-sibs (red) and the pedigree (green) are slight improvements.

```{r}
plot(mcmc.list(
  m1_fsurv$Deviance, m2_fsurv_fam$Deviance, m2_fsurv_anim$Deviance
  ), density = F)
```

The full-sib model suggests that there is little variance within clutches.

```{r}
plot(mcmc.list(
  m2_fsurv_fam$Deviance, m3_fsurv_mat$Deviance
  ), density = F)
```

Here are the posterior distributions for full-sib and parental effects. Both border zero.

```{r}
plot(m3_fsurv_mat$VCV)
```

```{r}
plot(mcmc.list(
  m2_fsurv_anim$Deviance, m3_fsurv_anim$Deviance
  ), density = F)
```

For, these models, we break down total variance into several components: fixed effects (starting conditions), shared environment, genetic effects, and parental effects.

```{r}
m_vc_fsurv <- get_var_comps("fsurv", dat_fsurv2_75dpi, 'tank_id', 
                            m1_fsurv_noib, m1_fsurv,
                            m2_fsurv_fam, m2_fsurv_anim,
                            m3_fsurv_mat, m3_fsurv_anim)

m_vc_fsurv2 <- m_vc_fsurv%>%filter(v_name %in% c("Ve2", "Vg2", "Vm2", "Vr2")) # var with fixed effects
m_vc_fsurv <- m_vc_fsurv%>%filter(!v_name %in% c("Ve2", "Vg2", "Vm2", "Vr2")) # var excluding fixed effects
```

Here is the percent variation in infected fish survival accounted for by starting conditions (initial fish size and sex). It is high, as small fish were less likely to survive.

```{r}
filter(m_vc_fsurv, v_name == "Vf")%>%
  mutate(across(v_fit:v_upr, function(x){round(x,3)}))
```

Here is the percent variation in fish survival accounted for by shared environments (tanks).

```{r}
filter(m_vc_fsurv, v_name == "Ve")%>%
  mutate(across(v_fit:v_upr, function(x){round(x,3)}))
```

Here is the percent variation in fish survival accounted for by genetic variance (i.e. the heritability). It diverges between pedigree and full sib models, though the CIs are broad. The addition of maternal effects did not decrease heritability much in the animal model, but it does in the full-sib model.

```{r}
filter(m_vc_fsurv, v_name == "Vg")%>%
  mutate(across(v_fit:v_upr, function(x){round(x,3)}))
```

Family effects were minimal, but they were also skewed towards higher values, i.e. some families seemed more likely to kill their host.

```{r}
mean_fsurv <- sum(dat_fsurv2_75dpi$fish_surv75, na.rm = T)/sum(!is.na(dat_fsurv2_75dpi$fish_surv75), na.rm = T)
re_fams_fsurv <- dat_fsurv2_75dpi%>%
  filter(is.na(pred))%>%
  group_by(worm_full_sib_fam_id, gen)%>%
  summarize(fam_avg = -1*(qnorm(sum(fish_surv75, na.rm = T)/sum(!is.na(fish_surv75), na.rm = T)) - qnorm(mean_fsurv)))%>%
  ungroup()%>%
  mutate(approach="unadjusted")

re <- colMeans(m2_fsurv_fam$Sol[,which(grepl("worm_full_sib", colnames(m2_fsurv_fam$Sol)))])
re <- data.frame(worm_full_sib_fam_id = gsub("worm_full_sib_fam_id.", "", x = names(re)),
                 fam_avg = -1*re,
                 approach="model-estimated")

re_fams_fsurv <- bind_rows(
  re_fams_fsurv,
  re
)%>%
  mutate(approach = fct_relevel(factor(approach), c("unadjusted", "model-estimated")))

ggplot(re_fams_fsurv, aes(x = fam_avg)) +
  geom_histogram(aes(y=..density..),
                 colour = "black",
                 fill = "white") +
  geom_function(data = filter(re_fams_fsurv, approach == "unadjusted"),
                fun = dnorm, 
                args = list(mean = 0, 
                            sd = re_fams_fsurv%>%
                              filter(approach == "unadjusted", !is.infinite(fam_avg))%>%.$fam_avg%>%sd(., na.rm=T)
                            ),
                color = 'red',
                linetype = "dashed") +
  geom_function(data = filter(re_fams_fsurv, approach == "model-estimated"),
                fun = dnorm, 
                args = list(mean = 0, 
                            sd = re_fams_fsurv%>%filter(approach == "model-estimated")%>%.$fam_avg%>%sd(., na.rm=T)
                            ),
                color = 'red',
                linetype = "dashed") +
  geom_density() +
  geom_label(data = re_fams_fsurv%>%
               filter(!is.infinite(fam_avg))%>%
               group_by(approach)%>%
               summarise(min_x = min(fam_avg, na.rm = T),
                         skew = e1071::skewness(fam_avg, na.rm = T))%>%
               mutate(skew_c = paste0("Skewness: ", round(skew, 2))),
             aes(label = skew_c,
                 x = min_x,
                 y = 2),
             hjust = -0.05, vjust = 0) +
  facet_wrap(~approach, scales = "free_x") +
  labs(y = "Density", x = "Family averages (probit transformed)")
```

Here are the parental influences on fish infection (i.e. the effect of dam/sire).

```{r}
filter(m_vc_fsurv, v_name == "Vm")%>%
  mutate(across(v_fit:v_upr, function(x){round(x,3)}))
```

Here is the total variance explained by each model.

```{r}
filter(m_vc_fsurv, v_name == "Vr")%>%
  mutate(across(v_fit:v_upr, function(x){round(1-x,3)}))
```

The fraction of phenotypic variance explained by each term in the different models is plotted here. 

```{r}
ggplot(m_vc_fsurv%>%
         mutate(v_name = fct_relevel(factor(v_name), c("Vf", "Ve", "Vg", "Vm", "Vr"))),
       aes(v_name, v_fit, color = mod)) +
  geom_pointrange(aes(ymin = v_lwr, ymax = v_upr),
                  position = position_dodge(width = 0.6)) +
  labs(y = "Fraction variance explained", title = "Fish survival") +
  scale_x_discrete(labels = c("Starting\nconditions", 
                              "Shared\nenvironment",
                              "Additive\ngenetic", "Parental", "Residual")) +
  theme(panel.grid.major.x = element_blank(),
        axis.title.x = element_blank(),
        legend.title = element_blank())
```

The estimated heritabilities go up some after excluding starting conditions from the denominator of the variance calculations.

```{r}
ggplot(m_vc_fsurv2%>%
         mutate(v_name = fct_relevel(factor(v_name), c("Ve2", "Vg2", "Vm2", "Vr2"))),
       aes(v_name, v_fit, color = mod)) +
  geom_pointrange(aes(ymin = v_lwr, ymax = v_upr),
                  position = position_dodge(width = 0.6)) +
  labs(y = "Fraction variance explained", title = "Fish survival") +
  scale_x_discrete(labels = c("Shared\nenvironment", "Additive\ngenetic", "Parental", "Residual")) +
  theme(panel.grid.major.x = element_blank(),
        axis.title.x = element_blank(),
        legend.title = element_blank())
```

## Plerocercoid size

Here are the number of full-sib families and the number of offspring phenotyped per family.

```{r}
dat_pler <- dat_pler%>%mutate(log_ww = log(worm_bm_mg))
```

```{r}
dat_pler%>%
  group_by(worm_full_sib_fam_id)%>%
  summarize(n_pheno = sum(!is.na(log_ww)))%>%
  ungroup()%>%
  filter(!is.na(worm_full_sib_fam_id))%>%
  summarize(fs_fams = n(),
            min_phenotyped = min(n_pheno),
            lwrIQR_phenotyped = quantile(n_pheno, prob = 0.25),
            avg_phenotyped = mean(n_pheno),
            uprIQR_phenotyped = quantile(n_pheno, prob = 0.75),
            max_phenotyped = max(n_pheno)
            )
```

Here are the number of offspring phenotyped per dam/sire. There are not enough phenotyped per dam where I feel comfortable in estimating the 'dam' effect.

```{r}
dat_pler%>%
  group_by(maternal_match, worm_maternal_id)%>%
  summarize(n_pheno = sum(!is.na(log_ww)))%>%
  group_by(maternal_match)%>%
  filter(!is.na(worm_maternal_id))%>%
  summarize(dams = n(),
            min_phenotyped = min(n_pheno),
            lwrIQR_phenotyped = quantile(n_pheno, prob = 0.25),
            avg_phenotyped = mean(n_pheno),
            uprIQR_phenotyped = quantile(n_pheno, prob = 0.75),
            max_phenotyped = max(n_pheno)
            )
```

```{r}
parents <- dat_pler%>%
  group_by(eggs_clutch_id)%>%
  summarise(n_parent = n(), pler_p = mean(log_ww))%>%
  na.omit()

offspring <- dat_pler%>%
  group_by(worm_full_sib_fam_id)%>%
  summarise(n_offspring = n(), pler_o = mean(log_ww, na.rm = T))%>%
  na.omit()

parent_offspring <- left_join(parents, offspring, by = c("eggs_clutch_id" = "worm_full_sib_fam_id"))
```

For plerocercoid size, there is not much of a parent-offspring regression.

```{r}
ggplot(parent_offspring, aes(pler_p, pler_o)) + geom_point() +geom_smooth(se=F)
```

This may partly be due to some small worms that were measured from dead fish. Let's look at the regression when we exclude any fish that died.

```{r}
parents <- dat_pler%>%
  filter(fish_surv == 0)%>%
  group_by(eggs_clutch_id)%>%
  summarise(n_parent = n(), pler_p = mean(log_ww))%>%
  na.omit()

offspring <- dat_pler%>%
  filter(fish_surv == 0)%>%
  group_by(worm_full_sib_fam_id)%>%
  summarise(n_offspring = n(), pler_o = mean(log_ww, na.rm = T))%>%
  na.omit()

parent_offspring <- left_join(parents, offspring, by = c("eggs_clutch_id" = "worm_full_sib_fam_id"))
```

The relationship is still flat, probably because plerocercoid size is strongly affected by environmental factors like fish size. This is also log-transformed body size, which reduces variance among larger worms.

```{r}
ggplot(parent_offspring, aes(pler_p, pler_o)) + geom_point() +geom_smooth(se=F)
```

But the trend is still flat with untransformed plerocercoid size.

```{r}
parents <- dat_pler%>%
  filter(fish_surv == 0)%>%
  group_by(eggs_clutch_id)%>%
  summarise(n_parent = n(), pler_p = mean(worm_bm_mg))%>%
  na.omit()

offspring <- dat_pler%>%
  filter(fish_surv == 0)%>%
  group_by(worm_full_sib_fam_id)%>%
  summarise(n_offspring = n(), pler_o = mean(worm_bm_mg, na.rm = T))%>%
  na.omit()

parent_offspring <- left_join(parents, offspring, by = c("eggs_clutch_id" = "worm_full_sib_fam_id"))

ggplot(parent_offspring, aes(pler_p, pler_o)) + geom_point() +geom_smooth(se=F)
```

The p-o regression does not account for environmental influences like fish size or tank. Thus, let's fit the same 5 models again, acknowledging that the models with 'dam' effects may be overfitted.

```{r}
p1_g <- list(R = list(V = 1, nu = 0.002),
             G = list(G1 = list(V = 1, nu = 0.002)))
p2_g <- list(R = list(V = 1, nu = 0.002),
             G = list(G1 = list(V = 1, nu = 0.002),
                      G2 = list(V = 1, nu = 0.002)))
p3_g <- list(R = list(V = 1, nu = 0.002),
             G = list(G1 = list(V = 1, nu = 0.002),
                      G2 = list(V = 1, nu = 0.002),
                      G3 = list(V = 1, nu = 0.002)))

thin_pler <- 100
burnin_pler <- 1000
nit_pler <- thin_pler * target_samples + burnin_pler

# fixed effects
m1_pler_noib <- MCMCglmm(log_ww ~ fish_sex_cen + fish_initial_bm_cen + age_diss_cen, 
                 random = ~tank_id,
                 family = "gaussian",
                 data = dat_pler,
                 prior = p1_g,
                 verbose=FALSE,
                 nitt = nit_pler, 
                 thin = thin_pler,
                 burnin = burnin_pler
                )
# fixed effects
m1_pler <- MCMCglmm(log_ww ~ ib + fish_sex_cen + fish_initial_bm_cen + age_diss_cen, 
                 random = ~tank_id,
                 family = "gaussian",
                 data = dat_pler,
                 prior = p1_g,
                 verbose=FALSE,
                 nitt = nit_pler, 
                 thin = thin_pler,
                 burnin = burnin_pler
                )
# fam
m2_pler_fam <- MCMCglmm(log_ww ~ ib + fish_sex_cen + fish_initial_bm_cen + age_diss_cen, 
                 random = ~tank_id + worm_full_sib_fam_id,
                 family = "gaussian",
                 data = dat_pler,
                 prior = p2_g,
                 verbose=FALSE,
                 pr = T,
                 nitt = nit_pler, 
                 thin = thin_pler,
                 burnin = burnin_pler
                )
# full sib x line
m2_pler_famXline <- MCMCglmm(
  log_ww ~ ib + fish_sex_cen + fish_initial_bm_cen + age_diss_cen, 
  random = ~tank_id + idh(line):worm_full_sib_fam_id,
  family = "gaussian",
  data = dat_pler,
  prior = list(R = list(V = 1, nu = 0.002),
               G = list(G1 = list(V = 1, nu = 0.002),
                        G2 = list(V = diag(4), nu = 0.002))),
  verbose=FALSE,
  pr = F,
  nitt = nit_pler, 
  thin = thin_pler,
  burnin = burnin_pler
  )
# anim model
m2_pler_anim <- MCMCglmm(log_ww ~ ib + fish_sex_cen + fish_initial_bm_cen + age_diss_cen, 
                 random = ~tank_id + id,
                 ginverse = list(id = Ainv),
                 family = "gaussian",
                 data = dat_pler,
                 prior = p2_g,
                 verbose=F,
                 nitt = nit_pler, 
                 thin = thin_pler,
                 burnin = burnin_pler
                )
# fam
m3_pler_mat <- MCMCglmm(log_ww ~ ib + fish_sex_cen + fish_initial_bm_cen + age_diss_cen, 
                 random = ~tank_id + worm_full_sib_fam_id + dam,
                 family = "gaussian",
                 data = dat_pler%>%mutate(dam = factor(dam)),
                 prior = p3_g,
                 verbose=FALSE,
                 nitt = nit_pler, 
                 thin = thin_pler,
                 burnin = burnin_pler
                )
# anim model
m3_pler_anim <- MCMCglmm(log_ww ~ ib + fish_sex_cen + fish_initial_bm_cen + age_diss_cen, 
                 random = ~tank_id + id + dam,
                 ginverse = list(id = Ainv),
                 family = "gaussian",
                 data = dat_pler%>%mutate(dam = factor(dam)),
                 prior = p3_g,
                 verbose=F,
                 nitt = nit_pler, 
                 thin = thin_pler,
                 burnin = burnin_pler
                )
```

Adding full-sibs (red) and then the pedigree (green) appear to be improvements.

```{r}
plot(mcmc.list(
  m1_pler$Deviance, m2_pler_fam$Deviance, m2_pler_anim$Deviance
  ), density = F)
```

The full-sib model suggests that little variance within clutches is explained by parental effects.

```{r}
plot(mcmc.list(
  m2_pler_fam$Deviance, m3_pler_mat$Deviance
  ), density = F)
```

By contrast, the addition of maternal effects seems to improve the animal model.

```{r}
plot(mcmc.list(
  m2_pler_anim$Deviance, m3_pler_anim$Deviance
  ), density = F)
```

For, these models, we break down total variance into several components: fixed effects (starting conditions), shared environment, genetic effects, and parental effects.

```{r}
m_vc_pler <- get_var_comps("pler", dat_pler, 'tank_id', 
                            m1_pler_noib, m1_pler,
                            m2_pler_fam, m2_pler_anim,
                            m3_pler_mat, m3_pler_anim)

m_vc_pler2 <- m_vc_pler%>%filter(v_name %in% c("Ve2", "Vg2", "Vm2", "Vr2")) # var with fixed effects
m_vc_pler <- m_vc_pler%>%filter(!v_name %in% c("Ve2", "Vg2", "Vm2", "Vr2")) # var excluding fixed effects
```

Here is the percent variation in log plerocercoid size accounted for by fish size, sex, and age at dissection across models. It is very high.

```{r}
filter(m_vc_pler, v_name == "Vf")%>%
  mutate(across(v_fit:v_upr, function(x){round(x,3)}))
```

Here is the percent variation in plerercoid size accounted for by shared environments (tanks). Worms from fish in the same tank were not much more similar in mass than expected.

```{r}
filter(m_vc_pler, v_name == "Ve")%>%
  mutate(across(v_fit:v_upr, function(x){round(x,3)}))
```

Here is the percent variation in plerercoid size accounted for by genetic variance (i.e. the heritability). It is low, but this is mainly because the fixed effect variance was high.

```{r}
filter(m_vc_pler, v_name == "Vg")%>%
  mutate(across(v_fit:v_upr, function(x){round(x,3)}))
```

The heritability appears much higher if we only consider the phenotypic variance left after accounting for fixed effects:

```{r}
filter(m_vc_pler2, v_name == "Vg2")%>%
  mutate(across(v_fit:v_upr, function(x){round(x,3)}))
```

Family effects were not skewed.

```{r}
mean_pler <- mean(dat_pler$log_ww, na.rm = T)
re_fams_pler <- dat_pler%>%
  filter(is.na(pred))%>%
  group_by(worm_full_sib_fam_id, gen)%>%
  summarize(fam_avg = mean(log_ww, na.rm = T) - mean_pler)%>%
  ungroup()%>%
  mutate(approach="unadjusted")

re <- colMeans(m2_pler_fam$Sol[,which(grepl("worm_full_sib", colnames(m2_pler_fam$Sol)))])
re <- data.frame(worm_full_sib_fam_id = gsub("worm_full_sib_fam_id.", "", x = names(re)),
                 fam_avg = re,
                 approach="model-estimated")

re_fams_pler <- bind_rows(
  re_fams_pler,
  re
)%>%
  mutate(approach = fct_relevel(factor(approach), c("unadjusted", "model-estimated")))

ggplot(re_fams_pler, aes(x = fam_avg)) +
  geom_histogram(aes(y=..density..),
                 colour = "black",
                 fill = "white") +
  geom_function(data = filter(re_fams_pler, approach == "unadjusted"),
                fun = dnorm, 
                args = list(mean = 0, 
                            sd = re_fams_pler%>%filter(approach == "unadjusted")%>%.$fam_avg%>%sd(., na.rm=T)
                            ),
                color = 'red',
                linetype = "dashed") +
  geom_function(data = filter(re_fams_pler, approach == "model-estimated"),
                fun = dnorm, 
                args = list(mean = 0, 
                            sd = re_fams_pler%>%filter(approach == "model-estimated")%>%.$fam_avg%>%sd(., na.rm=T)
                            ),
                color = 'red',
                linetype = "dashed") +
  geom_density() +
  geom_label(data = re_fams_pler%>%
               filter(!is.infinite(fam_avg))%>%
               group_by(approach)%>%
               summarise(min_x = min(fam_avg, na.rm = T),
                         skew = e1071::skewness(fam_avg, na.rm = T))%>%
               mutate(skew_c = paste0("Skewness: ", round(skew, 2))),
             aes(label = skew_c,
                 x = min_x,
                 y = 1),
             hjust = -0.05, vjust = 0) +
  facet_wrap(~approach, scales = "free_x") +
  labs(y = "Density", x = "Family averages")
```

The influence of 'parent' on plerercoid size are small. On the one hand, this is expected, as parental effects are more likely early in life. On the other hand, the data have limited data to detect parental effects, given that few plerocercoids were obtained per mother.

```{r}
filter(m_vc_pler, v_name == "Vm")%>%
  mutate(across(v_fit:v_upr, function(x){round(x,3)}))
```

Here is the total variance explained by each model.

```{r}
filter(m_vc_pler, v_name == "Vr")%>%
  mutate(across(v_fit:v_upr, function(x){round(1-x,3)}))
```

The fraction of phenotypic variance explained by each term in the different models is plotted here.

```{r}
ggplot(m_vc_pler%>%
         mutate(v_name = fct_relevel(factor(v_name), c("Vf", "Ve", "Vg", "Vm", "Vr"))),
       aes(v_name, v_fit, color = mod)) +
  geom_pointrange(aes(ymin = v_lwr, ymax = v_upr),
                  position = position_dodge(width = 0.6)) +
  labs(y = "Fraction variance explained", title = "Plerocercoid size") +
  scale_x_discrete(labels = c("Starting\nconditions", "Shared\nenvironment", "Additive\ngenetic", "Parental", "Residual")) +
  theme(panel.grid.major.x = element_blank(),
        axis.title.x = element_blank(),
        legend.title = element_blank())
```

Since starting condition have such a big influence on plerocercoid size, we also examine heritability after excluding this. It is much higher.

```{r}
ggplot(m_vc_pler2%>%
         mutate(v_name = fct_relevel(factor(v_name), c("Ve2", "Vg2", "Vm2", "Vr2"))),
       aes(v_name, v_fit, color = mod)) +
  geom_pointrange(aes(ymin = v_lwr, ymax = v_upr),
                  position = position_dodge(width = 0.6)) +
  labs(y = "Fraction variance explained", title = "Plerocercoid size") +
  scale_x_discrete(labels = c("Shared\nenvironment", "Additive\ngenetic", "Parental", "Residual")) +
  theme(panel.grid.major.x = element_blank(),
        axis.title.x = element_blank(),
        legend.title = element_blank())
```

### Excluding dead fish

We can also estimate the heritability excluding the fish that died during the experiment. Excluding the dead fish (and small worms) also removes the necessity to transform worm mass, since we mostly remove the exponential age x size relationship. 

```{r}
dat_pler2 <- dat_pler%>%
  filter(fish_surv == 0)
```

```{r}
# fixed effects
m1_pler2 <- MCMCglmm(worm_bm_mg ~ ib + fish_sex_cen + fish_initial_bm_cen + age_diss_cen, 
                 random = ~tank_id,
                 family = "gaussian",
                 data = dat_pler2,
                 prior = p1_g,
                 verbose=FALSE,
                 nitt = nit_pler, 
                 thin = thin_pler,
                 burnin = burnin_pler
                )
# fam
m2_pler_fam2 <- MCMCglmm(worm_bm_mg ~ ib + fish_sex_cen + fish_initial_bm_cen + age_diss_cen, 
                 random = ~tank_id + worm_full_sib_fam_id,
                 family = "gaussian",
                 data = dat_pler2,
                 prior = p2_g,
                 pr = T,
                 verbose=FALSE,
                 nitt = nit_pler, 
                 thin = thin_pler,
                 burnin = burnin_pler
                )
# anim model
m2_pler_anim2 <- MCMCglmm(worm_bm_mg ~ ib + fish_sex_cen + fish_initial_bm_cen + age_diss_cen, 
                 random = ~tank_id + id,
                 ginverse = list(id = Ainv),
                 family = "gaussian",
                 data = dat_pler2,
                 prior = p2_g,
                 verbose=F,
                 nitt = nit_pler, 
                 thin = thin_pler,
                 burnin = burnin_pler
                )
```

Adding full-sibs (red) is an improvemnt but adding the pedigree (green) is not clearly a further improvement.

```{r}
plot(mcmc.list(
  m1_pler2$Deviance, m2_pler_fam2$Deviance, m2_pler_anim2$Deviance
  ), density = F)
```

For, these models, we break down total variance into several components: fixed effects (starting conditions), shared environment, genetic effects, and parental effects.

```{r}
# first model
pdx <- m1_pler2$X # model matrix, fixed effx
n <- dim(pdx)[1] # number of data points in model, all traits
nt <- n/1 # number of data points per trait
p_i <- which(is.na(dat_pler2$pred)) # points where we can predicted vals and cred int
pdx <- pdx[c(p_i),] # restrict to only points where we want preds
p_n <- dim(pdx)[1]/1 # number of points for each trait we want to predict
num_fe <- m1_pler2$Fixed$nfl # number of fixed effx, rand effx marginalized
p_all <- as.matrix(pdx) %*% t(m1_pler2$Sol[,1:num_fe]) # predicteds via matrix mult 
m1_Vf <- apply(p_all, MARGIN = 2, var)

pdx <- m2_pler_fam2$X # model matrix, fixed effx
n <- dim(pdx)[1] # number of data points in model, all traits
nt <- n/1 # number of data points per trait
p_i <- which(is.na(dat_pler2$pred)) # points where we can predicted vals and cred int
pdx <- pdx[c(p_i),] # restrict to only points where we want preds
p_n <- dim(pdx)[1]/1 # number of points for each trait we want to predict
num_fe <- m2_pler_fam2$Fixed$nfl # number of fixed effx, rand effx marginalized
p_all <- as.matrix(pdx) %*% t(m2_pler_fam2$Sol[,1:num_fe]) # predicteds via matrix mult 
m2_Vf_fam <- apply(p_all, MARGIN = 2, var)

pdx <- m2_pler_anim2$X # model matrix, fixed effx
n <- dim(pdx)[1] # number of data points in model, all traits
nt <- n/1 # number of data points per trait
p_i <- which(is.na(dat_pler2$pred)) # points where we can predicted vals and cred int
pdx <- pdx[c(p_i),] # restrict to only points where we want preds
p_n <- dim(pdx)[1]/1 # number of points for each trait we want to predict
num_fe <- m2_pler_anim2$Fixed$nfl # number of fixed effx, rand effx marginalized
p_all <- as.matrix(pdx) %*% t(m2_pler_anim2$Sol[,1:num_fe]) # predicteds via matrix mult 
m2_Vf_anim <- apply(p_all, MARGIN = 2, var)
```
```{r}
# PHENOTYPIC VAR
m1_Vp <- m1_Vf + rowSums(m1_pler2$VCV)
m2_Vp_fam <- m2_Vf_fam + rowSums(m2_pler_fam2$VCV)
m2_Vp_anim <- m2_Vf_anim + rowSums(m2_pler_anim2$VCV)
```
```{r}
# vc for model 1
m1_vc <- data.frame(mod = "1b_base", 
                    v_est = c(
                      m1_Vf/m1_Vp, 
                      m1_pler2$VCV[,'tank_id']/m1_Vp,
                      m1_pler2$VCV[,'tank_id']/(m1_Vp - m1_Vf),
                      1-(m1_Vf+m1_pler2$VCV[,'tank_id'])/m1_Vp,
                      1-(m1_pler2$VCV[,'tank_id'])/(m1_Vp - m1_Vf)
                      ),
                    v_name = rep(c("Vf", "Ve", "Ve2", "Vr", "Vr2"), 
                                 times = rep(length(m1_Vf), 5))
                    )

# vc for model 2 full sib
m2_vc <- data.frame(mod = "2a_full_sibs", 
                    v_est = c(
                      m2_Vf_fam/m2_Vp_fam, 
                      m2_pler_fam2$VCV[,'tank_id']/m2_Vp_fam, 
                      m2_pler_fam2$VCV[,'tank_id']/(m2_Vp_fam - m2_Vf_fam), 
                      2*m2_pler_fam2$VCV[,'worm_full_sib_fam_id']/m2_Vp_fam, 
                      2*m2_pler_fam2$VCV[,'worm_full_sib_fam_id']/(m2_Vp_fam - m2_Vf_fam),
                      1-(m2_Vf_fam+m2_pler_fam2$VCV[,'tank_id'] +
                           2*m2_pler_fam2$VCV[,'worm_full_sib_fam_id'])/m2_Vp_fam,
                      1-(m2_pler_fam2$VCV[,'tank_id'] +
                           2*m2_pler_fam2$VCV[,'worm_full_sib_fam_id'])/(m2_Vp_fam - m2_Vf_fam)
                      ),
                    v_name = rep(c("Vf", "Ve", "Ve2", "Vg", "Vg2", "Vr", "Vr2"), 
                            times = rep(length(m2_Vf_fam), 7))
                    )

# vc for model 2 anim
m2_vc_anim <- data.frame(mod = "2b_animal", 
                    v_est = c(
                      m2_Vf_anim/m2_Vp_anim, 
                      m2_pler_anim2$VCV[,'tank_id']/m2_Vp_anim, 
                      m2_pler_anim2$VCV[,'tank_id']/(m2_Vp_anim - m2_Vf_anim), 
                      m2_pler_anim2$VCV[,'id']/m2_Vp_anim, 
                      m2_pler_anim2$VCV[,'id']/(m2_Vp_anim - m2_Vf_anim),
                      1-(m2_Vf_anim+m2_pler_anim2$VCV[,'tank_id'] +
                           m2_pler_anim2$VCV[,'id'])/m2_Vp_anim,
                      1-(m2_pler_anim2$VCV[,'tank_id'] +
                           m2_pler_anim2$VCV[,'id'])/(m2_Vp_anim - m2_Vf_anim)
                      ),
                    v_name = rep(c("Vf", "Ve", "Ve2", "Vg", "Vg2", "Vr", "Vr2"), 
                            times = rep(length(m2_Vf_anim), 7))
                    )


m_vc <- bind_rows(
  m1_vc,
  m2_vc, m2_vc_anim)


m_vc_pler_d <- m_vc%>%
  group_by(mod, v_name)%>%
  summarise(v_fit = median(v_est),
            v_lwr = quantile(v_est, probs = 0.025),
            v_upr = quantile(v_est, probs = 0.975),
            )%>%
  ungroup()%>%
  mutate(trait = "pler")

m_vc_pler_d2 <- m_vc_pler_d%>%filter(v_name %in% c("Ve2", "Vg2", "Vm2", "Vr2")) # var with fixed effects
m_vc_pler_d <- m_vc_pler_d%>%filter(!v_name %in% c("Ve2", "Vg2", "Vm2", "Vr2")) # var excluding fixed effects
```

Here is the percent variation in untransformed plerocercoid size accounted for by fish size, sex, and age at dissection across models. It is still high, but much lower than when we included the small worms from dead fish.

```{r}
filter(m_vc_pler_d, v_name == "Vf")%>%
  mutate(across(v_fit:v_upr, function(x){round(x,3)}))
```

Here is the percent variation in plerercoid size accounted for by shared environments (tanks). There is a tank effect.

```{r}
filter(m_vc_pler_d, v_name == "Ve")%>%
  mutate(across(v_fit:v_upr, function(x){round(x,3)}))
```

Here is the percent variation in plerercoid size accounted for by genetic variance (i.e. the heritability). It is significant.

```{r}
filter(m_vc_pler_d, v_name == "Vg")%>%
  mutate(across(v_fit:v_upr, function(x){round(x,3)}))
```

It is quite a bit higher after excluding the fixed effects.

```{r}
filter(m_vc_pler_d2, v_name == "Vg2")%>%
  mutate(across(v_fit:v_upr, function(x){round(x,3)}))
```

The distribution of the family effects still looks good after excluding the dead fish.

```{r}
mean_pler <- mean(dat_pler2$worm_bm_mg, na.rm = T)
re_fams_pler2 <- dat_pler2%>%
  filter(is.na(pred))%>%
  group_by(worm_full_sib_fam_id, gen)%>%
  summarize(fam_avg = mean(worm_bm_mg, na.rm = T) - mean_pler)%>%
  ungroup()%>%
  mutate(approach="unadjusted")

re <- colMeans(m2_pler_fam2$Sol[,which(grepl("worm_full_sib", colnames(m2_pler_fam2$Sol)))])
re <- data.frame(worm_full_sib_fam_id = gsub("worm_full_sib_fam_id.", "", x = names(re)),
                 fam_avg = re,
                 approach="model-estimated")

re_fams_pler2 <- bind_rows(
  re_fams_pler2,
  re
)%>%
  mutate(approach = fct_relevel(factor(approach), c("unadjusted", "model-estimated")))

ggplot(re_fams_pler2, aes(x = fam_avg)) +
  geom_histogram(aes(y=..density..),
                 colour = "black",
                 fill = "white") +
  geom_function(data = filter(re_fams_pler2, approach == "unadjusted"),
                fun = dnorm, 
                args = list(mean = 0, 
                            sd = re_fams_pler2%>%filter(approach == "unadjusted")%>%.$fam_avg%>%sd(., na.rm=T)
                            ),
                color = 'red',
                linetype = "dashed") +
  geom_function(data = filter(re_fams_pler2, approach == "model-estimated"),
                fun = dnorm, 
                args = list(mean = 0, 
                            sd = re_fams_pler2%>%filter(approach == "model-estimated")%>%.$fam_avg%>%sd(., na.rm=T)
                            ),
                color = 'red',
                linetype = "dashed") +
  geom_density() +
  geom_label(data = re_fams_pler2%>%
               filter(!is.infinite(fam_avg))%>%
               group_by(approach)%>%
               summarise(min_x = min(fam_avg, na.rm = T),
                         skew = e1071::skewness(fam_avg, na.rm = T))%>%
               mutate(skew_c = paste0("Skewness: ", round(skew, 2))),
             aes(label = skew_c,
                 x = min_x,
                 y = 0.025),
             hjust = -0.05, vjust = 0) +
  facet_wrap(~approach, scales = "free_x") +
  labs(y = "Density", x = "Family averages")
```

Here is the total variance explained by each model.

```{r}
filter(m_vc_pler2, v_name == "Vr")%>%
  mutate(across(v_fit:v_upr, function(x){round(1-x,3)}))
```

The fraction of phenotypic variance explained by each term in the different models is plotted here

```{r}
ggplot(m_vc_pler_d%>%
         mutate(v_name = fct_relevel(factor(v_name), c("Vf", "Ve", "Vg", "Vm", "Vr"))),
       aes(v_name, v_fit, color = mod)) +
  geom_pointrange(aes(ymin = v_lwr, ymax = v_upr),
                  position = position_dodge(width = 0.6)) +
  labs(y = "Fraction variance explained", title = "Plerocercoid size") +
  scale_x_discrete(labels = c("Starting\nconditions", "Shared\nenvironment", "Additive\ngenetic", "Residual")) +
  theme(panel.grid.major.x = element_blank(),
        axis.title.x = element_blank(),
        legend.title = element_blank())
```
```{r}
# ggplot(m_vc_pler_d2%>%
#          mutate(v_name = fct_relevel(factor(v_name), c("Ve2", "Vg2", "Vm2", "Vr2"))),
#        aes(v_name, v_fit, color = mod)) +
#   geom_pointrange(aes(ymin = v_lwr, ymax = v_upr),
#                   position = position_dodge(width = 0.6)) +
#   labs(y = "Fraction variance explained", title = "Plerocercoid size") +
#   scale_x_discrete(labels = c("Shared\nenvironment", "Additive\ngenetic", "Parental", "Residual")) +
#   theme(panel.grid.major.x = element_blank(),
#         axis.title.x = element_blank(),
#         legend.title = element_blank())
```

## Liver weight

Liver weight was measured in dissected fish, so all fish that were scored for infection also had a liver weight measurement. Having a large liver, relative to body mass, is considered a sign of good energetic condition.

Here are the number of full-sib families and the number of offspring phenotyped per family.

```{r}
dat_liver2 <- dat_liver2%>%mutate(log_liver = log(fish_liver_mg))
```

```{r}
dat_liver2%>%
  group_by(worm_full_sib_fam_id)%>%
  summarize(n_pheno = sum(!is.na(fish_liver_mg)))%>%
  ungroup()%>%
  filter(!is.na(worm_full_sib_fam_id))%>%
  summarize(fs_fams = n(),
            min_phenotyped = min(n_pheno),
            lwrIQR_phenotyped = quantile(n_pheno, prob = 0.25),
            avg_phenotyped = mean(n_pheno),
            uprIQR_phenotyped = quantile(n_pheno, prob = 0.75),
            max_phenotyped = max(n_pheno)
            )
```

Here are the number of offspring phenotyped per dam/sire. There are not enough phenotyped per dam where I feel comfortable in estimating the 'dam' effect.

```{r}
dat_liver2%>%
  group_by(maternal_match, worm_maternal_id)%>%
  summarize(n_pheno = sum(!is.na(fish_liver_mg)))%>%
  group_by(maternal_match)%>%
  filter(!is.na(worm_maternal_id))%>%
  summarize(dams = n(),
            min_phenotyped = min(n_pheno),
            lwrIQR_phenotyped = quantile(n_pheno, prob = 0.25),
            avg_phenotyped = mean(n_pheno),
            uprIQR_phenotyped = quantile(n_pheno, prob = 0.75),
            max_phenotyped = max(n_pheno)
            )
```

```{r}
parents <- dat_liver2%>%
  group_by(eggs_clutch_id)%>%
  summarise(n_parent = n(), liver_p = mean(log_liver))%>%
  na.omit()

offspring <- dat_liver2%>%
  group_by(worm_full_sib_fam_id)%>%
  summarise(n_offspring = n(), liver_o = mean(log_liver, na.rm = T))%>%
  na.omit()

parent_offspring <- left_join(parents, offspring, by = c("eggs_clutch_id" = "worm_full_sib_fam_id"))
```

For liver size, there is not much of a parent-offspring regression, which is not surprising, since this does not account for fish size.

```{r}
ggplot(parent_offspring, aes(liver_p, liver_o)) + geom_point() +geom_smooth(se=F)
```

So, let's fit the same 5 models again, acknowledging that the models with 'dam' effects may be overfitted.

```{r}
thin_liver <- 100
burnin_liver <- 1000
nit_liver <- thin_liver * target_samples + burnin_liver

# fixed effects, no ib
m1_liver_noib <- MCMCglmm(log_liver ~ age_diss_cen + fish_sex_cen + log_fish_initial_bm_cen, 
                 random = ~tank_id,
                 family = "gaussian",
                 data = dat_liver2,
                 prior = p1_g,
                 verbose=FALSE,
                 nitt = nit_liver, 
                 thin = thin_liver,
                 burnin = burnin_liver
                )
# fixed effects
m1_liver <- MCMCglmm(log_liver ~ ib + age_diss_cen + fish_sex_cen + log_fish_initial_bm_cen, 
                 random = ~tank_id,
                 family = "gaussian",
                 data = dat_liver2,
                 prior = p1_g,
                 verbose=FALSE,
                 nitt = nit_liver, 
                 thin = thin_liver,
                 burnin = burnin_liver
                )
# fam
m2_liver_fam <- MCMCglmm(log_liver ~ ib + age_diss_cen + fish_sex_cen + log_fish_initial_bm_cen, 
                 random = ~tank_id + worm_full_sib_fam_id,
                 family = "gaussian",
                 data = dat_liver2,
                 prior = p2_g,
                 pr = T,
                 verbose=FALSE,
                 nitt = nit_liver, 
                 thin = thin_liver,
                 burnin = burnin_liver
                )
# full sib x line
m2_liver_famXline <- MCMCglmm(
  log_liver ~ ib + age_diss_cen + fish_sex_cen + log_fish_initial_bm_cen, 
  random = ~tank_id + idh(line):worm_full_sib_fam_id,
  family = "gaussian",
  data = dat_liver2,
  prior = list(R = list(V = 1, nu = 0.002),
               G = list(G1 = list(V = 1, nu = 0.002),
                        G2 = list(V = diag(4), nu = 0.002))),
  verbose=FALSE,
  pr = F,
  nitt = nit_liver, 
  thin = thin_liver,
  burnin = burnin_liver
  )
# anim model
m2_liver_anim <- MCMCglmm(log_liver ~ ib + age_diss_cen + fish_sex_cen + log_fish_initial_bm_cen, 
                 random = ~tank_id + id,
                 ginverse = list(id = Ainv),
                 family = "gaussian",
                 data = dat_liver2,
                 prior = p2_g,
                 verbose=F,
                 nitt = nit_liver, 
                 thin = thin_liver,
                 burnin = burnin_liver
                )
# fam
m3_liver_mat <- MCMCglmm(log_liver ~ ib + age_diss_cen + fish_sex_cen + log_fish_initial_bm_cen, 
                 random = ~tank_id + worm_full_sib_fam_id + dam,
                 family = "gaussian",
                 data = dat_liver2%>%mutate(dam = factor(dam)),
                 prior = p3_g,
                 verbose=FALSE,
                 nitt = nit_liver, 
                 thin = thin_liver,
                 burnin = burnin_liver
                )
# anim model
m3_liver_anim <- MCMCglmm(log_liver ~ ib + age_diss_cen + fish_sex_cen + log_fish_initial_bm_cen, 
                 random = ~tank_id + id + dam,
                 ginverse = list(id = Ainv),
                 family = "gaussian",
                 data = dat_liver2%>%mutate(dam = factor(dam)),
                 prior = p3_g,
                 verbose=F,
                 nitt = nit_liver, 
                 thin = thin_liver,
                 burnin = burnin_liver
                )
```

Adding full-sibs (red) and then the pedigree (green) appear to be only slight improvements.

```{r}
plot(mcmc.list(
  m1_liver$Deviance, m2_liver_fam$Deviance, m2_liver_anim$Deviance
  ), density = F)
```

The full-sib model suggests that little variance within clutches is explained by parental effects.

```{r}
plot(mcmc.list(
  m2_liver_fam$Deviance, m3_liver_mat$Deviance
  ), density = F)
```

Same for the animal model.

```{r}
plot(mcmc.list(
  m2_liver_anim$Deviance, m3_liver_anim$Deviance
  ), density = F)
```

For, these models, we break down total variance into several components: fixed effects (starting conditions), shared environment, genetic effects, and parental effects.

```{r}
m_vc_liver <- get_var_comps("liver", dat_liver2, 'tank_id', 
                            m1_liver_noib, m1_liver,
                            m2_liver_fam, m2_liver_anim,
                            m3_liver_mat, m3_liver_anim)

m_vc_liver2 <- m_vc_liver%>%filter(v_name %in% c("Ve2", "Vg2", "Vm2", "Vr2")) # var with fixed effects
m_vc_liver <- m_vc_liver%>%filter(!v_name %in% c("Ve2", "Vg2", "Vm2", "Vr2")) # var excluding fixed effects
```

Here is the percent variation in log liver mass accounted for by fish size, sex, and age at dissection across models. It is high, which is not surprising (big fish have big livers).

```{r}
filter(m_vc_liver, v_name == "Vf")%>%
  mutate(across(v_fit:v_upr, function(x){round(x,3)}))
```

Here is the percent variation in liver size accounted for by shared environments (tanks). Infected fish from the same tank tended to have similar liver weights.

```{r}
filter(m_vc_liver, v_name == "Ve")%>%
  mutate(across(v_fit:v_upr, function(x){round(x,3)}))
```

Here is the percent variation in liver size accounted for by genetic variance (i.e. the heritability). It is low.

```{r}
filter(m_vc_liver, v_name == "Vg")%>%
  mutate(across(v_fit:v_upr, function(x){round(x,3)}))
```

The heritability is not much higher if we only consider the phenotypic variance left after accounting for fixed effects (here for the final animal model):

```{r}
filter(m_vc_liver2, v_name == "Vg2")%>%
  mutate(across(v_fit:v_upr, function(x){round(x,3)}))
```

Family effects were minimal (~5% in either direction) and not clearly skewed.

```{r}
mean_liver <- mean(dat_liver2$log_liver, na.rm = T)
re_fams_liver <- dat_liver2%>%
  filter(is.na(pred))%>%
  group_by(worm_full_sib_fam_id, gen)%>%
  summarize(fam_avg = mean(log_liver, na.rm = T) - mean_liver)%>%
  ungroup()%>%
  mutate(approach="unadjusted")

re <- colMeans(m2_liver_fam$Sol[,which(grepl("worm_full_sib", colnames(m2_liver_fam$Sol)))])
re <- data.frame(worm_full_sib_fam_id = gsub("worm_full_sib_fam_id.", "", x = names(re)),
                 fam_avg = re,
                 approach="model-estimated")

re_fams_liver <- bind_rows(
  re_fams_liver,
  re
)%>%
  mutate(approach = fct_relevel(factor(approach), c("unadjusted", "model-estimated")))

ggplot(re_fams_liver, aes(x = fam_avg)) +
  geom_histogram(aes(y=..density..),
                 colour = "black",
                 fill = "white") +
  geom_function(data = filter(re_fams_liver, approach == "unadjusted"),
                fun = dnorm, 
                args = list(mean = 0, 
                            sd = re_fams_liver%>%filter(approach == "unadjusted")%>%.$fam_avg%>%sd(., na.rm=T)
                            ),
                color = 'red',
                linetype = "dashed") +
  geom_function(data = filter(re_fams_liver, approach == "model-estimated"),
                fun = dnorm, 
                args = list(mean = 0, 
                            sd = re_fams_liver%>%filter(approach == "model-estimated")%>%.$fam_avg%>%sd(., na.rm=T)
                            ),
                color = 'red',
                linetype = "dashed") +
  geom_density() +
  geom_label(data = re_fams_liver%>%
               filter(!is.infinite(fam_avg))%>%
               group_by(approach)%>%
               summarise(min_x = min(fam_avg, na.rm = T),
                         skew = e1071::skewness(fam_avg, na.rm = T))%>%
               mutate(skew_c = paste0("Skewness: ", round(skew, 2))),
             aes(label = skew_c,
                 x = min_x,
                 y = 1),
             hjust = -0.05, vjust = 0) +
  facet_wrap(~approach, scales = "free") +
  labs(y = "Density", x = "Family averages")
```

The influence of 'parent' on liver size is small. But there was limited data to test such parental effects.

```{r}
filter(m_vc_liver, v_name == "Vm")%>%
  mutate(across(v_fit:v_upr, function(x){round(x,3)}))
```

Here is the total variance explained by each model.

```{r}
filter(m_vc_liver, v_name == "Vr")%>%
  mutate(across(v_fit:v_upr, function(x){round(1-x,3)}))
```


The fraction of phenotypic variance explained by each term in the different models is plotted here

```{r}
ggplot(m_vc_liver%>%
         mutate(v_name = fct_relevel(factor(v_name), c("Vf", "Ve", "Vg", "Vm", "Vr"))),
       aes(v_name, v_fit, color = mod)) +
  geom_pointrange(aes(ymin = v_lwr, ymax = v_upr),
                  position = position_dodge(width = 0.6)) +
  labs(y = "Fraction variance explained", title = "Liver weight") +
  scale_x_discrete(labels = c("Starting\nconditions", "Shared\nenvironment", "Additive\ngenetic", "Parental", "Residual")) +
  theme(panel.grid.major.x = element_blank(),
        axis.title.x = element_blank(),
        legend.title = element_blank())
```

Removing the starting conditions did not have much effect on the heritability estimate.

```{r}
ggplot(m_vc_liver2%>%
         mutate(v_name = fct_relevel(factor(v_name), c("Ve2", "Vg2", "Vm2", "Vr2"))),
       aes(v_name, v_fit, color = mod)) +
  geom_pointrange(aes(ymin = v_lwr, ymax = v_upr),
                  position = position_dodge(width = 0.6)) +
  labs(y = "Fraction variance explained", title = "Liver size") +
  scale_x_discrete(labels = c("Shared\nenvironment", "Additive\ngenetic", "Parental", "Residual")) +
  theme(panel.grid.major.x = element_blank(),
        axis.title.x = element_blank(),
        legend.title = element_blank())
```

## Fish growth

Does fish growth depend on parasite genes? We'll focus on just the infected fish. 

```{r}
dat_fgrow2 <- dat_fgrow%>%filter(schisto_inf == 1)
```

Here are the number of full-sib families and the number of offspring phenotyped per family.

```{r}
dat_fgrow2%>%
  group_by(worm_full_sib_fam_id)%>%
  summarize(n_pheno = sum(!is.na(fish_final_bm_g_noworm)))%>%
  ungroup()%>%
  filter(!is.na(worm_full_sib_fam_id))%>%
  summarize(fs_fams = n(),
            min_phenotyped = min(n_pheno),
            lwrIQR_phenotyped = quantile(n_pheno, prob = 0.25),
            avg_phenotyped = mean(n_pheno),
            uprIQR_phenotyped = quantile(n_pheno, prob = 0.75),
            max_phenotyped = max(n_pheno)
            )
```

Here are the number of offspring phenotyped per dam/sire. There are not enough phenotyped per dam where I feel comfortable in estimating the 'dam' effect.

```{r}
dat_fgrow2%>%
  group_by(maternal_match, worm_maternal_id)%>%
  summarize(n_pheno = sum(!is.na(fish_final_bm_g_noworm)))%>%
  group_by(maternal_match)%>%
  filter(!is.na(worm_maternal_id))%>%
  summarize(dams = n(),
            min_phenotyped = min(n_pheno),
            lwrIQR_phenotyped = quantile(n_pheno, prob = 0.25),
            avg_phenotyped = mean(n_pheno),
            uprIQR_phenotyped = quantile(n_pheno, prob = 0.75),
            max_phenotyped = max(n_pheno)
            )
```

```{r}
parents <- dat_fgrow2%>%
  group_by(eggs_clutch_id)%>%
  summarise(n_parent = n(), fgrow_p = mean(fish_final_bm_g_noworm - fish_initial_bm_g))%>%
  na.omit()

offspring <- dat_fgrow2%>%
  group_by(worm_full_sib_fam_id)%>%
  summarise(n_offspring = n(), fgrow_o = mean(fish_final_bm_g_noworm - fish_initial_bm_g, na.rm = T))%>%
  na.omit()

parent_offspring <- left_join(parents, offspring, by = c("eggs_clutch_id" = "worm_full_sib_fam_id"))
```

For fish growth, there is not much of a parent-offspring regression.

```{r}
ggplot(parent_offspring, aes(fgrow_p, fgrow_o)) + geom_point() +geom_smooth(se=F)
```

So, let's fit the same 5 models again, acknowledging that the models with 'dam' effects may be overfitted.

```{r}
thin_fgrow <- 50
burnin_fgrow <- 1000
nit_fgrow <- thin_fgrow * target_samples + burnin_fgrow

# fixed effects
m1_fgrow_noib <- MCMCglmm(fish_final_bm_g_noworm ~ age_diss_cen + fish_sex_cen + fish_initial_bm_cen, 
                 random = ~tank_id,
                 family = "gaussian",
                 data = dat_fgrow2,
                 prior = p1_g,
                 verbose=FALSE,
                 nitt = nit_fgrow, 
                 thin = thin_fgrow,
                 burnin = burnin_fgrow
                )
# fixed effects
m1_fgrow <- MCMCglmm(fish_final_bm_g_noworm ~ ib + age_diss_cen + fish_sex_cen + fish_initial_bm_cen, 
                 random = ~tank_id,
                 family = "gaussian",
                 data = dat_fgrow2,
                 prior = p1_g,
                 verbose=FALSE,
                 nitt = nit_fgrow, 
                 thin = thin_fgrow,
                 burnin = burnin_fgrow
                )
# fam
m2_fgrow_fam <- MCMCglmm(fish_final_bm_g_noworm ~ ib + age_diss_cen + fish_sex_cen + fish_initial_bm_cen, 
                 random = ~tank_id + worm_full_sib_fam_id,
                 family = "gaussian",
                 data = dat_fgrow2,
                 prior = p2_g,
                 verbose=FALSE,
                 pr=T,
                 nitt = nit_fgrow, 
                 thin = thin_fgrow,
                 burnin = burnin_fgrow
                )
# full sib x line
m2_fgrow_famXline <- MCMCglmm(
  fish_final_bm_g_noworm ~ ib + age_diss_cen + fish_sex_cen + fish_initial_bm_cen,
  random = ~tank_id + idh(line):worm_full_sib_fam_id,
  family = "gaussian",
  data = dat_fgrow2,
  prior = list(R = list(V = 1, nu = 0.002),
               G = list(G1 = list(V = 1, nu = 0.002),
                        G2 = list(V = diag(4), nu = 0.002))),
  verbose=FALSE,
  pr = F,
  nitt = nit_fgrow, 
  thin = thin_fgrow,
  burnin = burnin_fgrow
  )
# anim model
m2_fgrow_anim <- MCMCglmm(fish_final_bm_g_noworm ~ ib + age_diss_cen + fish_sex_cen + fish_initial_bm_cen, 
                 random = ~tank_id + id,
                 ginverse = list(id = Ainv),
                 family = "gaussian",
                 data = dat_fgrow2,
                 prior = p2_g,
                 verbose=F,
                 nitt = nit_fgrow, 
                 thin = thin_fgrow,
                 burnin = burnin_fgrow
                )
# fam
m3_fgrow_mat <- MCMCglmm(fish_final_bm_g_noworm ~ ib + age_diss_cen + fish_sex_cen + fish_initial_bm_cen, 
                 random = ~tank_id + worm_full_sib_fam_id + dam,
                 family = "gaussian",
                 data = dat_fgrow2%>%mutate(dam = factor(dam)),
                 prior = p3_g,
                 verbose=FALSE,
                 nitt = nit_fgrow, 
                 thin = thin_fgrow,
                 burnin = burnin_fgrow
                )
# anim model
m3_fgrow_anim <- MCMCglmm(fish_final_bm_g_noworm ~ ib + age_diss_cen + fish_sex_cen + fish_initial_bm_cen, 
                 random = ~tank_id + id + dam,
                 ginverse = list(id = Ainv),
                 family = "gaussian",
                 data = dat_fgrow2%>%mutate(dam = factor(dam)),
                 prior = p3_g,
                 verbose=F,
                 nitt = nit_fgrow, 
                 thin = thin_fgrow,
                 burnin = burnin_fgrow
                )
```

Adding full-sibs (red) and then the pedigree (green) appear to be only slight improvements.

```{r}
plot(mcmc.list(
  m1_fgrow$Deviance, m2_fgrow_fam$Deviance, m2_fgrow_anim$Deviance
  ), density = F)
```

The full-sib model suggests that little variance within clutches is explained by parental effects.

```{r}
plot(mcmc.list(
  m2_fgrow_fam$Deviance, m3_fgrow_mat$Deviance
  ), density = F)
```

Same for the animal model.

```{r}
plot(mcmc.list(
  m2_fgrow_anim$Deviance, m3_fgrow_anim$Deviance
  ), density = F)
```

For, these models, we break down total variance into several components: fixed effects (starting conditions), shared environment, genetic effects, and parental effects.

```{r}
m_vc_fgrow <- get_var_comps("fgrow", dat_fgrow2, 'tank_id', 
                            m1_fgrow_noib, m1_fgrow,
                            m2_fgrow_fam, m2_fgrow_anim,
                            m3_fgrow_mat, m3_fgrow_anim)

m_vc_fgrow2 <- m_vc_fgrow%>%filter(v_name %in% c("Ve2", "Vg2", "Vm2", "Vr2")) # var with fixed effects
m_vc_fgrow <- m_vc_fgrow%>%filter(!v_name %in% c("Ve2", "Vg2", "Vm2", "Vr2")) # var excluding fixed effects
```

Here is the percent variation in final fish size accounted for by initial fish size, sex, and age at dissection across models. It is high, which is not surprising.

```{r}
filter(m_vc_fgrow, v_name == "Vf")%>%
  mutate(across(v_fit:v_upr, function(x){round(x,3)}))
```

Here is the percent variation in fish growth accounted for by shared environments (tanks). Infected fish from the same tank tended to exhibit similar growth.

```{r}
filter(m_vc_fgrow, v_name == "Ve")%>%
  mutate(across(v_fit:v_upr, function(x){round(x,3)}))
```

Here is the percent variation in fgrow size accounted for by genetic variance (i.e. the heritability). It is low.

```{r}
filter(m_vc_fgrow, v_name == "Vg")%>%
  mutate(across(v_fit:v_upr, function(x){round(x,3)}))
```

The heritability is higher if we only consider the phenotypic variance left after accounting for fixed effects (here for the animal model):

```{r}
filter(m_vc_fgrow2, v_name == "Vg2")%>%
  mutate(across(v_fit:v_upr, function(x){round(x,3)}))
```

The distribution of family effects is normal.

```{r}
mean_fgrow <- mean(dat_fgrow2$fish_final_bm_g_noworm, na.rm = T)
re_fams_fgrow <- dat_fgrow2%>%
  filter(is.na(pred))%>%
  group_by(worm_full_sib_fam_id, gen)%>%
  summarize(fam_avg = mean(fish_final_bm_g_noworm, na.rm = T) - mean_fgrow)%>%
  ungroup()%>%
  mutate(approach="unadjusted")

re <- colMeans(m2_fgrow_fam$Sol[,which(grepl("worm_full_sib", colnames(m2_fgrow_fam$Sol)))])
re <- data.frame(worm_full_sib_fam_id = gsub("worm_full_sib_fam_id.", "", x = names(re)),
                 fam_avg = re,
                 approach="model-estimated")

re_fams_fgrow <- bind_rows(
  re_fams_fgrow,
  re
)%>%
  mutate(approach = fct_relevel(factor(approach), c("unadjusted", "model-estimated")))

ggplot(re_fams_fgrow, aes(x = fam_avg)) +
  geom_histogram(aes(y=..density..),
                 colour = "black",
                 fill = "white") +
  geom_function(data = filter(re_fams_fgrow, approach == "unadjusted"),
                fun = dnorm, 
                args = list(mean = 0, 
                            sd = re_fams_fgrow%>%filter(approach == "unadjusted")%>%.$fam_avg%>%sd(., na.rm=T)
                            ),
                color = 'red',
                linetype = "dashed") +
  geom_function(data = filter(re_fams_fgrow, approach == "model-estimated"),
                fun = dnorm, 
                args = list(mean = 0, 
                            sd = re_fams_fgrow%>%filter(approach == "model-estimated")%>%.$fam_avg%>%sd(., na.rm=T)
                            ),
                color = 'red',
                linetype = "dashed") +
  geom_density() +
  geom_label(data = re_fams_fgrow%>%
               filter(!is.infinite(fam_avg))%>%
               group_by(approach)%>%
               summarise(min_x = min(fam_avg, na.rm = T),
                         skew = e1071::skewness(fam_avg, na.rm = T))%>%
               mutate(skew_c = paste0("Skewness: ", round(skew, 2))),
             aes(label = skew_c,
                 x = min_x,
                 y = 1),
             hjust = -0.05, vjust = 0) +
  facet_wrap(~approach, scales = "free") +
  labs(y = "Density", x = "Family averages")
```

The influence of 'parent' on fish growth size is small. But there was limited data to test such parental effects.

```{r}
filter(m_vc_fgrow, v_name == "Vm")%>%
  mutate(across(v_fit:v_upr, function(x){round(x,3)}))
```

Here is the total variance explained by each model.

```{r}
filter(m_vc_fgrow, v_name == "Vr")%>%
  mutate(across(v_fit:v_upr, function(x){round(1-x,3)}))
```


The fraction of phenotypic variance explained by each term in the different models is plotted here

```{r}
ggplot(m_vc_fgrow%>%
         mutate(v_name = fct_relevel(factor(v_name), c("Vf", "Ve", "Vg", "Vm", "Vr"))),
       aes(v_name, v_fit, color = mod)) +
  geom_pointrange(aes(ymin = v_lwr, ymax = v_upr),
                  position = position_dodge(width = 0.6)) +
  labs(y = "Fraction variance explained", title = "Fish growth") +
  scale_x_discrete(labels = c("Starting\nconditions", "Shared\nenvironment", "Additive\ngenetic", "Parental", "Residual")) +
  theme(panel.grid.major.x = element_blank(),
        axis.title.x = element_blank(),
        legend.title = element_blank())
```

Since fish final size depended strongly on initial size, we should look at these fractions excluding the effect of starting conditons. The effect of parasite genes now appears a bit larger.

```{r}
ggplot(m_vc_fgrow2%>%
         mutate(v_name = fct_relevel(factor(v_name), c("Ve2", "Vg2", "Vm2", "Vr2"))),
       aes(v_name, v_fit, color = mod)) +
  geom_pointrange(aes(ymin = v_lwr, ymax = v_upr),
                  position = position_dodge(width = 0.6)) +
  labs(y = "Fraction variance explained", title = "Fish growth") +
  scale_x_discrete(labels = c("Shared\nenvironment", "Additive\ngenetic", "Parental", "Residual")) +
  theme(panel.grid.major.x = element_blank(),
        axis.title.x = element_blank(),
        legend.title = element_blank())
```

# Breeding traits

## Fecundity 

```{r}
dat_bred <- dat_bred%>%mutate(log_eggs = log(eggnum_d2_6_indiv))
```

Here are the number of full-sib families and the number of offspring phenotyped per family.

```{r}
dat_bred%>%
  group_by(worm_full_sib_fam_id)%>%
  summarize(n_pheno = sum(!is.na(log_eggs)))%>%
  ungroup()%>%
  filter(!is.na(worm_full_sib_fam_id))%>%
  summarize(fs_fams = n(),
            min_phenotyped = min(n_pheno),
            lwrIQR_phenotyped = quantile(n_pheno, prob = 0.25),
            avg_phenotyped = mean(n_pheno),
            uprIQR_phenotyped = quantile(n_pheno, prob = 0.75),
            max_phenotyped = max(n_pheno)
            )
```

Here are the number of offspring phenotyped per dam/sire. There are not enough phenotyped per dam to estimate the 'dam' effect.

```{r}
dat_bred%>%
  group_by(maternal_match, worm_maternal_id)%>%
  summarize(n_pheno = sum(!is.na(log_eggs)))%>%
  group_by(maternal_match)%>%
  filter(!is.na(worm_maternal_id))%>%
  summarize(dams = n(),
            min_phenotyped = min(n_pheno),
            lwrIQR_phenotyped = quantile(n_pheno, prob = 0.25),
            avg_phenotyped = mean(n_pheno),
            uprIQR_phenotyped = quantile(n_pheno, prob = 0.75),
            max_phenotyped = max(n_pheno)
            )
```

```{r}
parents <- dat_bred%>%
  group_by(eggs_clutch_id)%>%
  summarise(n_parent = n(), eggs_p = mean(log_eggs))%>%
  na.omit()

offspring <- dat_bred%>%
  group_by(worm_full_sib_fam_id)%>%
  summarise(n_offspring = n(), eggs_o = mean(log_eggs, na.rm = T))%>%
  na.omit()

parent_offspring <- left_join(parents, offspring, by = c("eggs_clutch_id" = "worm_full_sib_fam_id"))
```

Fecundity does not exhibit much of a parent-offspring regression. This is expected, since it does not account for environmental variance in parasite size and thus fecundity.

```{r}
ggplot(parent_offspring, aes(eggs_p, eggs_o)) + geom_point() +geom_smooth(se=F)
```



```{r}
thin_bred <- 200
burnin_bred <- 1000
nit_bred <- thin_bred * target_samples + burnin_bred

# fixed effects, noIB
m1_bred_noib <- MCMCglmm(log_eggs ~ log_fish_initial_bm_cen + fish_sex_cen + age_diss_cen, 
                 random = ~breeding_block,
                 family = "gaussian",
                 data = dat_bred,
                 prior = p1_g,
                 verbose=FALSE,
                 nitt = nit_bred, 
                 thin = thin_bred,
                 burnin = burnin_bred
                )
# fixed effects
m1_bred <- MCMCglmm(log_eggs ~ ib + log_fish_initial_bm_cen + fish_sex_cen + age_diss_cen, 
                 random = ~breeding_block,
                 family = "gaussian",
                 data = dat_bred,
                 prior = p1_g,
                 verbose=FALSE,
                 nitt = nit_bred, 
                 thin = thin_bred,
                 burnin = burnin_bred
                )
# fam
m2_bred_fam <- MCMCglmm(log_eggs ~ ib + log_fish_initial_bm_cen + fish_sex_cen + age_diss_cen, 
                 random = ~breeding_block + worm_full_sib_fam_id,
                 family = "gaussian",
                 data = dat_bred,
                 prior = p2_g,
                 pr = T,
                 verbose=FALSE,
                 nitt = nit_bred, 
                 thin = thin_bred,
                 burnin = burnin_bred
                )
# full sib x line
m2_bred_famXline <- MCMCglmm(
  log_eggs ~ ib + log_fish_initial_bm_cen + fish_sex_cen + age_diss_cen, 
  random = ~breeding_block + idh(line):worm_full_sib_fam_id,
  family = "gaussian",
  data = dat_bred,
  prior = list(R = list(V = 1, nu = 0.002),
               G = list(G1 = list(V = 1, nu = 0.002),
                        G2 = list(V = diag(4), nu = 0.002))),
  verbose=FALSE,
  pr = F,
  nitt = nit_bred, 
  thin = thin_bred,
  burnin = burnin_bred
  )

# anim model
m2_bred_anim <- MCMCglmm(log_eggs ~ ib + log_fish_initial_bm_cen + fish_sex_cen + age_diss_cen, 
                 random = ~breeding_block + id,
                 ginverse = list(id = Ainv),
                 family = "gaussian",
                 data = dat_bred,
                 prior = p2_g,
                 verbose=F,
                 nitt = nit_bred, 
                 thin = thin_bred,
                 burnin = burnin_bred
                )
```

Adding full-sibs (red) and the pedigree (green) is an improvement.

```{r}
plot(mcmc.list(
  m1_bred$Deviance, m2_bred_fam$Deviance, m2_bred_anim$Deviance
  ), density = F)
```

For, these models, we break down total variance into several components: fixed effects (starting conditions), shared environment, genetic effects, and parental effects.

```{r}
# first model, no inbreeding
pdx <- m1_bred_noib$X # model matrix, fixed effx
n <- dim(pdx)[1] # number of data points in model, all traits
nt <- n/1 # number of data points per trait
p_i <- which(is.na(dat_bred$pred)) # points where we can predicted vals and cred int
pdx <- pdx[c(p_i),] # restrict to only points where we want preds
p_n <- dim(pdx)[1]/1 # number of points for each trait we want to predict
num_fe <- m1_bred_noib$Fixed$nfl # number of fixed effx, rand effx marginalized
p_all <- as.matrix(pdx) %*% t(m1_bred_noib$Sol[,1:num_fe]) # predicteds via matrix mult 
m1_Vf_noib <- apply(p_all, MARGIN = 2, var)

# first model
pdx <- m1_bred$X # model matrix, fixed effx
n <- dim(pdx)[1] # number of data points in model, all traits
nt <- n/1 # number of data points per trait
p_i <- which(is.na(dat_bred$pred)) # points where we can predicted vals and cred int
pdx <- pdx[c(p_i),] # restrict to only points where we want preds
p_n <- dim(pdx)[1]/1 # number of points for each trait we want to predict
num_fe <- m1_bred$Fixed$nfl # number of fixed effx, rand effx marginalized
p_all <- as.matrix(pdx) %*% t(m1_bred$Sol[,1:num_fe]) # predicteds via matrix mult 
m1_Vf <- apply(p_all, MARGIN = 2, var)

pdx <- m2_bred_fam$X # model matrix, fixed effx
n <- dim(pdx)[1] # number of data points in model, all traits
nt <- n/1 # number of data points per trait
p_i <- which(is.na(dat_bred$pred)) # points where we can predicted vals and cred int
pdx <- pdx[c(p_i),] # restrict to only points where we want preds
p_n <- dim(pdx)[1]/1 # number of points for each trait we want to predict
num_fe <- m2_bred_fam$Fixed$nfl # number of fixed effx, rand effx marginalized
p_all <- as.matrix(pdx) %*% t(m2_bred_fam$Sol[,1:num_fe]) # predicteds via matrix mult 
m2_Vf_fam <- apply(p_all, MARGIN = 2, var)

pdx <- m2_bred_anim$X # model matrix, fixed effx
n <- dim(pdx)[1] # number of data points in model, all traits
nt <- n/1 # number of data points per trait
p_i <- which(is.na(dat_bred$pred)) # points where we can predicted vals and cred int
pdx <- pdx[c(p_i),] # restrict to only points where we want preds
p_n <- dim(pdx)[1]/1 # number of points for each trait we want to predict
num_fe <- m2_bred_anim$Fixed$nfl # number of fixed effx, rand effx marginalized
p_all <- as.matrix(pdx) %*% t(m2_bred_anim$Sol[,1:num_fe]) # predicteds via matrix mult 
m2_Vf_anim <- apply(p_all, MARGIN = 2, var)
```
```{r}
# PHENOTYPIC VAR
m1_Vp <- m1_Vf + rowSums(m1_bred$VCV)
m1_Vp_noib <- m1_Vf_noib + rowSums(m1_bred_noib$VCV)
m2_Vp_fam <- m2_Vf_fam + rowSums(m2_bred_fam$VCV)
m2_Vp_anim <- m2_Vf_anim + rowSums(m2_bred_anim$VCV)
```
```{r}
# vc for model 1, no inbreeding
m1_vc_noib <- data.frame(mod = "1a_base_noIB", 
                    v_est = c(
                      m1_Vf_noib/m1_Vp_noib, 
                      m1_bred_noib$VCV[,'breeding_block']/m1_Vp_noib,
                      m1_bred_noib$VCV[,'breeding_block']/(m1_Vp_noib - m1_Vf_noib),
                      1-(m1_Vf_noib+m1_bred_noib$VCV[,'breeding_block'])/m1_Vp_noib,
                      1-(m1_bred_noib$VCV[,'breeding_block'])/(m1_Vp_noib - m1_Vf_noib)
                      ),
                    v_name = rep(c("Vf", "Ve", "Ve2", "Vr", "Vr2"), 
                                 times = rep(length(m1_Vf), 5))
                    )

# vc for model 1
m1_vc <- data.frame(mod = "1b_base", 
                    v_est = c(
                      m1_Vf/m1_Vp, 
                      m1_bred$VCV[,'breeding_block']/m1_Vp,
                      m1_bred$VCV[,'breeding_block']/(m1_Vp - m1_Vf),
                      1-(m1_Vf+m1_bred$VCV[,'breeding_block'])/m1_Vp,
                      1-(m1_bred$VCV[,'breeding_block'])/(m1_Vp - m1_Vf)
                      ),
                    v_name = rep(c("Vf", "Ve", "Ve2", "Vr", "Vr2"), 
                                 times = rep(length(m1_Vf), 5))
                    )

# vc for model 2 full sib
m2_vc <- data.frame(mod = "2a_full_sibs", 
                    v_est = c(
                      m2_Vf_fam/m2_Vp_fam, 
                      m2_bred_fam$VCV[,'breeding_block']/m2_Vp_fam, 
                      m2_bred_fam$VCV[,'breeding_block']/(m2_Vp_fam - m2_Vf_fam), 
                      2*m2_bred_fam$VCV[,'worm_full_sib_fam_id']/m2_Vp_fam, 
                      2*m2_bred_fam$VCV[,'worm_full_sib_fam_id']/(m2_Vp_fam - m2_Vf_fam),
                      1-(m2_Vf_fam+m2_bred_fam$VCV[,'breeding_block'] +
                           2*m2_bred_fam$VCV[,'worm_full_sib_fam_id'])/m2_Vp_fam,
                      1-(m2_bred_fam$VCV[,'breeding_block'] +
                           2*m2_bred_fam$VCV[,'worm_full_sib_fam_id'])/(m2_Vp_fam - m2_Vf_fam)
                      ),
                    v_name = rep(c("Vf", "Ve", "Ve2", "Vg", "Vg2", "Vr", "Vr2"), 
                            times = rep(length(m2_Vf_fam), 7))
                    )

# vc for model 2 anim
m2_vc_anim <- data.frame(mod = "2b_animal", 
                    v_est = c(
                      m2_Vf_anim/m2_Vp_anim, 
                      m2_bred_anim$VCV[,'breeding_block']/m2_Vp_anim, 
                      m2_bred_anim$VCV[,'breeding_block']/(m2_Vp_anim - m2_Vf_anim), 
                      m2_bred_anim$VCV[,'id']/m2_Vp_anim, 
                      m2_bred_anim$VCV[,'id']/(m2_Vp_anim - m2_Vf_anim),
                      1-(m2_Vf_anim+m2_bred_anim$VCV[,'breeding_block'] +
                           m2_bred_anim$VCV[,'id'])/m2_Vp_anim,
                      1-(m2_bred_anim$VCV[,'breeding_block'] +
                           m2_bred_anim$VCV[,'id'])/(m2_Vp_anim - m2_Vf_anim)
                      ),
                    v_name = rep(c("Vf", "Ve", "Ve2", "Vg", "Vg2", "Vr", "Vr2"), 
                            times = rep(length(m2_Vf_anim), 7))
                    )


m_vc <- bind_rows(
  m1_vc_noib, m1_vc,
  m2_vc, m2_vc_anim
)


m_vc_bred <- m_vc%>%
  group_by(mod, v_name)%>%
  summarise(v_fit = median(v_est),
            v_lwr = quantile(v_est, probs = 0.025),
            v_upr = quantile(v_est, probs = 0.975),
            )%>%
  ungroup()%>%
  mutate(trait = "eggs")

m_vc_bred2 <- m_vc_bred%>%filter(v_name %in% c("Ve2", "Vg2", "Vm2", "Vr2")) # var with fixed effects
m_vc_bred <- m_vc_bred%>%filter(!v_name %in% c("Ve2", "Vg2", "Vm2", "Vr2")) # var excluding fixed effects
```

Here is the percent variation in log-transformed egg quantity accounted for by fish size, sex, and age at dissection across models. It is high, since bigger, older fish have bigger worms that produce more eggs.

```{r}
filter(m_vc_bred, v_name == "Vf")%>%
  mutate(across(v_fit:v_upr, function(x){round(x,3)}))
```

Here is the percent variation in fecundity attributable to shared environments (i.e. being bred at the same time). Breeding block does not affect fecundity.

```{r}
filter(m_vc_bred, v_name == "Ve")%>%
  mutate(across(v_fit:v_upr, function(x){round(x,3)}))
```

Here is the percent variation in fecundity accounted for by genetic variance (i.e. the heritability). It is significant.

```{r}
filter(m_vc_bred, v_name == "Vg")%>%
  mutate(across(v_fit:v_upr, function(x){round(x,3)}))
```

When we plot the family-level random effects, we see that this could be driven by an outlier with very low fecundity.

```{r}
mean_bred <- mean(dat_bred$log_eggs, na.rm = T)
re_fams_bred <- dat_bred%>%
  filter(is.na(pred))%>%
  group_by(worm_full_sib_fam_id, gen)%>%
  summarize(fam_avg = mean(log_eggs, na.rm = T) - mean_bred)%>%
  ungroup()%>%
  mutate(approach="unadjusted")

re <- colMeans(m2_bred_fam$Sol[,which(grepl("worm_full_sib", colnames(m2_bred_fam$Sol)))])
re <- data.frame(worm_full_sib_fam_id = gsub("worm_full_sib_fam_id.", "", x = names(re)),
                 fam_avg = re,
                 approach="model-estimated")

re_fams_bred <- bind_rows(
  re_fams_bred,
  re
)%>%
  mutate(approach = fct_relevel(factor(approach), c("unadjusted", "model-estimated")))

ggplot(re_fams_bred, aes(x = fam_avg)) +
  geom_histogram(aes(y=..density..),
                 colour = "black",
                 fill = "white") +
  geom_function(data = filter(re_fams_bred, approach == "unadjusted"),
                fun = dnorm, 
                args = list(mean = 0, 
                            sd = re_fams_bred%>%filter(approach == "unadjusted")%>%.$fam_avg%>%sd(., na.rm=T)
                            ),
                color = 'red',
                linetype = "dashed") +
  geom_function(data = filter(re_fams_bred, approach == "model-estimated"),
                fun = dnorm, 
                args = list(mean = 0, 
                            sd = re_fams_bred%>%filter(approach == "model-estimated")%>%.$fam_avg%>%sd(., na.rm=T)
                            ),
                color = 'red',
                linetype = "dashed") +
  geom_density() +
  geom_label(data = re_fams_bred%>%
               filter(!is.infinite(fam_avg))%>%
               group_by(approach)%>%
               summarise(min_x = min(fam_avg, na.rm = T),
                         skew = e1071::skewness(fam_avg, na.rm = T))%>%
               mutate(skew_c = paste0("Skewness: ", round(skew, 2))),
             aes(label = skew_c,
                 x = min_x,
                 y = 1),
             hjust = -0.05, vjust = 0) +
  facet_wrap(~approach, scales = "free") +
  labs(y = "Density", x = "Family averages")
```

When we exclude the outlier with very low fecundity, the distribution is still skewed, but much less so.

```{r}
re_fams_bred_no_outlier <- re_fams_bred%>%filter(worm_full_sib_fam_id != "1_S2")

ggplot(re_fams_bred_no_outlier, aes(x = fam_avg)) +
  geom_histogram(aes(y=..density..),
                 colour = "black",
                 fill = "white") +
  geom_function(data = filter(re_fams_bred_no_outlier, approach == "unadjusted"),
                fun = dnorm, 
                args = list(mean = 0, 
                            sd = re_fams_bred_no_outlier%>%filter(approach == "unadjusted")%>%.$fam_avg%>%sd(., na.rm=T)
                            ),
                color = 'red',
                linetype = "dashed") +
  geom_function(data = filter(re_fams_bred_no_outlier, approach == "model-estimated"),
                fun = dnorm, 
                args = list(mean = 0, 
                            sd = re_fams_bred_no_outlier%>%filter(approach == "model-estimated")%>%.$fam_avg%>%sd(., na.rm=T)
                            ),
                color = 'red',
                linetype = "dashed") +
  geom_density() +
  geom_label(data = re_fams_bred_no_outlier%>%
               filter(!is.infinite(fam_avg))%>%
               group_by(approach)%>%
               summarise(min_x = min(fam_avg, na.rm = T),
                         skew = e1071::skewness(fam_avg, na.rm = T))%>%
               mutate(skew_c = paste0("Skewness: ", round(skew, 2))),
             aes(label = skew_c,
                 x = min_x,
                 y = 1),
             hjust = -0.05, vjust = 0) +
  facet_wrap(~approach, scales = "free") +
  labs(y = "Density", x = "Family averages")
```

Here is the total variance explained by each model.

```{r}
filter(m_vc_bred, v_name == "Vr")%>%
  mutate(across(v_fit:v_upr, function(x){round(1-x,3)}))
```

The fraction of phenotypic variance explained by each term in the different models is plotted here.

```{r}
ggplot(m_vc_bred%>%
         mutate(v_name = fct_relevel(factor(v_name), c("Vf", "Ve", "Vg", "Vm", "Vr"))),
       aes(v_name, v_fit, color = mod)) +
  geom_pointrange(aes(ymin = v_lwr, ymax = v_upr),
                  position = position_dodge(width = 0.6)) +
  labs(y = "Fraction variance explained", title = "Fecundity") +
  scale_x_discrete(labels = c("Starting\nconditions", "Shared\nenvironment", "Additive\ngenetic", "Residual")) +
  theme(panel.grid.major.x = element_blank(),
        axis.title.x = element_blank(),
        legend.title = element_blank())
```

The fraction of variance attributed to genes goes up after accounting for fish characteristics.

```{r}
ggplot(m_vc_bred2%>%
         mutate(v_name = fct_relevel(factor(v_name), c("Ve2", "Vg2", "Vm2", "Vr2"))),
       aes(v_name, v_fit, color = mod)) +
  geom_pointrange(aes(ymin = v_lwr, ymax = v_upr),
                  position = position_dodge(width = 0.6)) +
  labs(y = "Fraction variance explained", title = "Plerocercoid size") +
  scale_x_discrete(labels = c("Shared\nenvironment", "Additive\ngenetic", "Parental", "Residual")) +
  theme(panel.grid.major.x = element_blank(),
        axis.title.x = element_blank(),
        legend.title = element_blank())
```

## Hatching

We do not consider heritability in hatching rate because it is much more a clutch-level than individual-level trait. That is, hatching is determined less by individual genotype and more by the combination of parental genotypes.


# Conclusions

## Trait variance

Most traits exhibited some genetic variation and non-zero heritability. However, phenotypes were often strongly determined by the uncontrolled starting conditions, particularly for traits in fish, as fish size has a big effect on worm size. 

```{r}
m_vc_comb <- bind_rows(
  m_vc_cinf, m_vc_cerc,
  m_vc_fail, m_vc_proc,
  m_vc_surv,
  m_vc_finf, m_vc_fsurv,
  m_vc_pler, m_vc_liver,
  m_vc_fgrow,
  m_vc_bred
)%>%
  mutate(trait = fct_relevel(factor(trait), 
                             levels = c(
                               "cinf", "cerc", "proc", "fail", "surv",
                               "finf", "fsurv", "pler", "liver", "fgrow",
                               "eggs"
                             )))

m_vc_comb2 <- bind_rows(
  m_vc_cinf2, m_vc_cerc2,
  m_vc_fail2, m_vc_proc2,
  m_vc_surv2,
  m_vc_finf2, m_vc_fsurv2,
  m_vc_pler2, m_vc_liver2,
  m_vc_fgrow2,
  m_vc_bred2
)%>%
  mutate(trait = fct_relevel(factor(trait), 
                             levels = c(
                               "cinf", "cerc", "proc", "fail", "surv",
                               "finf", "fsurv", "pler", "liver", "fgrow",
                               "eggs"
                             )))
```

Here is the breakdown of phenotypic variance for each trait. 

```{r}
ggplot(m_vc_comb%>%
         mutate(v_name = fct_relevel(factor(v_name), c("Vf", "Ve", "Vg", "Vm", "Vr"))),
       aes(v_name, v_fit, color = mod)) +
  geom_pointrange(aes(ymin = v_lwr, ymax = v_upr),
                  position = position_dodge(width = 0.75)) +
  labs(y = "Fraction variance explained") +
  scale_x_discrete(labels = c("Starting\nconditions", "Shared\nenvironment", "Additive\ngenetic", "Parental", "Residual")) +
  facet_wrap(~ trait, ncol = 5) +
  scale_color_brewer(palette = "Paired") +
  theme(panel.grid.major.x = element_blank(),
        axis.title.x = element_blank(),
        legend.title = element_blank())
```

Maybe it is easier to understand if we put trait on the x-axis and panel by variance component.

```{r}
m_vc_comb_plotx <- m_vc_comb%>%
  filter(
    !(mod %in% c("3a_full_sibs", "3b_animal") & !trait %in% c("cinf", "cerc", "proc", "surv"))
    )%>%
  mutate(v_name = case_when(v_name == "Vf" ~ "Starting conditions",
                            v_name == "Ve" ~ "Shared environment",
                            v_name == "Vg" ~ "Additive genetic",
                            v_name == "Vm" ~ "Parental",
                            v_name == "Vr" ~ "Residual"))%>%
  mutate(v_name = fct_relevel(factor(v_name), 
                              c("Starting conditions", 
                                "Shared environment",
                                "Additive genetic",
                                "Parental",
                                "Residual")))
```
```{r}
px_modcomps <- ggplot(m_vc_comb_plotx%>%
         filter(v_name != "Residual", mod != "1a_base_noIB"),
       aes(trait, v_fit, color = mod)) +
  geom_pointrange(
    aes(ymin = v_lwr, ymax = v_upr),
    position = position_dodge(width = 0.75),
    size=0.15) +
  labs(y = "Fraction variance explained", color = "Model") +
  facet_wrap(~ v_name, ncol = 2, scales = "free_y") +
  scale_x_discrete(
    labels = c("cinf", "cerc", "proc", "fail", "csurv",
               "finf", "fsurv", "pler", "liver", "fgrow", "eggs"
               )
    ) +
  scale_color_manual(values = RColorBrewer::brewer.pal(n=6,name = "Paired")[2:6],
                     labels = c("(1) base",
                                "(2a) base + sibship",
                                "(2b) base + pedigree",
                                "(3a) 2a + dam",
                                "(3a) 2b + dam")) +
  guides(colour = guide_legend(override.aes = list(size = .5))) +
  theme(panel.grid.major.x = element_blank(),
        axis.title.x = element_blank(),
        axis.text.x = element_text(angle = 45, hjust = 1),
        legend.background = element_rect(color="black"),
        legend.justification = c(1,1),
        legend.position = c(0.99,0.42))
```

Now we can see that different traits were affected by starting conditions and shared environments differently. However, the variance attributed to these effects is not model-dependent. The bottom left plot shows the heritability estimated by the various models. This estimate includes the entire phenotypic variance, i.e. we do not remove the variance accounted for by starting condition (the fixed effects). This is why traits in fish have low heritabilities - those traits depend stongly on fish characteristics. Also, the various models yield similar heritabilities, suggesting that 'parental' effects are not strong, nor are full-sib variances strongly inflated.

```{r}
px_modcomps
```

```{r}
ggsave(px_modcomps, filename = "../figs/figS_modcomps.png", width = 90*2, height = 60*2, units = "mm")
ggsave(px_modcomps, filename = "../figs/figS_modcomps.svg", width = 90*2, height = 60*2, units = "mm")
```

Here are the models and traits where *starting conditions* explained essentially no variation (i.e. the 95% CI includes zero < 0.5%).

```{r}
m_vc_comb%>%
  filter(v_name == "Vf", v_lwr < 0.005)
```

Here are the models and traits where *environmental* effects explained essentially no variation.

```{r}
m_vc_comb%>%
  filter(v_name == "Ve", v_lwr < 0.005)
```

Here are the models and traits where *genetic* effects explained essentially no variation.

```{r}
m_vc_comb%>%
  filter(v_name == "Vg", v_lwr < 0.005)
```

Here are the models and traits where *maternal* effects explained essentially no variation.

```{r}
m_vc_comb%>%
  filter(v_name == "Vm", v_lwr < 0.005)
```

```{r}
# ggplot(m_vc_comb%>%
#          filter(v_name == "Vg"), aes(trait, v_fit)) +
#   geom_pointrange(aes(ymin = v_lwr, ymax = v_upr, color = mod),
#                   position = position_dodge(width = 0.6)) +
#   # facet_wrap(~mod) +
#   labs(y = "Vg/Vp") +
#   theme(axis.title.x = element_blank(),
#         panel.grid.major.x = element_blank())
```

We can also examine phenotypic variance excluding the variance explained by starting conditions (i.e. the fixed effects). In general this does not change much.

```{r}
m_vc_comb_plotx <- m_vc_comb2%>%
  filter(
    !(mod %in% c("3a_full_sibs", "3b_animal") & !trait %in% c("cinf", "cerc", "proc", "surv"))
    )%>%
  mutate(v_name = case_when(v_name == "Ve2" ~ "Shared environment",
                            v_name == "Vg2" ~ "Additive genetic",
                            v_name == "Vm2" ~ "Parental",
                            v_name == "Vr2" ~ "Residual"))%>%
  mutate(v_name = fct_relevel(factor(v_name), 
                              c("Shared environment",
                                "Additive genetic",
                                "Parental",
                                "Residual")))
```
```{r}
px_modcomps2 <- ggplot(m_vc_comb_plotx%>%
         filter(v_name != "Residual", mod != "1a_base_noIB"),
       aes(trait, v_fit, color = mod)) +
  geom_pointrange(
    aes(ymin = v_lwr, ymax = v_upr),
    position = position_dodge(width = 0.75),
    size=0.15) +
  labs(y = "Fraction variance explained", color = "Model") +
  facet_wrap(~ v_name, ncol = 2, scales = "free_y") +
  scale_x_discrete(
    labels = c("cinf", "cerc", "proc", "fail", "csurv",
               "finf", "fsurv", "pler", "liver", "fgrow", "eggs"
               )
    ) +
  scale_color_manual(values = RColorBrewer::brewer.pal(n=6,name = "Paired")[2:6],
                     labels = c("(1) base",
                                "(2a) base + sibship",
                                "(2b) base + pedigree",
                                "(3a) 2a + dam",
                                "(3a) 2b + dam")) +
  guides(colour = guide_legend(override.aes = list(size = .5))) +
  theme(panel.grid.major.x = element_blank(),
        axis.title.x = element_blank(),
        axis.text.x = element_text(angle = 45, hjust = 1),
        legend.background = element_rect(color="black"),
        legend.justification = c(1,1),
        legend.position = c(0.99,0.42))
px_modcomps2
```

Here is a comparison of heritability before and after removing the fixed effects. The heritability of plerocercoid weight and infected fish growth goes up. Most heritabilities do not change much though.

```{r}
m_vc_comb_vg <- bind_rows(m_vc_comb%>%
  filter(v_name == "Vg")%>%
  mutate(h2 = "with fixed effects"),
m_vc_comb2%>%
  filter(v_name == "Vg2")%>%
  mutate(h2 = "after fixed effects")
)
```
```{r}
ggplot(m_vc_comb_vg, aes(trait, v_fit, color = h2)) +
  geom_pointrange(aes(ymin = v_lwr, ymax = v_upr, color = h2),
                  position = position_dodge(width = 0.6)) +
  labs(y = "Vg/Vp") +
  theme(axis.title.x = element_blank(),
        panel.grid.major.x = element_blank()) +
  facet_wrap(~mod)
```

## Skewness

We also noted that family-effects were skewed for some traits. 

```{r}
# combine RE estimates
re_fams_comb <- bind_rows(
  re_fams_cinf%>%mutate(trait = "cinf"),
  re_fams_proc%>%mutate(trait = "proc"),
  re_fams_cerc%>%mutate(trait = "cerc"),
  re_fams_surv%>%mutate(trait = "csurv"),
  re_fams_fail%>%mutate(trait = "fail"),
  re_fams_finf%>%mutate(trait = "finf"),
  re_fams_fsurv%>%mutate(trait = "fsurv"),
  # re_fams_pler%>%mutate(trait = "pler"),
  re_fams_pler2%>%mutate(trait = "pler"),
  re_fams_liver%>%mutate(trait = "liver"),
  re_fams_fgrow%>%mutate(trait = "fgrow"),
  # re_fams_bred%>%mutate(trait = "eggs")
  re_fams_bred_no_outlier%>%mutate(trait = "eggs")
)

# add gen and line info
re_fams_comb <- left_join(
  select(re_fams_comb, -gen),
  select(dat, line, gen, worm_full_sib_fam_id)%>%distinct()
  )

# arrange variables
re_fams_comb <- re_fams_comb%>%
  mutate(trait = fct_relevel(factor(trait), 
                             levels = c(
                               "cinf", "cerc", "proc", "fail", "csurv",
                               "finf", "fsurv", "pler", "liver", "fgrow",
                               "eggs"
                             )),
         line = fct_relevel(factor(line), levels = c("base", "slow", "control", "fast") ))

```
```{r}
spssSkewKurtosis=function(x) {
  w=length(x)
  m1=mean(x)
  m2=sum((x-m1)^2)
  m3=sum((x-m1)^3)
  m4=sum((x-m1)^4)
  s1=sd(x)
  skew=w*m3/(w-1)/(w-2)/s1^3
  sdskew=sqrt( 6*w*(w-1) / ((w-2)*(w+1)*(w+3)) )
  kurtosis=(w*(w+1)*m4 - 3*m2^2*(w-1)) / ((w-1)*(w-2)*(w-3)*s1^4)
  sdkurtosis=sqrt( 4*(w^2-1) * sdskew^2 / ((w-3)*(w+5)) )
  mat=matrix(c(skew,kurtosis, sdskew,sdkurtosis), 2,
        dimnames=list(c("skew","kurtosis"), c("estimate","se")))
  return(sdskew)
}
```
```{r}
skew_by_trait <- re_fams_comb%>%
  filter(!is.infinite(fam_avg))%>%
  group_by(trait, approach)%>%
  summarise(n = n(),
            var = var(fam_avg, na.rm = T),
            min_x = min(fam_avg, na.rm = T),
            max_y = max(density(fam_avg)$y),
            skew = e1071::skewness(fam_avg, na.rm = T),
            skew_se = spssSkewKurtosis(fam_avg))%>%
  mutate(skew_lwr = skew - 1.96*skew_se, skew_upr = skew + 1.96*skew_se)%>%
  mutate(skew_sig = if_else(skew_lwr > 0 | skew_upr < 0, "*", ""))%>%
  mutate(skew_c = paste0(round(skew, 2), skew_sig))
```

Here is the [skewness](https://en.wikipedia.org/wiki/Skewness) in the raw and the model-estimated family averages (i.e. the random effects). For most traits, the distribution of family averages were more skewed after correcting for starting conditions and shared environments. Usually, the skew was negative, indicating there are more families with low values than high values. This suggests much of the genetic variance in the traits are for lower fitness. Notably, the trait we selected, cercomere presence, did not exhibit skewed family effects, which is consistent with the symmetric response to selection observed.

```{r}
ggplot(skew_by_trait, aes(x = trait, y = skew)) +
  geom_hline(yintercept = 0, linetype = "dashed") +
  geom_line(aes(group=trait), arrow = arrow(length = unit(0.1, "in"), type="open")) +
  geom_point(aes(color = approach, shape = skew_sig)) +
  labs(y = "Skewness") +
  scale_y_continuous(breaks = c(-1.25, -1, -0.75, -0.5, -0.25, 0, 0.25)) +
  scale_shape_manual(values = c(16,8)) +
  guides(shape = "none") +
  theme(axis.title.x = element_blank(),
        panel.grid.major.x = element_blank())
```
Let's plot the model-estimated family means.

```{r}
px_skew <- ggplot(re_fams_comb%>%
         filter(approach == "model-estimated"),
       aes(x = fam_avg)) +
  geom_histogram(aes(y=..density..),
                 colour = "black",
                 fill = "white") +
  geom_density() +
  geom_vline(xintercept = 0, linetype = "dotted", color = "black") +
  geom_label(
    data = skew_by_trait%>%
      filter(approach == "model-estimated"),
    aes(label = skew_c,
        x = min_x,
        y = max_y),
    hjust = 0, vjust = 0.5) +
  facet_wrap(~trait, scales = "free", nrow = 3) +
  labs(y = "Density", x = "Sibship means") +
  theme(
    axis.text = element_blank(),
    axis.ticks = element_blank(),
    panel.grid = element_blank())
px_skew
```

```{r}
ggsave(px_skew, filename = "../figs/figS_skew.png", width = 40*4, height = 40*3, units = "mm")
```

The family averages include families from all three lines. Let's examine skewness in those distributions separately.

```{r}
skew_by_trait_grp <- re_fams_comb%>%
  filter(!is.infinite(fam_avg))%>%
  group_by(trait, approach, line)%>%
  summarise(n = n(),
            var = var(fam_avg, na.rm =T),
            sd = var(fam_avg, na.rm =T),
            min_x = min(fam_avg, na.rm = T),
            max_y = max(density(fam_avg)$y),
            skew = e1071::skewness(fam_avg, na.rm = T),
            skew_se = spssSkewKurtosis(fam_avg))%>%
  mutate(var_lwr = ((n-1)*var/qchisq(0.025, n-1)), var_upr = ((n-1)*var/qchisq(0.975, n-1)),
         skew_lwr = skew - 1.96*skew_se, skew_upr = skew + 1.96*skew_se)%>%
  mutate(skew_sig = if_else(skew_lwr > 0 | skew_upr < 0, "*", ""))%>%
  mutate(skew_c = paste0(round(skew, 2), skew_sig))
```

For fecundity and copepod infection, the families from the selected lines have more negative skew in their distributions. The dashed line is the skew in the combined distribution.

```{r}
ggplot(skew_by_trait_grp%>%
         filter(approach == "model-estimated", line != "base"),
       aes(trait, y = skew, color = line)) +
  # geom_point(position = position_dodge(width = 0.4)) +
  geom_hline(data = skew_by_trait%>%filter(approach == "model-estimated"),
             aes(yintercept = skew),
             linetype = "dashed") +
  geom_pointrange(
    aes(ymin = skew_lwr, ymax = skew_upr),
    position = position_dodge(width = 0.4)) +
  scale_color_manual(values = c("blue", "black", "red")) +
  theme(
    panel.grid.major.x = element_blank(),
    legend.title = element_blank(),
    legend.position = c(0.875, 0.15),
    ) +
  facet_wrap(~trait, scales = "free_x")
```


```{r}
# ggplot(skew_by_trait_grp%>%
#          filter(approach == "model-estimated", line != "base"),
#        aes(trait, y = skew, color = line)) +
#   geom_pointrange(
#     aes(ymin = skew_lwr, ymax = skew_upr),
#     position = position_dodge(width = 0.4)) +
#   scale_color_manual(values = c("blue", "black", "red")) +
#   theme(panel.grid.major.x = element_blank())
```

It is probably easier to see this by examining the distributions directly.

```{r}
px_sibXlin_disA <- ggplot(re_fams_comb%>%
         filter(approach == "model-estimated",
                # trait %in% c("cinf", "fail", "fsurv", "eggs"),
                gen>0),
       aes(x = fam_avg, color = line, fill = line)) +
  geom_density(alpha = 0.1, adjust = 1) +
  geom_vline(xintercept = 0, linetype = "dotted", color = "black") +
  facet_wrap(~trait, scales = "free", nrow = 3) +
  scale_color_manual(values = c("blue", "black", "red")) +
  scale_fill_manual(values = c("blue", "black", "red")) +
  labs(y = "Density", x = "Sibship means") +
  theme(
    axis.text = element_blank(),
    axis.ticks = element_blank(),
    legend.title = element_blank(),
    legend.position = c(0.875, 0.15),
    panel.grid = element_blank())
px_sibXlin_disA
```

We can also look at conditional densitity plots, i.e. what percent of the points along the x-axis belong to a given line. However, this removes all information about density. 

```{r}
ggplot(re_fams_comb%>%
         filter(approach == "model-estimated",
                # trait %in% c("cinf", "fail", "fsurv", "eggs"),
                gen>0),
       aes(x = fam_avg, y = after_stat(count), color = line, fill = line)) +
  geom_density(position = "fill", alpha = 0.2, adjust = 1/2) +
  geom_vline(xintercept = 0, linetype = "dotted", color = "black") +
  facet_wrap(~trait, scales = "free_x", nrow = 3) +
  scale_color_manual(values = c("blue", "black", "red")) +
  scale_fill_manual(values = c("blue", "black", "red")) +
  labs(y = "Density", x = "Family averages") +
  theme(
    # axis.text = element_blank(),
    axis.ticks = element_blank(),
    panel.grid = element_blank())
```

Simpler than skew, we can examine the variance in family means in each lines. That is, are variances higher in the selected lines because they are skewed? To test this, I fit models in which the sibship-level variance was estimated for each selection line separately.

```{r}
vc_ci <- function(model, term){
  tx <- summary(model)
  ci_df <- tx$Gcovariances[grepl(term, row.names(tx$Gcovariances)), 1:3]
  if(length(ci_df) == 3){
    ci_df <- data.frame(ci_df[1], ci_df[2], ci_df[3])
  } else {
    ci_df <- data.frame(ci_df)
  }
  names(ci_df) <- c("var", "var_lwr", "var_upr")
  return(ci_df)
}

# combine RE estimates
vc_fams_comb <- bind_rows(
  vc_ci(m2_cinf_fam, "worm_full_sib_fam_id")%>%mutate(trait = "cinf", line = "all"),
  vc_ci(m2_cerc_fam, "worm_full_sib_fam_id")%>%mutate(trait = "cerc", line = "all"),
  vc_ci(m2_proc_fam, "worm_full_sib_fam_id")%>%mutate(trait = "proc", line = "all"),
  vc_ci(m2_fail_fam, "worm_full_sib_fam_id")%>%mutate(trait = "fail", line = "all"),
  vc_ci(m2_surv_fam, "worm_full_sib_fam_id")%>%mutate(trait = "csurv", line = "all"),
  vc_ci(m2_finf_fam, "worm_full_sib_fam_id")%>%mutate(trait = "finf", line = "all"),
  vc_ci(m2_fsurv_fam, "worm_full_sib_fam_id")%>%mutate(trait = "fsurv", line = "all"),
  vc_ci(m2_pler_fam, "worm_full_sib_fam_id")%>%mutate(trait = "pler", line = "all"),
  vc_ci(m2_liver_fam, "worm_full_sib_fam_id")%>%mutate(trait = "liver", line = "all"),
  vc_ci(m2_fgrow_fam, "worm_full_sib_fam_id")%>%mutate(trait = "fgrow", line = "all"),
  vc_ci(m2_bred_fam, "worm_full_sib_fam_id")%>%mutate(trait = "eggs", line = "all")
)

vc_fams_comb_by_line <- bind_rows(
  vc_ci(m2_cinf_famXline, "line")%>%mutate(trait = "cinf"),
  vc_ci(m2_cerc_famXline, "line")%>%mutate(trait = "cerc"),
  vc_ci(m2_proc_famXline, "line")%>%mutate(trait = "proc"),
  vc_ci(m2_fail_famXline, "line")%>%mutate(trait = "fail"),
  vc_ci(m2_surv_famXline, "line")%>%mutate(trait = "csurv"),
  vc_ci(m2_finf_famXline, "line")%>%mutate(trait = "finf"),
  vc_ci(m2_fsurv_famXline, "line")%>%mutate(trait = "fsurv"),
  vc_ci(m2_pler_famXline, "line")%>%mutate(trait = "pler"),
  vc_ci(m2_liver_famXline, "line")%>%mutate(trait = "liver"),
  vc_ci(m2_fgrow_famXline, "line")%>%mutate(trait = "fgrow"),
  vc_ci(m2_bred_famXline, "line")%>%mutate(trait = "eggs")
)

vc_fams_comb_by_line <- vc_fams_comb_by_line%>%
  mutate(line = gsub(pattern = "line", replacement = "", x = row.names(vc_fams_comb_by_line)))%>%
  mutate(line = gsub(pattern = ".worm_full_sib_fam_id.*$", replacement = "", x = line))

vc_fams_comb <- bind_rows(vc_fams_comb, vc_fams_comb_by_line)
vc_fams_comb <- vc_fams_comb%>%
  mutate(
    trait = fct_relevel(factor(trait),
                             levels = c(
                               "cinf", "cerc", "proc", "fail", "csurv",
                               "finf", "fsurv", "pler", "liver", "fgrow",
                               "eggs"
                             )),
         line = fct_relevel(factor(line), levels = c("base", "slow", "control", "fast") ))
```

This plot shows the estimated variances among sibships for the 3 selection lines. In 3 of the 4 skewed traits (copepod infection, failed development, fecundity), there was more variation in one or both of the selected lines. And in some of the other traits,like plerocercoid weight, there was also more variance in the selected lines.

```{r}
px_sibXlin_disB <- ggplot(vc_fams_comb%>%
         filter(line %in% c("control", "fast", "slow")),
       aes(trait, y = var, color = line)) +
  # geom_ribbon(
  #   data = vc_fams_comb%>%filter(line == "all")%>%select(-line),
  #   aes(ymin = var_lwr, ymax = var_upr),
  #   color = NA, fill = "gray", alpha = 0.2) +
  geom_hline(
    data = vc_fams_comb%>%filter(line == "all")%>%select(-line),
    aes(yintercept = var),
    color = "black", linetype = "dashed") +
  geom_hline(
    data = vc_fams_comb%>%filter(line == "all")%>%select(-line),
    aes(yintercept = var_upr),
    color = "black", linetype = "dotted") +
  geom_hline(
    data = vc_fams_comb%>%filter(line == "all")%>%select(-line),
    aes(yintercept = var_lwr),
    color = "black", linetype = "dotted") +
  geom_pointrange(
    aes(ymin = var_lwr, ymax = var_upr),
    position = position_dodge(width = 0.4)) +
  labs(y = "Among-sibship variance") + 
  scale_x_discrete(expand = expansion(mult = c(0.33,0.33))) +
  scale_color_manual(values = c("blue", "black", "red")) +
  theme(
    axis.title.x = element_blank(),
    axis.text.x = element_blank(),
    axis.ticks = element_blank(),
    legend.title = element_blank(),
    legend.position = c(0.875, 0.15),
    # legend.justification = c(1, 0.1),
    panel.grid.major.x = element_blank()) +
  facet_wrap(~trait, scales = "free")
px_sibXlin_disB
```

```{r}
px_skew2 <- cowplot::plot_grid(
  px_sibXlin_disA,
  px_sibXlin_disB,
  labels = "AUTO",
  align = "h"
)
# px_skew2
ggsave(px_skew2, filename = "../figs/figS_skew_byline.png", width = 80*4, height = 40*3, units = "mm")
```

So, these trends are consistent with the idea that directional selection on larval developmental rate releases cryptic genetic variation and that most of this variation is detrimental.

```{r}
save.image(file = "04quant_gen_univariate.RData")
```

```{r}
sessionInfo()
```

